{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-10-02T07:55:55.284130Z",
     "start_time": "2023-10-02T07:55:53.625394Z"
    }
   },
   "outputs": [],
   "source": [
    "# Importing the libraries\n",
    "from functions import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## HMM: Data preparation\n",
    "In this part we generate HMM models based on MSA and retrieve the results from hmmsearch and then from Pfam."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set the maximum width of the columns\n",
    "pd.set_option('display.max_colwidth', 20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of rows with the disordered regions: 7393\n"
     ]
    }
   ],
   "source": [
    "# Load the disordered dataframe\n",
    "disordered = pd.read_csv('disordered_df.csv')\n",
    "curated_disprot = pd.read_csv('curated_disprot.csv')\n",
    "\n",
    "print(f'The number of rows with the disordered regions: {len(disordered)}')\n",
    "# disordered.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of the proteins with the disordered regions: 39\n",
      "The number of the disordered regions in the proteins: 53\n"
     ]
    }
   ],
   "source": [
    "# Collect the disordered regions in a dictionary\n",
    "dis_regs = set()\n",
    "\n",
    "for i, row in disordered.iterrows():\n",
    "    dis_id = row[0]\n",
    "    matching_row = curated_disprot[curated_disprot['acc'] == dis_id]\n",
    "    if not matching_row.empty:\n",
    "        region = matching_row['region']\n",
    "        dis_regs.update(region)\n",
    "        dis_regs_list = list(dis_regs) # convert set to a list\n",
    "        \n",
    "# Define an array of disordered regions ids\n",
    "disprot_ids = disordered['query_id'].unique()\n",
    "\n",
    "# print('The disordered regions:', dis_regs_list)\n",
    "print(f'The number of the proteins with the disordered regions: {len(disprot_ids)}')\n",
    "print(f'The number of the disordered regions in the proteins: {len(dis_regs_list)}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The results above show how many proteins with the disordered regions are in the provided data, as well as how many disordered regions do they contain."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-02T13:38:09.492889Z",
     "start_time": "2023-08-02T13:38:08.673895Z"
    }
   },
   "source": [
    "## 1. hmmbuild\n",
    "We build an HMM of each disordered region, using trimmed MSA as an input. We will build the models separately for BLAST and ClustalOmega to be able to compare the quality of the models based on the local and the global alignment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set the paths to HMM and MSA files - BLAST\n",
    "hmm_dir_blast = f'{directory}/results/hmms/hmmbuild/blast'\n",
    "align_dir_blast = f'{directory}/results/alignments/output_files/disordered/blast'\n",
    "\n",
    "# Set the paths to HMM and MSA files - ClustalOmega\n",
    "hmm_dir_clustal = f'{directory}/results/hmms/hmmbuild/clustal'\n",
    "align_dir_clustal = f'{directory}/results/alignments/output_files/disordered/clustal'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/S7W634_1-32.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/S7W634_1-32.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     S7W634_1-32            200    31    31    97.18  1.740 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.03\n",
      "hmmbuild completed for S7W634_1-32.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q8R464_25-120.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q8R464_25-120.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q8R464_25-120          200    95    95     1.05  0.602 \n",
      "\n",
      "# CPU time: 0.04u 0.00s 00:00:00.04 Elapsed: 00:00:00.03\n",
      "hmmbuild completed for Q8R464_25-120.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/P01097_28-50.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/P01097_28-50.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P01097_28-50           200    22    22    84.30  2.408 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for P01097_28-50.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q9Z2F5_306-430.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q9Z2F5_306-430.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q9Z2F5_306-430         200   124   124     0.41  0.586 \n",
      "\n",
      "# CPU time: 0.04u 0.00s 00:00:00.04 Elapsed: 00:00:00.04\n",
      "hmmbuild completed for Q9Z2F5_306-430.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/P01019_436-450.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/P01019_436-450.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P01019_436-450         200    14    14   200.00  2.756 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for P01019_436-450.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q8K4J6_67-98.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q8K4J6_67-98.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q8K4J6_67-98           200    31    31     4.57  1.740 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q8K4J6_67-98.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/A8AZZ3_116-134.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/A8AZZ3_116-134.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     A8AZZ3_116-134         106    18    18   106.00  2.337 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for A8AZZ3_116-134.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/A8AZZ3_146-195.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/A8AZZ3_146-195.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     A8AZZ3_146-195         106    49    49     7.76  1.128 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for A8AZZ3_146-195.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/P00736_193-305.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/P00736_193-305.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P00736_193-305         200   112   112     1.10  0.591 \n",
      "\n",
      "# CPU time: 0.04u 0.00s 00:00:00.04 Elapsed: 00:00:00.04\n",
      "hmmbuild completed for P00736_193-305.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/O35274_1-154.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/O35274_1-154.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O35274_1-154           200   153   153     0.88  0.590 \n",
      "\n",
      "# CPU time: 0.05u 0.00s 00:00:00.05 Elapsed: 00:00:00.05\n",
      "hmmbuild completed for O35274_1-154.fasta\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q9H832_1-99.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q9H832_1-99.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q9H832_1-99            200    98    98     2.94  0.590 \n",
      "\n",
      "# CPU time: 0.04u 0.00s 00:00:00.04 Elapsed: 00:00:00.04\n",
      "hmmbuild completed for Q9H832_1-99.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/O00585_71-111.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/O00585_71-111.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O00585_71-111          200    40    40     5.01  1.367 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for O00585_71-111.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/J8TM36_236-249.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/J8TM36_236-249.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     J8TM36_236-249         200    13    13   200.00  0.820 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for J8TM36_236-249.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q86FP8_20-80.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q86FP8_20-80.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q86FP8_20-80            14    60    60     3.05  0.930 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for Q86FP8_20-80.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/O14727_1-104.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/O14727_1-104.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O14727_1-104           200   103   103     0.55  0.592 \n",
      "\n",
      "# CPU time: 0.04u 0.00s 00:00:00.04 Elapsed: 00:00:00.04\n",
      "hmmbuild completed for O14727_1-104.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q93KQ4_51-81.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q93KQ4_51-81.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q93KQ4_51-81            73    30    30    73.00  1.706 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for Q93KQ4_51-81.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/O15922_1-89.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/O15922_1-89.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O15922_1-89            200    88    88     6.37  0.647 \n",
      "\n",
      "# CPU time: 0.03u 0.00s 00:00:00.03 Elapsed: 00:00:00.03\n",
      "hmmbuild completed for O15922_1-89.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q84852_17-299.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q84852_17-299.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q84852_17-299          200   282   282     1.52  0.591 \n",
      "\n",
      "# CPU time: 0.10u 0.00s 00:00:00.10 Elapsed: 00:00:00.10\n",
      "hmmbuild completed for Q84852_17-299.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q8K4J6_111-142.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q8K4J6_111-142.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q8K4J6_111-142         200    31    31     4.39  1.740 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q8K4J6_111-142.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q9H832_327-354.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q9H832_327-354.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q9H832_327-354         200    27    27     5.36  1.984 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q9H832_327-354.fasta\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q8K4J6_155-186.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q8K4J6_155-186.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q8K4J6_155-186         200    31    31     4.06  1.741 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q8K4J6_155-186.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/A1L1Q4_1-71.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/A1L1Q4_1-71.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     A1L1Q4_1-71            200    70    70     2.32  0.804 \n",
      "\n",
      "# CPU time: 0.03u 0.00s 00:00:00.03 Elapsed: 00:00:00.03\n",
      "hmmbuild completed for A1L1Q4_1-71.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/A4L7I2_1658-1856.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/A4L7I2_1658-1856.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     A4L7I2_1658-1856       200   198   194     0.44  0.593 \n",
      "\n",
      "# CPU time: 0.07u 0.00s 00:00:00.07 Elapsed: 00:00:00.07\n",
      "hmmbuild completed for A4L7I2_1658-1856.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/P04370-5_1-169.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/P04370-5_1-169.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P04370-5_1-169         201   168   168     0.46  0.589 \n",
      "\n",
      "# CPU time: 0.06u 0.00s 00:00:00.06 Elapsed: 00:00:00.06\n",
      "hmmbuild completed for P04370-5_1-169.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q8IU57_270-299.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q8IU57_270-299.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q8IU57_270-299         200    29    29    10.48  1.854 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q8IU57_270-299.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/O88339_144-575.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/O88339_144-575.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O88339_144-575         200   431   431     0.50  0.591 \n",
      "\n",
      "# CPU time: 0.15u 0.00s 00:00:00.15 Elapsed: 00:00:00.15\n",
      "hmmbuild completed for O88339_144-575.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q5VZK9_1005-1020.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q5VZK9_1005-1020.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q5VZK9_1005-1020       200    15    15    21.91  3.460 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q5VZK9_1005-1020.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q8WUG5_1-105.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q8WUG5_1-105.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q8WUG5_1-105           200   104   104     0.41  0.586 \n",
      "\n",
      "# CPU time: 0.04u 0.00s 00:00:00.04 Elapsed: 00:00:00.04\n",
      "hmmbuild completed for Q8WUG5_1-105.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/O00308_662-702.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/O00308_662-702.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O00308_662-702         200    40    40     1.03  1.370 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for O00308_662-702.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q8IW19_399-420.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q8IW19_399-420.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q8IW19_399-420         200    21    21     6.59  2.517 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q8IW19_399-420.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/O15922_230-240.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/O15922_230-240.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O15922_230-240         200    10    10   200.00  1.018 \n",
      "\n",
      "# CPU time: 0.00u 0.00s 00:00:00.00 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for O15922_230-240.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/O43474_1-130.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/O43474_1-130.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O43474_1-130           200   129   128     1.11  0.590 \n",
      "\n",
      "# CPU time: 0.05u 0.00s 00:00:00.05 Elapsed: 00:00:00.04\n",
      "hmmbuild completed for O43474_1-130.fasta\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/S6B291_240-465.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/S6B291_240-465.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     S6B291_240-465         200   225   225     1.19  0.589 \n",
      "\n",
      "# CPU time: 0.08u 0.00s 00:00:00.08 Elapsed: 00:00:00.08\n",
      "hmmbuild completed for S6B291_240-465.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/O35274_417-494.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/O35274_417-494.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O35274_417-494         200    77    77     0.75  0.734 \n",
      "\n",
      "# CPU time: 0.03u 0.00s 00:00:00.03 Elapsed: 00:00:00.03\n",
      "hmmbuild completed for O35274_417-494.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q9H0E2_1-53.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q9H0E2_1-53.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q9H0E2_1-53            200    52    52     2.37  1.067 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for Q9H0E2_1-53.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/P07342_647-687.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/P07342_647-687.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P07342_647-687         200    40    40     4.13  1.367 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for P07342_647-687.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/O43791_357-374.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/O43791_357-374.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O43791_357-374         200    17    17     3.01  3.075 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for O43791_357-374.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/A8AZZ3_24-44.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/A8AZZ3_24-44.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     A8AZZ3_24-44           106    20    20   106.00  1.574 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for A8AZZ3_24-44.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/P00742_90-126.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/P00742_90-126.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P00742_90-126          200    36    36     3.20  1.510 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for P00742_90-126.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/P01019_44-64.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/P01019_44-64.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P01019_44-64           200    20    20    62.77  2.636 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for P01019_44-64.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q99967_220-269.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q99967_220-269.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q99967_220-269         200    49    49     0.83  1.126 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for Q99967_220-269.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q9BYF1_769-805.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q9BYF1_769-805.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q9BYF1_769-805         200    36    36     3.49  1.511 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q9BYF1_769-805.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q8IW19_450-511.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q8IW19_450-511.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q8IW19_450-511         200    61    61     1.28  0.915 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for Q8IW19_450-511.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/O15922_315-324.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/O15922_315-324.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O15922_315-324         200     9     9   200.00  1.125 \n",
      "\n",
      "# CPU time: 0.00u 0.00s 00:00:00.00 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for O15922_315-324.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/O35274_584-602.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/O35274_584-602.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O35274_584-602         200    18    18     3.37  2.911 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for O35274_584-602.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q9CXY6_29-44.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q9CXY6_29-44.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q9CXY6_29-44           200    15    15     3.72  3.461 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q9CXY6_29-44.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/P07342_580-595.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/P07342_580-595.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P07342_580-595         200    15    15   200.00  3.318 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.00\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hmmbuild completed for P07342_580-595.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q9BYI3_149-253.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q9BYI3_149-253.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q9BYI3_149-253         200   104   104     0.41  0.589 \n",
      "\n",
      "# CPU time: 0.04u 0.00s 00:00:00.04 Elapsed: 00:00:00.04\n",
      "hmmbuild completed for Q9BYI3_149-253.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/P00392_550-561.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/P00392_550-561.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P00392_550-561         200    11    11   200.00  4.187 \n",
      "\n",
      "# CPU time: 0.00u 0.00s 00:00:00.00 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for P00392_550-561.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q9CXY6_347-390.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q9CXY6_347-390.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q9CXY6_347-390         200    43    43     1.10  1.275 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for Q9CXY6_347-390.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/O43791_169-178.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/O43791_169-178.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O43791_169-178         200     9     9   200.00  4.418 \n",
      "\n",
      "# CPU time: 0.00u 0.00s 00:00:00.00 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for O43791_169-178.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q5T4W7_108-120.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q5T4W7_108-120.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q5T4W7_108-120         200    12    12   200.00  1.367 \n",
      "\n",
      "# CPU time: 0.00u 0.00s 00:00:00.00 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for Q5T4W7_108-120.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/O14958_371-399.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/O14958_371-399.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O14958_371-399         200    28    28     2.79  1.917 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for O14958_371-399.fasta\n"
     ]
    }
   ],
   "source": [
    "# Build HMM for BLAST\n",
    "for filename in os.listdir(align_dir_blast):\n",
    "    if filename.endswith('.fasta'):\n",
    "        input_file = os.path.join(align_dir_blast, filename)\n",
    "        output_file = os.path.join(hmm_dir_blast, os.path.splitext(filename)[0] + '.hmm')\n",
    "\n",
    "        subprocess.run(['hmmbuild', output_file, input_file])\n",
    "        print(f'hmmbuild completed for {filename}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/S7W634_1-32.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/S7W634_1-32.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     S7W634_1-32            196    31    30   196.00  1.348 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for S7W634_1-32.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q8R464_25-120.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q8R464_25-120.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q8R464_25-120          194    95    95     0.41  0.598 \n",
      "\n",
      "# CPU time: 0.04u 0.00s 00:00:00.04 Elapsed: 00:00:00.03\n",
      "hmmbuild completed for Q8R464_25-120.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/P01097_28-50.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/P01097_28-50.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P01097_28-50           195    22    22    51.76  2.408 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for P01097_28-50.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q9Z2F5_306-430.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q9Z2F5_306-430.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q9Z2F5_306-430         192   124   124     0.41  0.588 \n",
      "\n",
      "# CPU time: 0.04u 0.00s 00:00:00.04 Elapsed: 00:00:00.04\n",
      "hmmbuild completed for Q9Z2F5_306-430.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/P01019_436-450.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/P01019_436-450.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P01019_436-450         192    14    13   192.00  1.794 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for P01019_436-450.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q8K4J6_67-98.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q8K4J6_67-98.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q8K4J6_67-98           194    31    31     2.54  1.741 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q8K4J6_67-98.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/A8AZZ3_116-134.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/A8AZZ3_116-134.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     A8AZZ3_116-134         102    18    18   102.00  1.994 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for A8AZZ3_116-134.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/A8AZZ3_146-195.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/A8AZZ3_146-195.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     A8AZZ3_146-195         102    49    45    13.26  1.222 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for A8AZZ3_146-195.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/P00736_193-305.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/P00736_193-305.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P00736_193-305         190   112   112     0.43  0.589 \n",
      "\n",
      "# CPU time: 0.04u 0.00s 00:00:00.04 Elapsed: 00:00:00.04\n",
      "hmmbuild completed for P00736_193-305.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/O35274_1-154.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/O35274_1-154.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O35274_1-154           193   153   152     2.09  0.591 \n",
      "\n",
      "# CPU time: 0.05u 0.00s 00:00:00.05 Elapsed: 00:00:00.05\n",
      "hmmbuild completed for O35274_1-154.fasta\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q9H832_1-99.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q9H832_1-99.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q9H832_1-99            190    98    91     7.24  0.627 \n",
      "\n",
      "# CPU time: 0.04u 0.00s 00:00:00.04 Elapsed: 00:00:00.03\n",
      "hmmbuild completed for Q9H832_1-99.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/O00585_71-111.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/O00585_71-111.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O00585_71-111          191    40    40     6.73  1.367 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for O00585_71-111.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/J8TM36_236-249.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/J8TM36_236-249.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     J8TM36_236-249         186    13    13   186.00  0.706 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for J8TM36_236-249.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q86FP8_20-80.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q86FP8_20-80.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q86FP8_20-80            13    60    60     5.22  0.930 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for Q86FP8_20-80.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/O14727_1-104.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/O14727_1-104.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O14727_1-104           193   103   103     0.56  0.590 \n",
      "\n",
      "# CPU time: 0.04u 0.00s 00:00:00.04 Elapsed: 00:00:00.04\n",
      "hmmbuild completed for O14727_1-104.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q93KQ4_51-81.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q93KQ4_51-81.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q93KQ4_51-81            62    30    30    62.00  1.764 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for Q93KQ4_51-81.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/O15922_1-89.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/O15922_1-89.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O15922_1-89            181    88    71    46.21  0.793 \n",
      "\n",
      "# CPU time: 0.03u 0.00s 00:00:00.03 Elapsed: 00:00:00.03\n",
      "hmmbuild completed for O15922_1-89.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q84852_17-299.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q84852_17-299.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q84852_17-299          200   282   279     0.78  0.590 \n",
      "\n",
      "# CPU time: 0.10u 0.00s 00:00:00.10 Elapsed: 00:00:00.10\n",
      "hmmbuild completed for Q84852_17-299.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q8K4J6_111-142.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q8K4J6_111-142.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q8K4J6_111-142         194    31    31     2.38  1.740 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q8K4J6_111-142.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q9H832_327-354.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q9H832_327-354.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q9H832_327-354         190    27    27     7.91  1.984 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q9H832_327-354.fasta\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q8K4J6_155-186.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q8K4J6_155-186.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q8K4J6_155-186         194    31    31     1.68  1.738 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q8K4J6_155-186.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/A1L1Q4_1-71.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/A1L1Q4_1-71.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     A1L1Q4_1-71            191    70    70     1.63  0.804 \n",
      "\n",
      "# CPU time: 0.03u 0.00s 00:00:00.03 Elapsed: 00:00:00.03\n",
      "hmmbuild completed for A1L1Q4_1-71.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/A4L7I2_1658-1856.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/A4L7I2_1658-1856.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     A4L7I2_1658-1856       200   198   194     0.44  0.593 \n",
      "\n",
      "# CPU time: 0.08u 0.00s 00:00:00.08 Elapsed: 00:00:00.08\n",
      "hmmbuild completed for A4L7I2_1658-1856.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/P04370-5_1-169.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/P04370-5_1-169.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P04370-5_1-169         192   168   166     0.52  0.587 \n",
      "\n",
      "# CPU time: 0.06u 0.00s 00:00:00.06 Elapsed: 00:00:00.06\n",
      "hmmbuild completed for P04370-5_1-169.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q8IU57_270-299.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q8IU57_270-299.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q8IU57_270-299         190    29    29    11.29  1.854 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q8IU57_270-299.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/O88339_144-575.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/O88339_144-575.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O88339_144-575         190   431   431     0.54  0.592 \n",
      "\n",
      "# CPU time: 0.15u 0.00s 00:00:00.15 Elapsed: 00:00:00.15\n",
      "hmmbuild completed for O88339_144-575.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q5VZK9_1005-1020.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q5VZK9_1005-1020.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q5VZK9_1005-1020       195    15    15    21.97  3.460 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q5VZK9_1005-1020.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q8WUG5_1-105.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q8WUG5_1-105.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q8WUG5_1-105           189   104   103     0.46  0.589 \n",
      "\n",
      "# CPU time: 0.04u 0.00s 00:00:00.04 Elapsed: 00:00:00.04\n",
      "hmmbuild completed for Q8WUG5_1-105.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/O00308_662-702.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/O00308_662-702.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O00308_662-702         191    40    40     1.03  1.370 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for O00308_662-702.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q8IW19_399-420.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q8IW19_399-420.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q8IW19_399-420         190    21    21     8.17  2.517 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q8IW19_399-420.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/O15922_230-240.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/O15922_230-240.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O15922_230-240         181    10    10   181.00  2.008 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for O15922_230-240.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/O43474_1-130.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/O43474_1-130.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O43474_1-130           188   129   128     0.59  0.589 \n",
      "\n",
      "# CPU time: 0.05u 0.00s 00:00:00.05 Elapsed: 00:00:00.04\n",
      "hmmbuild completed for O43474_1-130.fasta\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/S6B291_240-465.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/S6B291_240-465.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     S6B291_240-465         193   225   220     0.64  0.590 \n",
      "\n",
      "# CPU time: 0.08u 0.00s 00:00:00.08 Elapsed: 00:00:00.08\n",
      "hmmbuild completed for S6B291_240-465.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/O35274_417-494.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/O35274_417-494.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O35274_417-494         193    77    77     0.75  0.735 \n",
      "\n",
      "# CPU time: 0.03u 0.00s 00:00:00.03 Elapsed: 00:00:00.03\n",
      "hmmbuild completed for O35274_417-494.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q9H0E2_1-53.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q9H0E2_1-53.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q9H0E2_1-53            195    52    52     3.32  1.066 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for Q9H0E2_1-53.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/P07342_647-687.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/P07342_647-687.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P07342_647-687         194    40    40     3.34  1.367 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for P07342_647-687.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/O43791_357-374.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/O43791_357-374.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O43791_357-374         193    17    17     3.01  3.073 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for O43791_357-374.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/A8AZZ3_24-44.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/A8AZZ3_24-44.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     A8AZZ3_24-44           102    20    18   102.00  1.863 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for A8AZZ3_24-44.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/P00742_90-126.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/P00742_90-126.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P00742_90-126          192    36    36     1.49  1.511 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for P00742_90-126.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/P01019_44-64.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/P01019_44-64.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P01019_44-64           192    20    20   192.00  2.609 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for P01019_44-64.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q99967_220-269.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q99967_220-269.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q99967_220-269         193    49    49     0.82  1.130 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for Q99967_220-269.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q9BYF1_769-805.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q9BYF1_769-805.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q9BYF1_769-805         196    36    36     5.47  1.510 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q9BYF1_769-805.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q8IW19_450-511.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q8IW19_450-511.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q8IW19_450-511         190    61    61     7.08  0.916 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for Q8IW19_450-511.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/O15922_315-324.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/O15922_315-324.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O15922_315-324         181     9     5   181.00  1.387 \n",
      "\n",
      "# CPU time: 0.00u 0.00s 00:00:00.00 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for O15922_315-324.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/O35274_584-602.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/O35274_584-602.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O35274_584-602         193    18    18     3.37  2.912 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for O35274_584-602.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q9CXY6_29-44.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q9CXY6_29-44.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q9CXY6_29-44           195    15    15     3.72  3.460 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q9CXY6_29-44.fasta\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/P07342_580-595.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/P07342_580-595.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P07342_580-595         194    15    15     5.31  3.461 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for P07342_580-595.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q9BYI3_149-253.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q9BYI3_149-253.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q9BYI3_149-253         192   104   104     0.41  0.592 \n",
      "\n",
      "# CPU time: 0.04u 0.00s 00:00:00.04 Elapsed: 00:00:00.04\n",
      "hmmbuild completed for Q9BYI3_149-253.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/P00392_550-561.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/P00392_550-561.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P00392_550-561         185    11    11   185.00  4.132 \n",
      "\n",
      "# CPU time: 0.00u 0.00s 00:00:00.00 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for P00392_550-561.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q9CXY6_347-390.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q9CXY6_347-390.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q9CXY6_347-390         195    43    43     1.16  1.276 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for Q9CXY6_347-390.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/O43791_169-178.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/O43791_169-178.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O43791_169-178         193     9     9   193.00  4.418 \n",
      "\n",
      "# CPU time: 0.00u 0.00s 00:00:00.00 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for O43791_169-178.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q5T4W7_108-120.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q5T4W7_108-120.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q5T4W7_108-120         190    12    11   190.00  1.521 \n",
      "\n",
      "# CPU time: 0.00u 0.00s 00:00:00.00 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for Q5T4W7_108-120.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/O14958_371-399.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/O14958_371-399.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O14958_371-399         194    28    28    34.42  1.917 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for O14958_371-399.fasta\n"
     ]
    }
   ],
   "source": [
    "# Build HMM for ClustalOmega\n",
    "for filename in os.listdir(align_dir_clustal):\n",
    "    if filename.endswith('.fasta'):\n",
    "        input_file = os.path.join(align_dir_clustal, filename)\n",
    "        output_file = os.path.join(hmm_dir_clustal, os.path.splitext(filename)[0] + '.hmm')\n",
    "\n",
    "        subprocess.run(['hmmbuild', output_file, input_file])\n",
    "        print(f'hmmbuild completed for {filename}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. hmmsearch\n",
    "\n",
    "After building the models, our objective is to assess if overlaps with the profiles in the sequence database (Reference Proteome 75%) exist. We retrive the data containing the most significant sequences, with a default E-value threshold of 0.01.\n",
    "\n",
    "The same as before, we do it for BLAST and ClustalOmega separately."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File O43791_169-178.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/O43791_169-178.hmm\n",
      "File P04370-5_1-169.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/P04370-5_1-169.hmm\n",
      "File Q9H832_327-354.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/Q9H832_327-354.hmm\n",
      "File O35274_584-602.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/O35274_584-602.hmm\n",
      "File Q8IW19_399-420.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/Q8IW19_399-420.hmm\n",
      "File Q8R464_25-120.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/Q8R464_25-120.hmm\n",
      "File Q93KQ4_51-81.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/Q93KQ4_51-81.hmm\n",
      "File P01019_436-450.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/P01019_436-450.hmm\n",
      "File S7W634_1-32.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/S7W634_1-32.hmm\n",
      "File Q99967_220-269.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/Q99967_220-269.hmm\n",
      "File Q5T4W7_108-120.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/Q5T4W7_108-120.hmm\n",
      "File A4L7I2_1658-1856.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/A4L7I2_1658-1856.hmm\n",
      "File Q5VZK9_1005-1020.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/Q5VZK9_1005-1020.hmm\n",
      "File Q9CXY6_29-44.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/Q9CXY6_29-44.hmm\n",
      "File O35274_417-494.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/O35274_417-494.hmm\n",
      "File A8AZZ3_116-134.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/A8AZZ3_116-134.hmm\n",
      "File Q9CXY6_347-390.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/Q9CXY6_347-390.hmm\n",
      "File P00736_193-305.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/P00736_193-305.hmm\n",
      "File Q9Z2F5_306-430.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/Q9Z2F5_306-430.hmm\n",
      "File Q8K4J6_111-142.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/Q8K4J6_111-142.hmm\n",
      "File Q84852_17-299.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/Q84852_17-299.hmm\n",
      "File P01097_28-50.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/P01097_28-50.hmm\n",
      "File O43474_1-130.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/O43474_1-130.hmm\n",
      "File A1L1Q4_1-71.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/A1L1Q4_1-71.hmm\n",
      "File O43791_357-374.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/O43791_357-374.hmm\n",
      "File Q8K4J6_155-186.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/Q8K4J6_155-186.hmm\n",
      "File O35274_1-154.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/O35274_1-154.hmm\n",
      "File Q9BYI3_149-253.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/Q9BYI3_149-253.hmm\n",
      "File Q8WUG5_1-105.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/Q8WUG5_1-105.hmm\n",
      "File Q86FP8_20-80.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/Q86FP8_20-80.hmm\n",
      "File P00392_550-561.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/P00392_550-561.hmm\n",
      "File Q9H832_1-99.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/Q9H832_1-99.hmm\n",
      "File J8TM36_236-249.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/J8TM36_236-249.hmm\n",
      "File O88339_144-575.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/O88339_144-575.hmm\n",
      "File O15922_1-89.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/O15922_1-89.hmm\n",
      "File O00585_71-111.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/O00585_71-111.hmm\n",
      "File Q8K4J6_67-98.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/Q8K4J6_67-98.hmm\n",
      "File Q8IW19_450-511.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/Q8IW19_450-511.hmm\n",
      "File S6B291_240-465.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/S6B291_240-465.hmm\n",
      "File O14958_371-399.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/O14958_371-399.hmm\n",
      "File P07342_580-595.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/P07342_580-595.hmm\n",
      "File P01019_44-64.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/P01019_44-64.hmm\n",
      "File O15922_315-324.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/O15922_315-324.hmm\n",
      "File A8AZZ3_24-44.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/A8AZZ3_24-44.hmm\n",
      "File A8AZZ3_146-195.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/A8AZZ3_146-195.hmm\n",
      "File Q9H0E2_1-53.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/Q9H0E2_1-53.hmm\n",
      "File Q9BYF1_769-805.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/Q9BYF1_769-805.hmm\n",
      "File O14727_1-104.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/O14727_1-104.hmm\n",
      "File Q8IU57_270-299.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/Q8IU57_270-299.hmm\n",
      "File P07342_647-687.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/P07342_647-687.hmm\n",
      "File O00308_662-702.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/O00308_662-702.hmm\n",
      "File O15922_230-240.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/O15922_230-240.hmm\n",
      "File P00742_90-126.hmm copied to alina@ecate:~/hmm/blast/hmmbuild/P00742_90-126.hmm\n"
     ]
    }
   ],
   "source": [
    "# Copy the HMMs from the BLAST folder\n",
    "for filename in os.listdir(hmm_dir_blast):\n",
    "    source_path = os.path.join(hmm_dir_blast, filename)\n",
    "    destination_path = f'{name}@{server}:~/hmm/blast/hmmbuild/{filename}'\n",
    "    subprocess.run(['scp', source_path, destination_path])\n",
    "    print(f'File {filename} copied to {destination_path}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File O43791_169-178.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/O43791_169-178.hmm\n",
      "File P04370-5_1-169.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/P04370-5_1-169.hmm\n",
      "File Q9H832_327-354.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/Q9H832_327-354.hmm\n",
      "File O35274_584-602.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/O35274_584-602.hmm\n",
      "File Q8IW19_399-420.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/Q8IW19_399-420.hmm\n",
      "File Q8R464_25-120.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/Q8R464_25-120.hmm\n",
      "File Q93KQ4_51-81.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/Q93KQ4_51-81.hmm\n",
      "File P01019_436-450.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/P01019_436-450.hmm\n",
      "File S7W634_1-32.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/S7W634_1-32.hmm\n",
      "File Q99967_220-269.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/Q99967_220-269.hmm\n",
      "File Q5T4W7_108-120.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/Q5T4W7_108-120.hmm\n",
      "File A4L7I2_1658-1856.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/A4L7I2_1658-1856.hmm\n",
      "File Q5VZK9_1005-1020.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/Q5VZK9_1005-1020.hmm\n",
      "File Q9CXY6_29-44.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/Q9CXY6_29-44.hmm\n",
      "File O35274_417-494.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/O35274_417-494.hmm\n",
      "File A8AZZ3_116-134.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/A8AZZ3_116-134.hmm\n",
      "File Q9CXY6_347-390.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/Q9CXY6_347-390.hmm\n",
      "File P00736_193-305.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/P00736_193-305.hmm\n",
      "File Q9Z2F5_306-430.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/Q9Z2F5_306-430.hmm\n",
      "File Q8K4J6_111-142.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/Q8K4J6_111-142.hmm\n",
      "File Q84852_17-299.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/Q84852_17-299.hmm\n",
      "File P01097_28-50.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/P01097_28-50.hmm\n",
      "File O43474_1-130.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/O43474_1-130.hmm\n",
      "File A1L1Q4_1-71.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/A1L1Q4_1-71.hmm\n",
      "File O43791_357-374.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/O43791_357-374.hmm\n",
      "File Q8K4J6_155-186.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/Q8K4J6_155-186.hmm\n",
      "File O35274_1-154.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/O35274_1-154.hmm\n",
      "File Q9BYI3_149-253.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/Q9BYI3_149-253.hmm\n",
      "File Q8WUG5_1-105.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/Q8WUG5_1-105.hmm\n",
      "File Q86FP8_20-80.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/Q86FP8_20-80.hmm\n",
      "File P00392_550-561.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/P00392_550-561.hmm\n",
      "File Q9H832_1-99.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/Q9H832_1-99.hmm\n",
      "File J8TM36_236-249.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/J8TM36_236-249.hmm\n",
      "File O88339_144-575.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/O88339_144-575.hmm\n",
      "File O15922_1-89.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/O15922_1-89.hmm\n",
      "File O00585_71-111.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/O00585_71-111.hmm\n",
      "File Q8K4J6_67-98.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/Q8K4J6_67-98.hmm\n",
      "File Q8IW19_450-511.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/Q8IW19_450-511.hmm\n",
      "File S6B291_240-465.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/S6B291_240-465.hmm\n",
      "File O14958_371-399.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/O14958_371-399.hmm\n",
      "File P07342_580-595.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/P07342_580-595.hmm\n",
      "File P01019_44-64.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/P01019_44-64.hmm\n",
      "File O15922_315-324.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/O15922_315-324.hmm\n",
      "File A8AZZ3_24-44.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/A8AZZ3_24-44.hmm\n",
      "File A8AZZ3_146-195.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/A8AZZ3_146-195.hmm\n",
      "File Q9H0E2_1-53.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/Q9H0E2_1-53.hmm\n",
      "File Q9BYF1_769-805.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/Q9BYF1_769-805.hmm\n",
      "File O14727_1-104.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/O14727_1-104.hmm\n",
      "File Q8IU57_270-299.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/Q8IU57_270-299.hmm\n",
      "File P07342_647-687.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/P07342_647-687.hmm\n",
      "File O00308_662-702.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/O00308_662-702.hmm\n",
      "File O15922_230-240.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/O15922_230-240.hmm\n",
      "File P00742_90-126.hmm copied to alina@ecate:~/hmm/clustal/hmmbuild/P00742_90-126.hmm\n"
     ]
    }
   ],
   "source": [
    "# Copy the HMMs from the Clustal folder\n",
    "for filename in os.listdir(hmm_dir_clustal):\n",
    "    source_path = os.path.join(hmm_dir_clustal, filename)\n",
    "    destination_path = f'{name}@{server}:~/hmm/clustal/hmmbuild/{filename}'\n",
    "    subprocess.run(['scp', source_path, destination_path])\n",
    "    print(f'File {filename} copied to {destination_path}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After copying the files, we run `hmmsearch` command. The following code should be uncommented if there are new files provided. However, it requires some time to retrieve the output."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Construct the source and the target directories for BLAST\n",
    "blast_hmmbuild_dir = f\"/home/{name}/hmm/blast/hmmbuild\"\n",
    "blast_hmmsearch_dir = f\"/home/{name}/hmm/blast/hmmsearch\"\n",
    "\n",
    "# List files in the source directory on the remote machine\n",
    "ls_command = f'ssh {name}@{server} \"ls {blast_hmmbuild_dir}\"'\n",
    "file_list = !{ls_command}\n",
    "\n",
    "# # Iterate over each filename - RUNS FOR A LONG TIME\n",
    "# for filename in file_list:\n",
    "#     filename = filename.strip()\n",
    "#     source_path = f'{blast_hmmbuild_dir}/{filename}'\n",
    "#     dest_filename = f'{os.path.splitext(filename)[0]}'\n",
    "#     destination_path = f'{blast_hmmsearch_dir}/hmmsearch_blast_{dest_filename}.txt'\n",
    "    \n",
    "#     # Run hmmsearch command using ssh without redirection\n",
    "#     !ssh {name}@{server} \"/software/packages/hmmer/hmmer-3.3.2/usr/bin/hmmsearch {source_path} /db/rp/rp-seqs-75.fasta.gz > {destination_path} 2>&1\"\n",
    "        \n",
    "#     print(f\"hmmsearch completed for {filename} - Results saved to {destination_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hmmsearch_blast_A1L1Q4_1-71.txt               100%  130MB   6.4MB/s   00:20    \n",
      "hmmsearch_blast_A4L7I2_1658-1856.txt          100%   43KB 571.8KB/s   00:00    \n",
      "hmmsearch_blast_A8AZZ3_116-134.txt            100%   24KB 640.9KB/s   00:00    \n",
      "hmmsearch_blast_A8AZZ3_146-195.txt            100%   13KB 220.4KB/s   00:00    \n",
      "hmmsearch_blast_A8AZZ3_24-44.txt              100%   75KB 935.2KB/s   00:00    \n",
      "hmmsearch_blast_J8TM36_236-249.txt            100% 1803    28.8KB/s   00:00    \n",
      "hmmsearch_blast_O00308_662-702.txt            100% 7658KB   5.1MB/s   00:01    \n",
      "hmmsearch_blast_O00585_71-111.txt             100% 3835KB   5.7MB/s   00:00    \n",
      "hmmsearch_blast_O14727_1-104.txt              100% 2407KB   5.8MB/s   00:00    \n",
      "hmmsearch_blast_O14958_371-399.txt            100% 2859    54.2KB/s   00:00    \n",
      "hmmsearch_blast_O15922_1-89.txt               100%   86KB   1.7MB/s   00:00    \n",
      "hmmsearch_blast_O15922_230-240.txt            100% 1805    18.1KB/s   00:00    \n",
      "hmmsearch_blast_O15922_315-324.txt            100% 1807    41.6KB/s   00:00    \n",
      "hmmsearch_blast_O35274_1-154.txt              100% 3430KB   5.3MB/s   00:00    \n",
      "hmmsearch_blast_O35274_417-494.txt            100% 3286KB   4.9MB/s   00:00    \n",
      "hmmsearch_blast_O35274_584-602.txt            100% 1917KB   5.5MB/s   00:00    \n",
      "hmmsearch_blast_O43474_1-130.txt              100%  654KB   4.4MB/s   00:00    \n",
      "hmmsearch_blast_O43791_169-178.txt            100%   19KB 347.1KB/s   00:00    \n",
      "hmmsearch_blast_O43791_357-374.txt            100% 1124KB   5.0MB/s   00:00    \n",
      "hmmsearch_blast_O88339_144-575.txt            100% 7652KB   5.8MB/s   00:01    \n",
      "hmmsearch_blast_P00392_550-561.txt            100%  239KB   2.2MB/s   00:00    \n",
      "hmmsearch_blast_P00736_193-305.txt            100%   72MB   6.3MB/s   00:11    \n",
      "hmmsearch_blast_P00742_90-126.txt             100%  137MB   6.2MB/s   00:22    \n",
      "hmmsearch_blast_P01019_436-450.txt            100%  131KB   2.2MB/s   00:00    \n",
      "hmmsearch_blast_P01019_44-64.txt              100%  142KB   2.6MB/s   00:00    \n",
      "hmmsearch_blast_P01097_28-50.txt              100%  452KB   3.9MB/s   00:00    \n",
      "hmmsearch_blast_P04370-5_1-169.txt            100% 2000KB   5.9MB/s   00:00    \n",
      "hmmsearch_blast_P07342_580-595.txt            100%  794KB   4.3MB/s   00:00    \n",
      "hmmsearch_blast_P07342_647-687.txt            100% 1054KB   5.1MB/s   00:00    \n",
      "hmmsearch_blast_Q5T4W7_108-120.txt            100% 1803    22.3KB/s   00:00    \n",
      "hmmsearch_blast_Q5VZK9_1005-1020.txt          100%  407KB   4.3MB/s   00:00    \n",
      "hmmsearch_blast_Q84852_17-299.txt             100%   38KB 917.4KB/s   00:00    \n",
      "hmmsearch_blast_Q86FP8_20-80.txt              100% 4812KB   5.8MB/s   00:00    \n",
      "hmmsearch_blast_Q8IU57_270-299.txt            100%  377KB   3.6MB/s   00:00    \n",
      "hmmsearch_blast_Q8IW19_399-420.txt            100%  163KB   2.3MB/s   00:00    \n",
      "hmmsearch_blast_Q8IW19_450-511.txt            100%  526KB   3.8MB/s   00:00    \n",
      "hmmsearch_blast_Q8K4J6_111-142.txt            100% 2517KB   4.5MB/s   00:00    \n",
      "hmmsearch_blast_Q8K4J6_155-186.txt            100% 1578KB   4.2MB/s   00:00    \n",
      "hmmsearch_blast_Q8K4J6_67-98.txt              100% 3471KB   4.2MB/s   00:00    \n",
      "hmmsearch_blast_Q8R464_25-120.txt             100% 3971KB   3.6MB/s   00:01    \n",
      "hmmsearch_blast_Q8WUG5_1-105.txt              100%  914KB   4.7MB/s   00:00    \n",
      "hmmsearch_blast_Q93KQ4_51-81.txt              100%  126KB   1.7MB/s   00:00    \n",
      "hmmsearch_blast_Q99967_220-269.txt            100% 1325KB   5.2MB/s   00:00    \n",
      "hmmsearch_blast_Q9BYF1_769-805.txt            100% 1319KB   5.5MB/s   00:00    \n",
      "hmmsearch_blast_Q9BYI3_149-253.txt            100% 2643KB   4.9MB/s   00:00    \n",
      "hmmsearch_blast_Q9CXY6_29-44.txt              100%  480KB   3.1MB/s   00:00    \n",
      "hmmsearch_blast_Q9CXY6_347-390.txt            100%  770KB   3.7MB/s   00:00    \n",
      "hmmsearch_blast_Q9H0E2_1-53.txt               100%  899KB   4.1MB/s   00:00    \n",
      "hmmsearch_blast_Q9H832_1-99.txt               100%  438KB   3.5MB/s   00:00    \n",
      "hmmsearch_blast_Q9H832_327-354.txt            100%  406KB   2.8MB/s   00:00    \n",
      "hmmsearch_blast_Q9Z2F5_306-430.txt            100% 4529KB   5.5MB/s   00:00    \n",
      "hmmsearch_blast_S6B291_240-465.txt            100%   32MB   5.2MB/s   00:06    \n",
      "hmmsearch_blast_S7W634_1-32.txt               100%  132KB   1.3MB/s   00:00    \n"
     ]
    }
   ],
   "source": [
    "# Copy BLAST results to the local folder\n",
    "for filename in file_list:\n",
    "    filename = filename.strip()\n",
    "    source_path = f'{blast_hmmbuild_dir}/{filename}'\n",
    "    dest_filename = f'{os.path.splitext(filename)[0]}'\n",
    "    \n",
    "    # Copy files\n",
    "    !scp {name}@{server}:~/hmm/blast/hmmsearch/hmmsearch_blast_{dest_filename}.txt hmm_dir_blast        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Construct the source and the target directories for ClustalOmega\n",
    "clustal_hmmbuild_dir = f\"/home/{name}/hmm/clustal/hmmbuild\"\n",
    "clustal_hmmsearch_dir = f\"/home/{name}/hmm/clustal/hmmsearch\"\n",
    "\n",
    "# List files in the source directory on the remote machine\n",
    "ls_command = f'ssh {name}@{server} \"ls {clustal_hmmbuild_dir}\"'\n",
    "file_list = !{ls_command}\n",
    "\n",
    "# # Iterate over each filename\n",
    "# for filename in file_list:\n",
    "#     filename = filename.strip()\n",
    "#     source_path = f'{clustal_hmmbuild_dir}/{filename}'\n",
    "#     dest_filename = f'{os.path.splitext(filename)[0]}'\n",
    "#     destination_path = f'{clustal_hmmsearch_dir}/hmmsearch_clustal_{dest_filename}.txt'\n",
    "    \n",
    "#     # Run hmmsearch command using ssh without redirection\n",
    "#     !ssh {name}@{server} \"/software/packages/hmmer/hmmer-3.3.2/usr/bin/hmmsearch {source_path} /db/rp/rp-seqs-75.fasta.gz > {destination_path} 2>&1\"\n",
    "        \n",
    "#     print(f\"hmmsearch completed for {filename} - Results saved to {destination_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hmmsearch_clustal_A1L1Q4_1-71.txt             100%  134MB   6.1MB/s   00:22    \n"
     ]
    }
   ],
   "source": [
    "# Copy ClustalOmega results to the local folder\n",
    "for filename in file_list:\n",
    "    filename = filename.strip()\n",
    "    source_path = f'{clustal_hmmbuild_dir}/{filename}'\n",
    "    dest_filename = f'{os.path.splitext(filename)[0]}'\n",
    "    \n",
    "    # Copy files\n",
    "    !scp {name}@{server}:~/hmm/clustal/hmmsearch/hmmsearch_clustal_{dest_filename}.txt hmm_dir_clustal"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.1 hmmsearch results preprocessing\n",
    "Then we preprocess the data obtained from hmmsearch into the dataframes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set the paths to hmmsearch results for BLAST and ClustalOmega\n",
    "hmmsearch_dir_blast = f'{directory}/results/hmms/hmmsearch/blast'\n",
    "hmmsearch_dir_clustal = f'{directory}/results/hmms/hmmsearch/clustal'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "'  ------ inclusion threshold ------\\n' is not in list",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Input \u001b[0;32mIn [22]\u001b[0m, in \u001b[0;36m<cell line: 3>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m filename\u001b[38;5;241m.\u001b[39mendswith(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m.txt\u001b[39m\u001b[38;5;124m'\u001b[39m):\n\u001b[1;32m      5\u001b[0m         file_path \u001b[38;5;241m=\u001b[39m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(hmmsearch_dir_blast, filename)\n\u001b[0;32m----> 6\u001b[0m         hmmsearch_file \u001b[38;5;241m=\u001b[39m \u001b[43mprocess_hmmsearch_file\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfile_path\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      7\u001b[0m \u001b[38;5;66;03m#         df = pd.read_csv(hmmsearch_file)\u001b[39;00m\n\u001b[1;32m      8\u001b[0m         df_blast\u001b[38;5;241m.\u001b[39mappend(hmmsearch_file)\n",
      "File \u001b[0;32m~/HMM/functions.py:448\u001b[0m, in \u001b[0;36mprocess_hmmsearch_file\u001b[0;34m(filename)\u001b[0m\n\u001b[1;32m    446\u001b[0m \u001b[38;5;66;03m# Extract the column names\u001b[39;00m\n\u001b[1;32m    447\u001b[0m column_names \u001b[38;5;241m=\u001b[39m lines[\u001b[38;5;241m12\u001b[39m]\u001b[38;5;241m.\u001b[39msplit()\n\u001b[0;32m--> 448\u001b[0m inclusion_threshold_index \u001b[38;5;241m=\u001b[39m \u001b[43mlines\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mindex\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m  ------ inclusion threshold ------\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    450\u001b[0m \u001b[38;5;66;03m# Extract the data rows\u001b[39;00m\n\u001b[1;32m    451\u001b[0m data_rows \u001b[38;5;241m=\u001b[39m [line\u001b[38;5;241m.\u001b[39msplit() \u001b[38;5;28;01mfor\u001b[39;00m line \u001b[38;5;129;01min\u001b[39;00m lines[\u001b[38;5;241m14\u001b[39m:inclusion_threshold_index]]\n",
      "\u001b[0;31mValueError\u001b[0m: '  ------ inclusion threshold ------\\n' is not in list"
     ]
    }
   ],
   "source": [
    "# BLAST hmmsearch dataframe\n",
    "df_blast = []\n",
    "\n",
    "for filename in os.listdir(hmmsearch_dir_blast):\n",
    "    if filename.endswith('.txt'):\n",
    "        file_path = os.path.join(hmmsearch_dir_blast, filename)\n",
    "        hmmsearch_file = process_hmmsearch_file(file_path)\n",
    "#         df = pd.read_csv(hmmsearch_file)\n",
    "        df_blast.append(hmmsearch_file)\n",
    "        \n",
    "hmmsearch_blast_df = pd.concat(df_blast, ignore_index=True)\n",
    "# hmmsearch_blast_df.to_csv('results/pfam/hmmsearch_blast_df.csv', index=False)\n",
    "print(len(hmmsearch_blast_df))\n",
    "hmmsearch_blast_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 820,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. Make a dataframe with the statistics - the first table of the hmmsearch output\n",
    "stats_rp_75 = process_hmmsearch_file(f\"{directory}/results/hmms/hmmsearch/hmmsearch_rp_75_{id_dis}_{i}.txt\")\n",
    "# stats_rp_75.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 821,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2. Create a dataframe with the extracted regions from HMM - from the rest of the file\n",
    "hmm_rp_75 = f\"{directory}/results/hmms/hmmsearch/hmmsearch_rp_75_{id_dis}_{i}.txt\"\n",
    "hmmsearch_regions_rp_75 = extract_table_from_output(hmm_rp_75)\n",
    "# hmmsearch_regions_rp_75.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 822,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of hmmsearch hits for the 1 disordered region of the Q86FP8 protein (RP 75%): 1339\n",
      "Stored 'hmmsearch_results_rp_75' (DataFrame)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>E-value</th>\n",
       "      <th>score</th>\n",
       "      <th>bias</th>\n",
       "      <th>E-value</th>\n",
       "      <th>score</th>\n",
       "      <th>bias</th>\n",
       "      <th>exp</th>\n",
       "      <th>N</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>ali_from</th>\n",
       "      <th>ali_to</th>\n",
       "      <th>ali_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7e-33</td>\n",
       "      <td>125.0</td>\n",
       "      <td>0.2</td>\n",
       "      <td>1.6e-13</td>\n",
       "      <td>63.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.5</td>\n",
       "      <td>2</td>\n",
       "      <td>A0A5J9WTQ4</td>\n",
       "      <td>286</td>\n",
       "      <td>340</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.7e-20</td>\n",
       "      <td>84.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.3e-20</td>\n",
       "      <td>83.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1</td>\n",
       "      <td>A0A0D3D564</td>\n",
       "      <td>283</td>\n",
       "      <td>339</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5.8e-20</td>\n",
       "      <td>83.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.2e-19</td>\n",
       "      <td>82.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1</td>\n",
       "      <td>A0A397YI80</td>\n",
       "      <td>277</td>\n",
       "      <td>333</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5.8e-20</td>\n",
       "      <td>83.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.2e-19</td>\n",
       "      <td>82.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1</td>\n",
       "      <td>M4DEU0</td>\n",
       "      <td>277</td>\n",
       "      <td>333</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7.7e-20</td>\n",
       "      <td>83.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.7e-19</td>\n",
       "      <td>82.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1</td>\n",
       "      <td>M4CCJ6</td>\n",
       "      <td>283</td>\n",
       "      <td>339</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>7.7e-20</td>\n",
       "      <td>83.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.7e-19</td>\n",
       "      <td>82.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1</td>\n",
       "      <td>A0A398A8X2</td>\n",
       "      <td>283</td>\n",
       "      <td>339</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7.8e-20</td>\n",
       "      <td>83.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.7e-19</td>\n",
       "      <td>82.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1</td>\n",
       "      <td>A0A0D3BCZ4</td>\n",
       "      <td>284</td>\n",
       "      <td>340</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2e-19</td>\n",
       "      <td>81.9</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.4e-19</td>\n",
       "      <td>80.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.6</td>\n",
       "      <td>1</td>\n",
       "      <td>A0A6J0NM38</td>\n",
       "      <td>277</td>\n",
       "      <td>333</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2e-19</td>\n",
       "      <td>81.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.5e-19</td>\n",
       "      <td>80.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.6</td>\n",
       "      <td>1</td>\n",
       "      <td>A9TSC0</td>\n",
       "      <td>310</td>\n",
       "      <td>360</td>\n",
       "      <td>51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2.9e-19</td>\n",
       "      <td>81.3</td>\n",
       "      <td>0.1</td>\n",
       "      <td>6.5e-19</td>\n",
       "      <td>80.2</td>\n",
       "      <td>0.1</td>\n",
       "      <td>1.6</td>\n",
       "      <td>1</td>\n",
       "      <td>A0A087HC55</td>\n",
       "      <td>290</td>\n",
       "      <td>345</td>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>3.4e-19</td>\n",
       "      <td>81.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.6e-19</td>\n",
       "      <td>80.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1</td>\n",
       "      <td>A0A6J0NNF8</td>\n",
       "      <td>281</td>\n",
       "      <td>337</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>4.3e-19</td>\n",
       "      <td>80.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.7e-19</td>\n",
       "      <td>79.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1</td>\n",
       "      <td>R0I7V2</td>\n",
       "      <td>284</td>\n",
       "      <td>340</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>5.1e-19</td>\n",
       "      <td>80.6</td>\n",
       "      <td>0.1</td>\n",
       "      <td>1e-18</td>\n",
       "      <td>79.6</td>\n",
       "      <td>0.1</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1</td>\n",
       "      <td>A0A6D2JP04</td>\n",
       "      <td>285</td>\n",
       "      <td>341</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2.9e-18</td>\n",
       "      <td>78.2</td>\n",
       "      <td>0.1</td>\n",
       "      <td>5.8e-18</td>\n",
       "      <td>77.2</td>\n",
       "      <td>0.1</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1</td>\n",
       "      <td>V4L0T1</td>\n",
       "      <td>291</td>\n",
       "      <td>346</td>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1.4e-17</td>\n",
       "      <td>76.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3e-17</td>\n",
       "      <td>74.9</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.6</td>\n",
       "      <td>1</td>\n",
       "      <td>D7MKE8</td>\n",
       "      <td>288</td>\n",
       "      <td>343</td>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1.6e-17</td>\n",
       "      <td>75.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.3e-17</td>\n",
       "      <td>74.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1</td>\n",
       "      <td>A0A384L621</td>\n",
       "      <td>284</td>\n",
       "      <td>337</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1.6e-17</td>\n",
       "      <td>75.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.3e-17</td>\n",
       "      <td>74.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1</td>\n",
       "      <td>Q9LIP9</td>\n",
       "      <td>284</td>\n",
       "      <td>337</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1.6e-17</td>\n",
       "      <td>75.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.5e-17</td>\n",
       "      <td>74.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.6</td>\n",
       "      <td>1</td>\n",
       "      <td>D7L496</td>\n",
       "      <td>335</td>\n",
       "      <td>390</td>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1.9e-17</td>\n",
       "      <td>75.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.4e-17</td>\n",
       "      <td>74.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.4</td>\n",
       "      <td>1</td>\n",
       "      <td>A0A1R3HHC1</td>\n",
       "      <td>296</td>\n",
       "      <td>352</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>6.5e-17</td>\n",
       "      <td>73.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.2e-16</td>\n",
       "      <td>73.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.4</td>\n",
       "      <td>1</td>\n",
       "      <td>A0A1R3K173</td>\n",
       "      <td>296</td>\n",
       "      <td>352</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    E-value  score bias  E-value score bias  exp  N    Sequence  ali_from  \\\n",
       "0     7e-33  125.0  0.2  1.6e-13  63.0  0.0  2.5  2  A0A5J9WTQ4       286   \n",
       "1   2.7e-20   84.6  0.0  5.3e-20  83.7  0.0  1.5  1  A0A0D3D564       283   \n",
       "2   5.8e-20   83.6  0.0  1.2e-19  82.6  0.0  1.5  1  A0A397YI80       277   \n",
       "3   5.8e-20   83.6  0.0  1.2e-19  82.6  0.0  1.5  1      M4DEU0       277   \n",
       "4   7.7e-20   83.2  0.0  1.7e-19  82.1  0.0  1.5  1      M4CCJ6       283   \n",
       "5   7.7e-20   83.2  0.0  1.7e-19  82.1  0.0  1.5  1  A0A398A8X2       283   \n",
       "6   7.8e-20   83.2  0.0  1.7e-19  82.1  0.0  1.5  1  A0A0D3BCZ4       284   \n",
       "7     2e-19   81.9  0.0  4.4e-19  80.8  0.0  1.6  1  A0A6J0NM38       277   \n",
       "8     2e-19   81.8  0.0  4.5e-19  80.7  0.0  1.6  1      A9TSC0       310   \n",
       "9   2.9e-19   81.3  0.1  6.5e-19  80.2  0.1  1.6  1  A0A087HC55       290   \n",
       "10  3.4e-19   81.1  0.0  6.6e-19  80.2  0.0  1.5  1  A0A6J0NNF8       281   \n",
       "11  4.3e-19   80.8  0.0  8.7e-19  79.8  0.0  1.5  1      R0I7V2       284   \n",
       "12  5.1e-19   80.6  0.1    1e-18  79.6  0.1  1.5  1  A0A6D2JP04       285   \n",
       "13  2.9e-18   78.2  0.1  5.8e-18  77.2  0.1  1.5  1      V4L0T1       291   \n",
       "14  1.4e-17   76.0  0.0    3e-17  74.9  0.0  1.6  1      D7MKE8       288   \n",
       "15  1.6e-17   75.8  0.0  3.3e-17  74.8  0.0  1.5  1  A0A384L621       284   \n",
       "16  1.6e-17   75.8  0.0  3.3e-17  74.8  0.0  1.5  1      Q9LIP9       284   \n",
       "17  1.6e-17   75.7  0.0  3.5e-17  74.7  0.0  1.6  1      D7L496       335   \n",
       "18  1.9e-17   75.6  0.0  3.4e-17  74.7  0.0  1.4  1  A0A1R3HHC1       296   \n",
       "19  6.5e-17   73.8  0.0  1.2e-16  73.0  0.0  1.4  1  A0A1R3K173       296   \n",
       "\n",
       "    ali_to  ali_length  \n",
       "0      340          55  \n",
       "1      339          57  \n",
       "2      333          57  \n",
       "3      333          57  \n",
       "4      339          57  \n",
       "5      339          57  \n",
       "6      340          57  \n",
       "7      333          57  \n",
       "8      360          51  \n",
       "9      345          56  \n",
       "10     337          57  \n",
       "11     340          57  \n",
       "12     341          57  \n",
       "13     346          56  \n",
       "14     343          56  \n",
       "15     337          54  \n",
       "16     337          54  \n",
       "17     390          56  \n",
       "18     352          57  \n",
       "19     352          57  "
      ]
     },
     "execution_count": 822,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 3. Combine the results into a dataframe and save the file\n",
    "hmmsearch_results_rp_75 = pd.merge(stats_rp_75, hmmsearch_regions_rp_75, left_on=\"Sequence\", right_on=\"id\", how=\"inner\")\n",
    "hmmsearch_results_rp_75 = hmmsearch_results_rp_75.drop(columns=[\"Description\", \"id\"])\n",
    "hmmsearch_results_rp_75.to_csv(f\"{directory}/results/hmms/hmmsearch/stats/stats_rp_75_{id_dis}_{i}.csv\", index=False)\n",
    "print(f\"Number of hmmsearch hits for the {i} disordered region of the {id_dis} protein (RP 75%):\", len(hmmsearch_results_rp_75))\n",
    "%store hmmsearch_results_rp_75\n",
    "hmmsearch_results_rp_75[:20]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 743,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of unique IDs for the Q86FP8 MSA: 1339\n"
     ]
    }
   ],
   "source": [
    "subject_ids_rp_75 = hmmsearch_results_rp_75.iloc[:, 8].unique()\n",
    "print(f\"The number of unique IDs for the {id_dis} MSA:\", len(subject_ids_rp_75))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Pfam\n",
    "### 3.1 Pfam results preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 744,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-25T07:30:38.221338Z",
     "start_time": "2023-08-25T07:30:38.183994Z"
    }
   },
   "outputs": [],
   "source": [
    "# # Copy the files with the statistics to the remote computer\n",
    "# !scp {directory}/results/hmms/hmmsearch/stats/stats_rp_75_{id_dis}_{i}.csv {name}@{server}:~/stats/stats_rp_75/stats_rp_75_{id_dis}_{i}.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 746,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Check the overlaps with Interpro domains (RP 75%)\n",
    "# !ssh {name}@{server} \"/home/alina/protein2ipr.py /home/alina/stats/stats_rp_75 /home/alina/filtered.tsv.gz protein2ipr_rp_75.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 747,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-25T07:40:00.248249Z",
     "start_time": "2023-08-25T07:40:00.231919Z"
    }
   },
   "outputs": [],
   "source": [
    "# # Copy the files to the local folder\n",
    "# !scp {name}@{server}:~/protein2ipr_rp_75.tsv {directory}/results/pfam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 823,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter only entries with Pfam ID and intercepting regions with the curated_disprot instances (RP 75%)\n",
    "filename = f'{directory}/results/pfam/protein2ipr_rp_75.tsv'\n",
    "pfam_rp_75 = pfam_processing(filename)\n",
    "\n",
    "# pfam_rp_75 = pfam_rp_75[pfam_rp_75['uniprot_id'].isin(subject_ids_rp_75)]\n",
    "# print(f\"The number of retrieved Pfam instances for the {id_dis}_{i} protein (RP 75%): {len(pfam_rp_75)}\")\n",
    "# # print(\"The number of Uniprot instances not covered by Pfam (RP 75%):\", len(hmmsearch_results_rp_75) - len(pfam_rp_75))\n",
    "# pfam_rp_75.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 825,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>pfam_id</th>\n",
       "      <th>ipr_id</th>\n",
       "      <th>start_pfam</th>\n",
       "      <th>end_pfam</th>\n",
       "      <th>length_pfam</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A010QN55</td>\n",
       "      <td>PF01380</td>\n",
       "      <td>IPR001347</td>\n",
       "      <td>381</td>\n",
       "      <td>508</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A010QN55</td>\n",
       "      <td>PF01380</td>\n",
       "      <td>IPR001347</td>\n",
       "      <td>553</td>\n",
       "      <td>682</td>\n",
       "      <td>130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A016UWB3</td>\n",
       "      <td>PF01380</td>\n",
       "      <td>IPR001347</td>\n",
       "      <td>173</td>\n",
       "      <td>300</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A016UWB3</td>\n",
       "      <td>PF01380</td>\n",
       "      <td>IPR001347</td>\n",
       "      <td>343</td>\n",
       "      <td>432</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A016UWX5</td>\n",
       "      <td>PF01380</td>\n",
       "      <td>IPR001347</td>\n",
       "      <td>202</td>\n",
       "      <td>329</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>A0A016UWX5</td>\n",
       "      <td>PF01380</td>\n",
       "      <td>IPR001347</td>\n",
       "      <td>372</td>\n",
       "      <td>461</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A0A017S8W7</td>\n",
       "      <td>PF01380</td>\n",
       "      <td>IPR001347</td>\n",
       "      <td>375</td>\n",
       "      <td>502</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>A0A017S8W7</td>\n",
       "      <td>PF01380</td>\n",
       "      <td>IPR001347</td>\n",
       "      <td>548</td>\n",
       "      <td>675</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A0A022QGG4</td>\n",
       "      <td>PF01380</td>\n",
       "      <td>IPR001347</td>\n",
       "      <td>365</td>\n",
       "      <td>493</td>\n",
       "      <td>129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>A0A022QGG4</td>\n",
       "      <td>PF01380</td>\n",
       "      <td>IPR001347</td>\n",
       "      <td>536</td>\n",
       "      <td>667</td>\n",
       "      <td>132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>A0A023ERG4</td>\n",
       "      <td>PF07528</td>\n",
       "      <td>IPR049401</td>\n",
       "      <td>20</td>\n",
       "      <td>229</td>\n",
       "      <td>210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>A0A023ERG4</td>\n",
       "      <td>PF20965</td>\n",
       "      <td>IPR049402</td>\n",
       "      <td>232</td>\n",
       "      <td>364</td>\n",
       "      <td>133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>A0A026WDI9</td>\n",
       "      <td>PF01380</td>\n",
       "      <td>IPR001347</td>\n",
       "      <td>367</td>\n",
       "      <td>494</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>A0A026WDI9</td>\n",
       "      <td>PF01380</td>\n",
       "      <td>IPR001347</td>\n",
       "      <td>538</td>\n",
       "      <td>667</td>\n",
       "      <td>130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>A0A059AY18</td>\n",
       "      <td>PF01380</td>\n",
       "      <td>IPR001347</td>\n",
       "      <td>192</td>\n",
       "      <td>320</td>\n",
       "      <td>129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>A0A059AY18</td>\n",
       "      <td>PF01380</td>\n",
       "      <td>IPR001347</td>\n",
       "      <td>363</td>\n",
       "      <td>494</td>\n",
       "      <td>132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>A0A059AYM9</td>\n",
       "      <td>PF01380</td>\n",
       "      <td>IPR001347</td>\n",
       "      <td>284</td>\n",
       "      <td>383</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>A0A059AYM9</td>\n",
       "      <td>PF13537</td>\n",
       "      <td>IPR017932</td>\n",
       "      <td>87</td>\n",
       "      <td>140</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>A0A059AYT0</td>\n",
       "      <td>PF01380</td>\n",
       "      <td>IPR001347</td>\n",
       "      <td>278</td>\n",
       "      <td>339</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>A0A059AYT0</td>\n",
       "      <td>PF01380</td>\n",
       "      <td>IPR001347</td>\n",
       "      <td>428</td>\n",
       "      <td>551</td>\n",
       "      <td>124</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    uniprot_id  pfam_id     ipr_id  start_pfam  end_pfam  length_pfam\n",
       "0   A0A010QN55  PF01380  IPR001347         381       508          128\n",
       "1   A0A010QN55  PF01380  IPR001347         553       682          130\n",
       "2   A0A016UWB3  PF01380  IPR001347         173       300          128\n",
       "3   A0A016UWB3  PF01380  IPR001347         343       432           90\n",
       "4   A0A016UWX5  PF01380  IPR001347         202       329          128\n",
       "5   A0A016UWX5  PF01380  IPR001347         372       461           90\n",
       "6   A0A017S8W7  PF01380  IPR001347         375       502          128\n",
       "7   A0A017S8W7  PF01380  IPR001347         548       675          128\n",
       "8   A0A022QGG4  PF01380  IPR001347         365       493          129\n",
       "9   A0A022QGG4  PF01380  IPR001347         536       667          132\n",
       "10  A0A023ERG4  PF07528  IPR049401          20       229          210\n",
       "11  A0A023ERG4  PF20965  IPR049402         232       364          133\n",
       "12  A0A026WDI9  PF01380  IPR001347         367       494          128\n",
       "13  A0A026WDI9  PF01380  IPR001347         538       667          130\n",
       "14  A0A059AY18  PF01380  IPR001347         192       320          129\n",
       "15  A0A059AY18  PF01380  IPR001347         363       494          132\n",
       "16  A0A059AYM9  PF01380  IPR001347         284       383          100\n",
       "17  A0A059AYM9  PF13537  IPR017932          87       140           54\n",
       "18  A0A059AYT0  PF01380  IPR001347         278       339           62\n",
       "19  A0A059AYT0  PF01380  IPR001347         428       551          124"
      ]
     },
     "execution_count": 825,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pfam_rp_75[:20]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 751,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>pfam_id</th>\n",
       "      <th>ipr_id</th>\n",
       "      <th>start_pfam</th>\n",
       "      <th>end_pfam</th>\n",
       "      <th>length_pfam</th>\n",
       "      <th>ali_from</th>\n",
       "      <th>ali_to</th>\n",
       "      <th>ali_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A010QN55</td>\n",
       "      <td>PF01380</td>\n",
       "      <td>IPR001347</td>\n",
       "      <td>381</td>\n",
       "      <td>508</td>\n",
       "      <td>128</td>\n",
       "      <td>330</td>\n",
       "      <td>362</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A010QN55</td>\n",
       "      <td>PF01380</td>\n",
       "      <td>IPR001347</td>\n",
       "      <td>553</td>\n",
       "      <td>682</td>\n",
       "      <td>130</td>\n",
       "      <td>330</td>\n",
       "      <td>362</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A016UWB3</td>\n",
       "      <td>PF01380</td>\n",
       "      <td>IPR001347</td>\n",
       "      <td>173</td>\n",
       "      <td>300</td>\n",
       "      <td>128</td>\n",
       "      <td>104</td>\n",
       "      <td>155</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A016UWB3</td>\n",
       "      <td>PF01380</td>\n",
       "      <td>IPR001347</td>\n",
       "      <td>343</td>\n",
       "      <td>432</td>\n",
       "      <td>90</td>\n",
       "      <td>104</td>\n",
       "      <td>155</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A016UWX5</td>\n",
       "      <td>PF01380</td>\n",
       "      <td>IPR001347</td>\n",
       "      <td>202</td>\n",
       "      <td>329</td>\n",
       "      <td>128</td>\n",
       "      <td>133</td>\n",
       "      <td>184</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   uniprot_id  pfam_id     ipr_id  start_pfam  end_pfam  length_pfam  \\\n",
       "0  A0A010QN55  PF01380  IPR001347         381       508          128   \n",
       "1  A0A010QN55  PF01380  IPR001347         553       682          130   \n",
       "2  A0A016UWB3  PF01380  IPR001347         173       300          128   \n",
       "3  A0A016UWB3  PF01380  IPR001347         343       432           90   \n",
       "4  A0A016UWX5  PF01380  IPR001347         202       329          128   \n",
       "\n",
       "   ali_from  ali_to  ali_length  \n",
       "0       330     362          33  \n",
       "1       330     362          33  \n",
       "2       104     155          52  \n",
       "3       104     155          52  \n",
       "4       133     184          52  "
      ]
     },
     "execution_count": 751,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge with hmmsearch result - for one MSA (RP 75%)\n",
    "pfam_rp_75 = pd.merge(pfam_rp_75, hmmsearch_results_rp_75[['Sequence', 'ali_from', 'ali_to', 'ali_length']], left_on='uniprot_id', right_on='Sequence', how='left')\n",
    "pfam_rp_75 = pfam_rp_75.dropna(axis=0)\n",
    "pfam_rp_75['ali_from'] = pfam_rp_75['ali_from'].astype(int)\n",
    "pfam_rp_75['ali_to'] = pfam_rp_75['ali_to'].astype(int)\n",
    "pfam_rp_75['ali_length'] = pfam_rp_75['ali_length'].astype(int)\n",
    "pfam_rp_75 = pfam_rp_75.drop(columns='Sequence')\n",
    "# pfam_rp_75.to_csv(f'results/pfam/pfam_overlap/pfam_overlap_{id_dis}.csv', index=False)\n",
    "# print(f\"The number of overlaps for {id_dis}_{i} (RP 75%):\", len(pfam_rp_75))\n",
    "pfam_rp_75.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 752,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the overlaps percentage\n",
    "def pfam_hmm_overlap(row_pfam):\n",
    "    start_pfam = row_pfam['start_pfam']\n",
    "    end_pfam = row_pfam['end_pfam']\n",
    "    start_hmm = row_pfam['ali_from']\n",
    "    end_hmm = row_pfam['ali_to']\n",
    "    len_pfam = row_pfam['length_pfam']\n",
    "    len_hmm = row_pfam['ali_length']\n",
    "\n",
    "    overlap_len = min(end_pfam, end_hmm) - max(start_pfam, start_hmm) + 1\n",
    "    overlap_pfam = overlap_len/len_pfam * 100\n",
    "    overlap_hmm = overlap_len/len_hmm * 100\n",
    "    max_length = len_pfam + len_hmm - overlap_len\n",
    "    non_overlap_len = max_length - overlap_len + 1\n",
    "\n",
    "    if overlap_len > 0:\n",
    "        overlap_perc = (overlap_len / max_length) * 100\n",
    "#         overlap_perc = 2 * overlap_len / (len_hmm + len_pfam)\n",
    "    else:\n",
    "        overlap_len = 0\n",
    "        overlap_pfam = 0\n",
    "        overlap_hmm = 0\n",
    "        overlap_perc = 0\n",
    "\n",
    "    return overlap_len, overlap_pfam, overlap_hmm, non_overlap_len, overlap_perc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 754,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>pfam_id</th>\n",
       "      <th>ipr_id</th>\n",
       "      <th>start_pfam</th>\n",
       "      <th>end_pfam</th>\n",
       "      <th>length_pfam</th>\n",
       "      <th>ali_from</th>\n",
       "      <th>ali_to</th>\n",
       "      <th>ali_length</th>\n",
       "      <th>overl_len</th>\n",
       "      <th>overl_pfam</th>\n",
       "      <th>overl_hmm</th>\n",
       "      <th>non_overl_len</th>\n",
       "      <th>overl_perc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A010QN55</td>\n",
       "      <td>PF01380</td>\n",
       "      <td>IPR001347</td>\n",
       "      <td>381</td>\n",
       "      <td>508</td>\n",
       "      <td>128</td>\n",
       "      <td>330</td>\n",
       "      <td>362</td>\n",
       "      <td>33</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>198</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A010QN55</td>\n",
       "      <td>PF01380</td>\n",
       "      <td>IPR001347</td>\n",
       "      <td>553</td>\n",
       "      <td>682</td>\n",
       "      <td>130</td>\n",
       "      <td>330</td>\n",
       "      <td>362</td>\n",
       "      <td>33</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>544</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A016UWB3</td>\n",
       "      <td>PF01380</td>\n",
       "      <td>IPR001347</td>\n",
       "      <td>173</td>\n",
       "      <td>300</td>\n",
       "      <td>128</td>\n",
       "      <td>104</td>\n",
       "      <td>155</td>\n",
       "      <td>52</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>215</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A016UWB3</td>\n",
       "      <td>PF01380</td>\n",
       "      <td>IPR001347</td>\n",
       "      <td>343</td>\n",
       "      <td>432</td>\n",
       "      <td>90</td>\n",
       "      <td>104</td>\n",
       "      <td>155</td>\n",
       "      <td>52</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>517</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A016UWX5</td>\n",
       "      <td>PF01380</td>\n",
       "      <td>IPR001347</td>\n",
       "      <td>202</td>\n",
       "      <td>329</td>\n",
       "      <td>128</td>\n",
       "      <td>133</td>\n",
       "      <td>184</td>\n",
       "      <td>52</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>215</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   uniprot_id  pfam_id     ipr_id  start_pfam  end_pfam  length_pfam  \\\n",
       "0  A0A010QN55  PF01380  IPR001347         381       508          128   \n",
       "1  A0A010QN55  PF01380  IPR001347         553       682          130   \n",
       "2  A0A016UWB3  PF01380  IPR001347         173       300          128   \n",
       "3  A0A016UWB3  PF01380  IPR001347         343       432           90   \n",
       "4  A0A016UWX5  PF01380  IPR001347         202       329          128   \n",
       "\n",
       "   ali_from  ali_to  ali_length  overl_len  overl_pfam  overl_hmm  \\\n",
       "0       330     362          33          0         0.0        0.0   \n",
       "1       330     362          33          0         0.0        0.0   \n",
       "2       104     155          52          0         0.0        0.0   \n",
       "3       104     155          52          0         0.0        0.0   \n",
       "4       133     184          52          0         0.0        0.0   \n",
       "\n",
       "   non_overl_len  overl_perc  \n",
       "0            198         0.0  \n",
       "1            544         0.0  \n",
       "2            215         0.0  \n",
       "3            517         0.0  \n",
       "4            215         0.0  "
      ]
     },
     "execution_count": 754,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add the overlaps to the dataframe (RP 75%)\n",
    "# overlap_percentages = []\n",
    "overlap_pfam_hmm = []\n",
    "\n",
    "# for index_pfam, row_pfam in pfam_rp_75.iterrows():\n",
    "#     row_disprot = curated_query.iloc[i-1]\n",
    "#     overl_perc_pfam, overl_perc_hmm, overl_len_pfam, overl_len_hmm = calculate_overlap(row_pfam, row_disprot)\n",
    "#     overl_perc_pfam = round(overl_perc_pfam, 2)\n",
    "#     overl_perc_hmm = round(overl_perc_hmm, 2)\n",
    "#     overlap_percentages.append((overl_perc_pfam, overl_perc_hmm, overl_len_pfam, overl_len_hmm))\n",
    "\n",
    "for index_pfam, row_pfam in pfam_rp_75.iterrows():\n",
    "    overl_len, overl_pfam, overl_hmm, non_overl_len, overl_perc = pfam_hmm_overlap(row_pfam)\n",
    "    overl_pfam = round(overl_pfam, 2)\n",
    "    overl_hmm = round(overl_hmm, 2)\n",
    "    overl_perc = round(overl_perc, 2)\n",
    "    overlap_pfam_hmm.append((overl_len, overl_pfam, overl_hmm, non_overl_len, overl_perc))\n",
    "\n",
    "# Extract overlap_pfam and overlap_hmm from the list of tuples\n",
    "# pfam_rp_75['overlap_pfam_%'], pfam_rp_75['overlap_hmm_%'], pfam_rp_75['overlap_pfam_len'], pfam_rp_75['overlap_hmm_len'] = zip(*overlap_percentages)\n",
    "pfam_rp_75['overl_len'], pfam_rp_75['overl_pfam'], pfam_rp_75['overl_hmm'], pfam_rp_75['non_overl_len'], pfam_rp_75['overl_perc'] = zip(*overlap_pfam_hmm)\n",
    "# print(\"DisProt start:\", curated_query['start'].iloc[i-1], \"\\n\"\n",
    "#       \"DisProt end:\", curated_query['end'].iloc[i-1], \"\\n\"\n",
    "#       \"DisProt length:\", curated_query['length'].iloc[i-1])\n",
    "pfam_rp_75.to_csv(f'results/pfam/pfam_overlap/rp_75/pfam_rp_75_{id_dis}_{i}.csv', index=False)\n",
    "pfam_rp_75.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 756,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>pfam_id</th>\n",
       "      <th>ipr_id</th>\n",
       "      <th>start_pfam</th>\n",
       "      <th>end_pfam</th>\n",
       "      <th>length_pfam</th>\n",
       "      <th>ali_from</th>\n",
       "      <th>ali_to</th>\n",
       "      <th>ali_length</th>\n",
       "      <th>overl_len</th>\n",
       "      <th>overl_pfam</th>\n",
       "      <th>overl_hmm</th>\n",
       "      <th>non_overl_len</th>\n",
       "      <th>overl_perc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>A0A0C4EII3</td>\n",
       "      <td>PF01380</td>\n",
       "      <td>IPR001347</td>\n",
       "      <td>367</td>\n",
       "      <td>472</td>\n",
       "      <td>106</td>\n",
       "      <td>326</td>\n",
       "      <td>367</td>\n",
       "      <td>42</td>\n",
       "      <td>1</td>\n",
       "      <td>0.94</td>\n",
       "      <td>2.38</td>\n",
       "      <td>147</td>\n",
       "      <td>0.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>794</th>\n",
       "      <td>A0A2T0WH78</td>\n",
       "      <td>PF00725</td>\n",
       "      <td>IPR006108</td>\n",
       "      <td>479</td>\n",
       "      <td>570</td>\n",
       "      <td>92</td>\n",
       "      <td>532</td>\n",
       "      <td>580</td>\n",
       "      <td>49</td>\n",
       "      <td>39</td>\n",
       "      <td>42.39</td>\n",
       "      <td>79.59</td>\n",
       "      <td>64</td>\n",
       "      <td>38.24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1573</th>\n",
       "      <td>A0A6P4D783</td>\n",
       "      <td>PF01380</td>\n",
       "      <td>IPR001347</td>\n",
       "      <td>181</td>\n",
       "      <td>309</td>\n",
       "      <td>129</td>\n",
       "      <td>305</td>\n",
       "      <td>350</td>\n",
       "      <td>46</td>\n",
       "      <td>5</td>\n",
       "      <td>3.88</td>\n",
       "      <td>10.87</td>\n",
       "      <td>166</td>\n",
       "      <td>2.94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1917</th>\n",
       "      <td>A0A8C4HUS4</td>\n",
       "      <td>PF01380</td>\n",
       "      <td>IPR001347</td>\n",
       "      <td>352</td>\n",
       "      <td>479</td>\n",
       "      <td>128</td>\n",
       "      <td>322</td>\n",
       "      <td>362</td>\n",
       "      <td>41</td>\n",
       "      <td>11</td>\n",
       "      <td>8.59</td>\n",
       "      <td>26.83</td>\n",
       "      <td>148</td>\n",
       "      <td>6.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1921</th>\n",
       "      <td>A0A8C4HYL4</td>\n",
       "      <td>PF01380</td>\n",
       "      <td>IPR001347</td>\n",
       "      <td>301</td>\n",
       "      <td>428</td>\n",
       "      <td>128</td>\n",
       "      <td>328</td>\n",
       "      <td>368</td>\n",
       "      <td>41</td>\n",
       "      <td>41</td>\n",
       "      <td>32.03</td>\n",
       "      <td>100.00</td>\n",
       "      <td>88</td>\n",
       "      <td>32.03</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      uniprot_id  pfam_id     ipr_id  start_pfam  end_pfam  length_pfam  \\\n",
       "107   A0A0C4EII3  PF01380  IPR001347         367       472          106   \n",
       "794   A0A2T0WH78  PF00725  IPR006108         479       570           92   \n",
       "1573  A0A6P4D783  PF01380  IPR001347         181       309          129   \n",
       "1917  A0A8C4HUS4  PF01380  IPR001347         352       479          128   \n",
       "1921  A0A8C4HYL4  PF01380  IPR001347         301       428          128   \n",
       "\n",
       "      ali_from  ali_to  ali_length  overl_len  overl_pfam  overl_hmm  \\\n",
       "107        326     367          42          1        0.94       2.38   \n",
       "794        532     580          49         39       42.39      79.59   \n",
       "1573       305     350          46          5        3.88      10.87   \n",
       "1917       322     362          41         11        8.59      26.83   \n",
       "1921       328     368          41         41       32.03     100.00   \n",
       "\n",
       "      non_overl_len  overl_perc  \n",
       "107             147        0.68  \n",
       "794              64       38.24  \n",
       "1573            166        2.94  \n",
       "1917            148        6.96  \n",
       "1921             88       32.03  "
      ]
     },
     "execution_count": 756,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pfam_rp_75[pfam_rp_75['overl_len'] != 0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.2 Pfam results preprocessing (overall)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 760,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23010\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>pfam_id</th>\n",
       "      <th>ipr_id</th>\n",
       "      <th>start_pfam</th>\n",
       "      <th>end_pfam</th>\n",
       "      <th>length_pfam</th>\n",
       "      <th>ali_from</th>\n",
       "      <th>ali_to</th>\n",
       "      <th>ali_length</th>\n",
       "      <th>overl_len</th>\n",
       "      <th>overl_pfam</th>\n",
       "      <th>overl_hmm</th>\n",
       "      <th>non_overl_len</th>\n",
       "      <th>overl_perc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A060VYC8</td>\n",
       "      <td>PF07686</td>\n",
       "      <td>IPR013106</td>\n",
       "      <td>21</td>\n",
       "      <td>101</td>\n",
       "      <td>81</td>\n",
       "      <td>15</td>\n",
       "      <td>110</td>\n",
       "      <td>96</td>\n",
       "      <td>81</td>\n",
       "      <td>100.0</td>\n",
       "      <td>84.38</td>\n",
       "      <td>16</td>\n",
       "      <td>84.38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A060VYC8</td>\n",
       "      <td>PF08205</td>\n",
       "      <td>IPR013162</td>\n",
       "      <td>119</td>\n",
       "      <td>197</td>\n",
       "      <td>79</td>\n",
       "      <td>15</td>\n",
       "      <td>110</td>\n",
       "      <td>96</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>192</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A060W1A4</td>\n",
       "      <td>PF07686</td>\n",
       "      <td>IPR013106</td>\n",
       "      <td>23</td>\n",
       "      <td>99</td>\n",
       "      <td>77</td>\n",
       "      <td>18</td>\n",
       "      <td>110</td>\n",
       "      <td>93</td>\n",
       "      <td>77</td>\n",
       "      <td>100.0</td>\n",
       "      <td>82.80</td>\n",
       "      <td>17</td>\n",
       "      <td>82.80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A060W1A4</td>\n",
       "      <td>PF08205</td>\n",
       "      <td>IPR013162</td>\n",
       "      <td>118</td>\n",
       "      <td>195</td>\n",
       "      <td>78</td>\n",
       "      <td>18</td>\n",
       "      <td>110</td>\n",
       "      <td>93</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>186</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A060WNT4</td>\n",
       "      <td>PF07686</td>\n",
       "      <td>IPR013106</td>\n",
       "      <td>27</td>\n",
       "      <td>108</td>\n",
       "      <td>82</td>\n",
       "      <td>21</td>\n",
       "      <td>116</td>\n",
       "      <td>96</td>\n",
       "      <td>82</td>\n",
       "      <td>100.0</td>\n",
       "      <td>85.42</td>\n",
       "      <td>15</td>\n",
       "      <td>85.42</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   uniprot_id  pfam_id     ipr_id  start_pfam  end_pfam  length_pfam  \\\n",
       "0  A0A060VYC8  PF07686  IPR013106          21       101           81   \n",
       "1  A0A060VYC8  PF08205  IPR013162         119       197           79   \n",
       "2  A0A060W1A4  PF07686  IPR013106          23        99           77   \n",
       "3  A0A060W1A4  PF08205  IPR013162         118       195           78   \n",
       "4  A0A060WNT4  PF07686  IPR013106          27       108           82   \n",
       "\n",
       "   ali_from  ali_to  ali_length  overl_len  overl_pfam  overl_hmm  \\\n",
       "0        15     110          96         81       100.0      84.38   \n",
       "1        15     110          96          0         0.0       0.00   \n",
       "2        18     110          93         77       100.0      82.80   \n",
       "3        18     110          93          0         0.0       0.00   \n",
       "4        21     116          96         82       100.0      85.42   \n",
       "\n",
       "   non_overl_len  overl_perc  \n",
       "0             16       84.38  \n",
       "1            192        0.00  \n",
       "2             17       82.80  \n",
       "3            186        0.00  \n",
       "4             15       85.42  "
      ]
     },
     "execution_count": 760,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Combine all pfam results (RP 75%)\n",
    "pfam_path = f'{directory}/results/pfam/pfam_overlap/rp_75'\n",
    "dataframes = []\n",
    "\n",
    "for filename in os.listdir(pfam_path):\n",
    "    if filename.endswith('.csv'):\n",
    "        file_path = os.path.join(pfam_path, filename)\n",
    "        df = pd.read_csv(file_path)\n",
    "        dataframes.append(df)\n",
    "        \n",
    "pfam_hmm_rp_75_1 = pd.concat(dataframes, ignore_index=True)\n",
    "pfam_hmm_rp_75_1.to_csv('results/pfam/pfam_hmm_rp_75_1.csv', index=False)\n",
    "print(len(pfam_hmm_rp_75_1))\n",
    "pfam_hmm_rp_75_1.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
