{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set the path and input parameters\n",
    "import os\n",
    "directory = os.getcwd() # the main directory of the project\n",
    "\n",
    "# The credentials for the remote cluster\n",
    "name = 'alina'\n",
    "server = 'ecate'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-10-02T07:55:55.284130Z",
     "start_time": "2023-10-02T07:55:53.625394Z"
    }
   },
   "outputs": [],
   "source": [
    "# Importing the libraries\n",
    "from functions import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## HMM: Data preparation\n",
    "In this part we generate HMM models based on MSA and retrieve the results from hmmsearch and then from Pfam. We start with the loading of the `disordered` dataframe and analyse how many proteins and disordered regions we have."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set the maximum width of the columns\n",
    "pd.set_option('display.max_colwidth', 20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of rows with the disordered regions: 7393\n"
     ]
    }
   ],
   "source": [
    "# Load the disordered and curated_disprot dataframes\n",
    "disordered_df = pd.read_csv('{}/disordered_df.csv'.format(directory))\n",
    "curated_disprot_df = pd.read_csv('{}/curated_disprot.csv'.format(directory))\n",
    "\n",
    "print('The number of rows with the disordered regions: {}'.format(len(disordered_df)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of proteins with disordered regions: 39\n",
      "The number of disordered regions in the proteins: 53\n"
     ]
    }
   ],
   "source": [
    "# Collect the disordered regions in a dictionary\n",
    "dis_regs = set()\n",
    "\n",
    "for i, row in disordered_df.iterrows():\n",
    "    dis_id = row[0]\n",
    "    matching_row = curated_disprot_df[curated_disprot_df['acc'] == dis_id]\n",
    "    if not matching_row.empty:\n",
    "        region = matching_row['region']\n",
    "        dis_regs.update(region)\n",
    "        dis_regs_list = list(dis_regs) # convert set to a list\n",
    "        \n",
    "# Define an array of disordered regions ids\n",
    "disprot_ids = disordered_df['query_id'].unique()\n",
    "\n",
    "print('The number of proteins with disordered regions: {}'.format(len(disprot_ids)))\n",
    "print('The number of disordered regions in the proteins: {}'.format(len(dis_regs_list)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The results above show how many proteins with the disordered regions are in the provided data, as well as how many disordered regions do they contain."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-02T13:38:09.492889Z",
     "start_time": "2023-08-02T13:38:08.673895Z"
    }
   },
   "source": [
    "## 1. hmmbuild\n",
    "We build an HMM of each disordered region, using trimmed MSA as an input. We will build the models separately for BLAST and ClustalOmega to be able to compare the quality of the models based on the local and the global alignment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set the paths to MSA (input) and HMM (output) files - BLAST\n",
    "msa_dir_blast = '{}/results/alignments/output_files/disordered/blast'.format(directory)\n",
    "hmmbuild_dir_blast = '{}/results/hmms/hmmbuild/blast'.format(directory)\n",
    "\n",
    "# Set the paths to  MSA (input) and HMM (output) files - ClustalOmega\n",
    "msa_dir_clustal = '{}/results/alignments/output_files/disordered/clustal'.format(directory)\n",
    "hmmbuild_dir_clustal = '{}/results/hmms/hmmbuild/clustal'.format(directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/S7W634_1-32.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/S7W634_1-32.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     S7W634_1-32            200    32    32    75.11  1.689 \n",
      "\n",
      "# CPU time: 0.03u 0.00s 00:00:00.03 Elapsed: 00:00:00.03\n",
      "hmmbuild completed for S7W634_1-32.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q8R464_25-120.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q8R464_25-120.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q8R464_25-120          200    96    96     1.03  0.595 \n",
      "\n",
      "# CPU time: 0.04u 0.00s 00:00:00.04 Elapsed: 00:00:00.04\n",
      "hmmbuild completed for Q8R464_25-120.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/P01097_28-50.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/P01097_28-50.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P01097_28-50           200    23    23    39.15  2.309 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for P01097_28-50.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q9Z2F5_306-430.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q9Z2F5_306-430.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q9Z2F5_306-430         200   125   125     0.41  0.586 \n",
      "\n",
      "# CPU time: 0.04u 0.00s 00:00:00.04 Elapsed: 00:00:00.04\n",
      "hmmbuild completed for Q9Z2F5_306-430.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/P01019_436-450.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/P01019_436-450.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P01019_436-450         200    15    15   200.00  2.870 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for P01019_436-450.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q8K4J6_67-98.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q8K4J6_67-98.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q8K4J6_67-98           200    32    32     4.31  1.689 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q8K4J6_67-98.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/A8AZZ3_116-134.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/A8AZZ3_116-134.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     A8AZZ3_116-134         106    19    19   106.00  2.338 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for A8AZZ3_116-134.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/A8AZZ3_146-195.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/A8AZZ3_146-195.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     A8AZZ3_146-195         106    50    50     6.62  1.106 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for A8AZZ3_146-195.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/P00736_193-305.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/P00736_193-305.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P00736_193-305         200   113   113     1.10  0.592 \n",
      "\n",
      "# CPU time: 0.04u 0.00s 00:00:00.04 Elapsed: 00:00:00.04\n",
      "hmmbuild completed for P00736_193-305.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/O35274_1-154.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/O35274_1-154.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O35274_1-154           200   154   154     0.88  0.591 \n",
      "\n",
      "# CPU time: 0.05u 0.00s 00:00:00.05 Elapsed: 00:00:00.05\n",
      "hmmbuild completed for O35274_1-154.fasta\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q9H832_1-99.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q9H832_1-99.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q9H832_1-99            200    99    99     2.96  0.590 \n",
      "\n",
      "# CPU time: 0.04u 0.00s 00:00:00.04 Elapsed: 00:00:00.04\n",
      "hmmbuild completed for Q9H832_1-99.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/O00585_71-111.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/O00585_71-111.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O00585_71-111          200    41    41     4.65  1.336 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for O00585_71-111.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/J8TM36_236-249.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/J8TM36_236-249.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     J8TM36_236-249         200    14    14   200.00  0.784 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for J8TM36_236-249.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q86FP8_20-80.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q86FP8_20-80.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q86FP8_20-80            14    61    61     2.96  0.916 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for Q86FP8_20-80.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/O14727_1-104.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/O14727_1-104.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O14727_1-104           200   104   104     0.55  0.589 \n",
      "\n",
      "# CPU time: 0.04u 0.00s 00:00:00.04 Elapsed: 00:00:00.04\n",
      "hmmbuild completed for O14727_1-104.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q93KQ4_51-81.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q93KQ4_51-81.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q93KQ4_51-81            73    31    31    73.00  1.727 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for Q93KQ4_51-81.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/O15922_1-89.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/O15922_1-89.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O15922_1-89            200    89    89     6.31  0.640 \n",
      "\n",
      "# CPU time: 0.03u 0.00s 00:00:00.03 Elapsed: 00:00:00.03\n",
      "hmmbuild completed for O15922_1-89.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q84852_17-299.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q84852_17-299.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q84852_17-299          200   283   283     1.50  0.590 \n",
      "\n",
      "# CPU time: 0.10u 0.00s 00:00:00.10 Elapsed: 00:00:00.10\n",
      "hmmbuild completed for Q84852_17-299.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q8K4J6_111-142.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q8K4J6_111-142.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q8K4J6_111-142         200    32    32     4.15  1.689 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q8K4J6_111-142.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q9H832_327-354.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q9H832_327-354.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q9H832_327-354         200    28    28     5.00  1.917 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q9H832_327-354.fasta\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q8K4J6_155-186.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q8K4J6_155-186.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q8K4J6_155-186         200    32    32     3.93  1.689 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q8K4J6_155-186.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/A1L1Q4_1-71.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/A1L1Q4_1-71.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     A1L1Q4_1-71            200    71    71     2.31  0.793 \n",
      "\n",
      "# CPU time: 0.03u 0.00s 00:00:00.03 Elapsed: 00:00:00.03\n",
      "hmmbuild completed for A1L1Q4_1-71.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/A4L7I2_1658-1856.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/A4L7I2_1658-1856.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     A4L7I2_1658-1856       200   199   195     0.44  0.593 \n",
      "\n",
      "# CPU time: 0.07u 0.00s 00:00:00.07 Elapsed: 00:00:00.07\n",
      "hmmbuild completed for A4L7I2_1658-1856.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/P04370-5_1-169.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/P04370-5_1-169.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P04370-5_1-169         201   169   169     0.46  0.590 \n",
      "\n",
      "# CPU time: 0.06u 0.00s 00:00:00.06 Elapsed: 00:00:00.06\n",
      "hmmbuild completed for P04370-5_1-169.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q8IU57_270-299.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q8IU57_270-299.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q8IU57_270-299         200    30    30     9.72  1.796 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q8IU57_270-299.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/O88339_144-575.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/O88339_144-575.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O88339_144-575         200   432   432     0.50  0.591 \n",
      "\n",
      "# CPU time: 0.15u 0.00s 00:00:00.15 Elapsed: 00:00:00.15\n",
      "hmmbuild completed for O88339_144-575.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q5VZK9_1005-1020.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q5VZK9_1005-1020.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q5VZK9_1005-1020       200    16    16    11.45  3.255 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q5VZK9_1005-1020.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q8WUG5_1-105.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q8WUG5_1-105.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q8WUG5_1-105           200   105   105     0.41  0.587 \n",
      "\n",
      "# CPU time: 0.04u 0.00s 00:00:00.04 Elapsed: 00:00:00.04\n",
      "hmmbuild completed for Q8WUG5_1-105.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/O00308_662-702.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/O00308_662-702.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O00308_662-702         200    41    41     1.02  1.338 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for O00308_662-702.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q8IW19_399-420.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q8IW19_399-420.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q8IW19_399-420         200    22    22     5.86  2.408 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q8IW19_399-420.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/O15922_230-240.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/O15922_230-240.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O15922_230-240         200    11    11   200.00  1.033 \n",
      "\n",
      "# CPU time: 0.00u 0.00s 00:00:00.00 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for O15922_230-240.fasta\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/O43474_1-130.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/O43474_1-130.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O43474_1-130           200   130   129     1.11  0.591 \n",
      "\n",
      "# CPU time: 0.05u 0.00s 00:00:00.05 Elapsed: 00:00:00.05\n",
      "hmmbuild completed for O43474_1-130.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/S6B291_240-465.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/S6B291_240-465.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     S6B291_240-465         200   226   226     1.18  0.589 \n",
      "\n",
      "# CPU time: 0.08u 0.00s 00:00:00.08 Elapsed: 00:00:00.08\n",
      "hmmbuild completed for S6B291_240-465.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/O35274_417-494.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/O35274_417-494.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O35274_417-494         200    78    78     0.74  0.727 \n",
      "\n",
      "# CPU time: 0.03u 0.00s 00:00:00.03 Elapsed: 00:00:00.03\n",
      "hmmbuild completed for O35274_417-494.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q9H0E2_1-53.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q9H0E2_1-53.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q9H0E2_1-53            200    53    53     2.32  1.046 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for Q9H0E2_1-53.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/P07342_647-687.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/P07342_647-687.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P07342_647-687         200    41    41     3.76  1.335 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for P07342_647-687.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/O43791_357-374.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/O43791_357-374.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O43791_357-374         200    18    18     2.71  2.912 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for O43791_357-374.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/A8AZZ3_24-44.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/A8AZZ3_24-44.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     A8AZZ3_24-44           106    21    21   106.00  1.637 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for A8AZZ3_24-44.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/P00742_90-126.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/P00742_90-126.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P00742_90-126          200    37    37     3.13  1.472 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for P00742_90-126.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/P01019_44-64.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/P01019_44-64.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P01019_44-64           200    21    21    25.91  2.517 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for P01019_44-64.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q99967_220-269.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q99967_220-269.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q99967_220-269         200    50    50     0.82  1.107 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for Q99967_220-269.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q9BYF1_769-805.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q9BYF1_769-805.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q9BYF1_769-805         200    37    37     3.20  1.472 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q9BYF1_769-805.fasta\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q8IW19_450-511.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q8IW19_450-511.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q8IW19_450-511         200    62    62     1.25  0.902 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for Q8IW19_450-511.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/O15922_315-324.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/O15922_315-324.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O15922_315-324         200    10    10   200.00  1.101 \n",
      "\n",
      "# CPU time: 0.00u 0.00s 00:00:00.00 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for O15922_315-324.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/O35274_584-602.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/O35274_584-602.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O35274_584-602         200    19    19     3.01  2.767 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for O35274_584-602.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q9CXY6_29-44.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q9CXY6_29-44.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q9CXY6_29-44           200    16    16     3.24  3.256 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q9CXY6_29-44.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/P07342_580-595.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/P07342_580-595.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P07342_580-595         200    16    16    65.67  3.255 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for P07342_580-595.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q9BYI3_149-253.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q9BYI3_149-253.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q9BYI3_149-253         200   105   105     0.41  0.589 \n",
      "\n",
      "# CPU time: 0.04u 0.00s 00:00:00.04 Elapsed: 00:00:00.04\n",
      "hmmbuild completed for Q9BYI3_149-253.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/P00392_550-561.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/P00392_550-561.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P00392_550-561         200    12    12   200.00  4.167 \n",
      "\n",
      "# CPU time: 0.00u 0.00s 00:00:00.00 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for P00392_550-561.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q9CXY6_347-390.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q9CXY6_347-390.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q9CXY6_347-390         200    44    44     1.08  1.246 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for Q9CXY6_347-390.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/O43791_169-178.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/O43791_169-178.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O43791_169-178         200    10    10   200.00  4.515 \n",
      "\n",
      "# CPU time: 0.00u 0.00s 00:00:00.00 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for O43791_169-178.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/Q5T4W7_108-120.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/Q5T4W7_108-120.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q5T4W7_108-120         200    13    13   200.00  1.349 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for Q5T4W7_108-120.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/blast/O14958_371-399.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/blast/O14958_371-399.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O14958_371-399         200    29    28     2.79  1.917 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for O14958_371-399.fasta\n"
     ]
    }
   ],
   "source": [
    "# Build HMM for BLAST\n",
    "for filename in os.listdir(msa_dir_blast):\n",
    "    if filename.endswith('.fasta'):\n",
    "        input_file = os.path.join(msa_dir_blast, filename)\n",
    "        output_file = os.path.join(hmmbuild_dir_blast, os.path.splitext(filename)[0] + '.hmm')\n",
    "\n",
    "        subprocess.run(['hmmbuild', output_file, input_file])\n",
    "        print('hmmbuild completed for {}'.format(filename))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/S7W634_1-32.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/S7W634_1-32.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     S7W634_1-32            196    32    31   196.00  1.341 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for S7W634_1-32.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q8R464_25-120.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q8R464_25-120.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q8R464_25-120          194    96    96     0.41  0.598 \n",
      "\n",
      "# CPU time: 0.04u 0.00s 00:00:00.04 Elapsed: 00:00:00.04\n",
      "hmmbuild completed for Q8R464_25-120.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/P01097_28-50.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/P01097_28-50.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P01097_28-50           195    23    23    28.85  2.309 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for P01097_28-50.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q9Z2F5_306-430.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q9Z2F5_306-430.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q9Z2F5_306-430         192   125   125     0.41  0.588 \n",
      "\n",
      "# CPU time: 0.04u 0.00s 00:00:00.04 Elapsed: 00:00:00.04\n",
      "hmmbuild completed for Q9Z2F5_306-430.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/P01019_436-450.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/P01019_436-450.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P01019_436-450         192    15    14   192.00  1.835 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for P01019_436-450.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q8K4J6_67-98.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q8K4J6_67-98.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q8K4J6_67-98           194    32    32     2.39  1.690 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q8K4J6_67-98.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/A8AZZ3_116-134.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/A8AZZ3_116-134.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     A8AZZ3_116-134         102    19    19   102.00  1.979 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for A8AZZ3_116-134.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/A8AZZ3_146-195.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/A8AZZ3_146-195.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     A8AZZ3_146-195         102    50    46    11.64  1.197 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for A8AZZ3_146-195.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/P00736_193-305.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/P00736_193-305.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P00736_193-305         190   113   113     0.43  0.587 \n",
      "\n",
      "# CPU time: 0.04u 0.00s 00:00:00.04 Elapsed: 00:00:00.04\n",
      "hmmbuild completed for P00736_193-305.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/O35274_1-154.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/O35274_1-154.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O35274_1-154           193   154   153     2.02  0.590 \n",
      "\n",
      "# CPU time: 0.05u 0.00s 00:00:00.05 Elapsed: 00:00:00.05\n",
      "hmmbuild completed for O35274_1-154.fasta\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q9H832_1-99.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q9H832_1-99.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q9H832_1-99            190    99    92     7.31  0.620 \n",
      "\n",
      "# CPU time: 0.04u 0.00s 00:00:00.04 Elapsed: 00:00:00.04\n",
      "hmmbuild completed for Q9H832_1-99.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/O00585_71-111.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/O00585_71-111.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O00585_71-111          191    41    41     6.53  1.335 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for O00585_71-111.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/J8TM36_236-249.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/J8TM36_236-249.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     J8TM36_236-249         186    14    12   186.00  0.703 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for J8TM36_236-249.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q86FP8_20-80.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q86FP8_20-80.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q86FP8_20-80            13    61    61     5.15  0.916 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for Q86FP8_20-80.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/O14727_1-104.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/O14727_1-104.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O14727_1-104           193   104   104     0.56  0.591 \n",
      "\n",
      "# CPU time: 0.04u 0.00s 00:00:00.04 Elapsed: 00:00:00.04\n",
      "hmmbuild completed for O14727_1-104.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q93KQ4_51-81.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q93KQ4_51-81.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q93KQ4_51-81            62    31    31    56.86  1.740 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q93KQ4_51-81.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/O15922_1-89.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/O15922_1-89.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O15922_1-89            181    89    72    44.89  0.783 \n",
      "\n",
      "# CPU time: 0.03u 0.00s 00:00:00.03 Elapsed: 00:00:00.03\n",
      "hmmbuild completed for O15922_1-89.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q84852_17-299.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q84852_17-299.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q84852_17-299          200   283   280     0.77  0.590 \n",
      "\n",
      "# CPU time: 0.10u 0.00s 00:00:00.10 Elapsed: 00:00:00.10\n",
      "hmmbuild completed for Q84852_17-299.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q8K4J6_111-142.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q8K4J6_111-142.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q8K4J6_111-142         194    32    32     2.17  1.689 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q8K4J6_111-142.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q9H832_327-354.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q9H832_327-354.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q9H832_327-354         190    28    28     7.71  1.917 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q9H832_327-354.fasta\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q8K4J6_155-186.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q8K4J6_155-186.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q8K4J6_155-186         194    32    32     1.63  1.688 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q8K4J6_155-186.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/A1L1Q4_1-71.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/A1L1Q4_1-71.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     A1L1Q4_1-71            191    71    71     1.63  0.794 \n",
      "\n",
      "# CPU time: 0.03u 0.00s 00:00:00.03 Elapsed: 00:00:00.03\n",
      "hmmbuild completed for A1L1Q4_1-71.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/A4L7I2_1658-1856.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/A4L7I2_1658-1856.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     A4L7I2_1658-1856       200   199   195     0.44  0.593 \n",
      "\n",
      "# CPU time: 0.07u 0.00s 00:00:00.07 Elapsed: 00:00:00.07\n",
      "hmmbuild completed for A4L7I2_1658-1856.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/P04370-5_1-169.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/P04370-5_1-169.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P04370-5_1-169         192   169   167     0.52  0.592 \n",
      "\n",
      "# CPU time: 0.06u 0.00s 00:00:00.06 Elapsed: 00:00:00.06\n",
      "hmmbuild completed for P04370-5_1-169.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q8IU57_270-299.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q8IU57_270-299.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q8IU57_270-299         190    30    30    10.09  1.795 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q8IU57_270-299.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/O88339_144-575.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/O88339_144-575.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O88339_144-575         190   432   432     0.54  0.592 \n",
      "\n",
      "# CPU time: 0.15u 0.00s 00:00:00.15 Elapsed: 00:00:00.15\n",
      "hmmbuild completed for O88339_144-575.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q5VZK9_1005-1020.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q5VZK9_1005-1020.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q5VZK9_1005-1020       195    16    16    11.46  3.256 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q5VZK9_1005-1020.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q8WUG5_1-105.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q8WUG5_1-105.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q8WUG5_1-105           189   105   104     0.46  0.590 \n",
      "\n",
      "# CPU time: 0.04u 0.00s 00:00:00.04 Elapsed: 00:00:00.04\n",
      "hmmbuild completed for Q8WUG5_1-105.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/O00308_662-702.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/O00308_662-702.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O00308_662-702         191    41    41     1.01  1.333 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for O00308_662-702.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q8IW19_399-420.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q8IW19_399-420.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q8IW19_399-420         190    22    22     7.59  2.408 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q8IW19_399-420.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/O15922_230-240.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/O15922_230-240.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O15922_230-240         181    11    11   181.00  1.988 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for O15922_230-240.fasta\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/O43474_1-130.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/O43474_1-130.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O43474_1-130           188   130   129     0.59  0.588 \n",
      "\n",
      "# CPU time: 0.05u 0.00s 00:00:00.05 Elapsed: 00:00:00.05\n",
      "hmmbuild completed for O43474_1-130.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/S6B291_240-465.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/S6B291_240-465.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     S6B291_240-465         193   226   221     0.64  0.588 \n",
      "\n",
      "# CPU time: 0.08u 0.00s 00:00:00.08 Elapsed: 00:00:00.08\n",
      "hmmbuild completed for S6B291_240-465.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/O35274_417-494.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/O35274_417-494.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O35274_417-494         193    78    78     0.73  0.726 \n",
      "\n",
      "# CPU time: 0.03u 0.00s 00:00:00.03 Elapsed: 00:00:00.03\n",
      "hmmbuild completed for O35274_417-494.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q9H0E2_1-53.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q9H0E2_1-53.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q9H0E2_1-53            195    53    53     3.17  1.047 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for Q9H0E2_1-53.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/P07342_647-687.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/P07342_647-687.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P07342_647-687         194    41    41     3.31  1.335 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for P07342_647-687.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/O43791_357-374.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/O43791_357-374.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O43791_357-374         193    18    18     2.71  2.911 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for O43791_357-374.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/A8AZZ3_24-44.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/A8AZZ3_24-44.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     A8AZZ3_24-44           102    21    18   102.00  1.918 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for A8AZZ3_24-44.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/P00742_90-126.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/P00742_90-126.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P00742_90-126          192    37    37     1.46  1.472 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for P00742_90-126.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/P01019_44-64.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/P01019_44-64.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P01019_44-64           192    21    21    75.23  2.517 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for P01019_44-64.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q99967_220-269.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q99967_220-269.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q99967_220-269         193    50    50     0.81  1.108 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for Q99967_220-269.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q9BYF1_769-805.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q9BYF1_769-805.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q9BYF1_769-805         196    37    37     5.03  1.472 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q9BYF1_769-805.fasta\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q8IW19_450-511.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q8IW19_450-511.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q8IW19_450-511         190    62    62     6.91  0.902 \n",
      "\n",
      "# CPU time: 0.03u 0.00s 00:00:00.03 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for Q8IW19_450-511.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/O15922_315-324.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/O15922_315-324.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O15922_315-324         181    10     6   181.00  1.331 \n",
      "\n",
      "# CPU time: 0.00u 0.00s 00:00:00.00 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for O15922_315-324.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/O35274_584-602.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/O35274_584-602.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O35274_584-602         193    19    19     3.01  2.768 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for O35274_584-602.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q9CXY6_29-44.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q9CXY6_29-44.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q9CXY6_29-44           195    16    16     3.25  3.256 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q9CXY6_29-44.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/P07342_580-595.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/P07342_580-595.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P07342_580-595         194    16    16     4.21  3.256 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for P07342_580-595.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q9BYI3_149-253.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q9BYI3_149-253.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q9BYI3_149-253         192   105   105     0.41  0.592 \n",
      "\n",
      "# CPU time: 0.04u 0.00s 00:00:00.04 Elapsed: 00:00:00.04\n",
      "hmmbuild completed for Q9BYI3_149-253.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/P00392_550-561.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/P00392_550-561.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     P00392_550-561         185    12    12   185.00  4.117 \n",
      "\n",
      "# CPU time: 0.00u 0.00s 00:00:00.00 Elapsed: 00:00:00.00\n",
      "hmmbuild completed for P00392_550-561.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q9CXY6_347-390.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q9CXY6_347-390.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q9CXY6_347-390         195    44    44     1.14  1.248 \n",
      "\n",
      "# CPU time: 0.02u 0.00s 00:00:00.02 Elapsed: 00:00:00.02\n",
      "hmmbuild completed for Q9CXY6_347-390.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/O43791_169-178.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/O43791_169-178.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O43791_169-178         193    10    10   193.00  4.515 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for O43791_169-178.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/Q5T4W7_108-120.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/Q5T4W7_108-120.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Q5T4W7_108-120         190    13    12   190.00  1.498 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for Q5T4W7_108-120.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /Users/alina/HMM/results/alignments/output_files/disordered/clustal/O14958_371-399.fasta\n",
      "# output HMM file:                  /Users/alina/HMM/results/hmms/hmmbuild/clustal/O14958_371-399.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     O14958_371-399         194    29    28    33.25  1.917 \n",
      "\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.01\n",
      "hmmbuild completed for O14958_371-399.fasta\n"
     ]
    }
   ],
   "source": [
    "# Build HMM for ClustalOmega\n",
    "for filename in os.listdir(msa_dir_clustal):\n",
    "    if filename.endswith('.fasta'):\n",
    "        input_file = os.path.join(msa_dir_clustal, filename)\n",
    "        output_file = os.path.join(hmmbuild_dir_clustal, os.path.splitext(filename)[0] + '.hmm')\n",
    "\n",
    "        subprocess.run(['hmmbuild', output_file, input_file])\n",
    "        print('hmmbuild completed for {}'.format(filename))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. hmmsearch\n",
    "\n",
    "After building the models, our objective is to assess if overlaps with the profiles in the sequence database (Reference Proteome 75%) exist. We retrive the data containing the most significant sequences, with a default E-value threshold of 0.01.\n",
    "\n",
    "The same as before, we do it for BLAST and ClustalOmega separately."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Copy the HMMs from the BLAST folder\n",
    "# for filename in os.listdir(hmmbuild_dir_blast):\n",
    "#     source_path = os.path.join(hmmbuild_dir_blast, filename)\n",
    "#     destination_path = '{}@{}:~/hmm/blast/hmmbuild/{}'.format(name, server, filename)\n",
    "#     subprocess.run(['scp', source_path, destination_path])\n",
    "#     print('File {} copied to {}'.format(filename, destination_path))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Copy the HMMs from the ClustalOmega folder\n",
    "# for filename in os.listdir(hmmbuild_dir_clustal):\n",
    "#     source_path = os.path.join(hmmbuild_dir_clustal, filename)\n",
    "#     destination_path = '{}@{}:~/hmm/clustal/hmmbuild/{}'.format(name, server, filename)\n",
    "#     subprocess.run(['scp', source_path, destination_path])\n",
    "#     print('File {} copied to {}'.format(filename, destination_path))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After copying the files, we run `hmmsearch` command. The following code should be uncommented if there are new files provided. However, it requires some time to retrieve the output."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Construct the source and the target directories for BLAST - remote computer\n",
    "blast_in_dir_rem = '/home/{}/hmm/blast/hmmbuild'.format(name)\n",
    "blast_out_dir_rem = '/home/{}/hmm/blast/hmmsearch'.format(name)\n",
    "\n",
    "# List files in the source directory on the remote machine\n",
    "ls_command = 'ssh {}@{} \"ls {}\"'.format(name, server, blast_in_dir_rem)\n",
    "file_list = get_ipython().getoutput(ls_command)\n",
    "\n",
    "# # Iterate over each filename - RUNS FOR A LONG TIME\n",
    "# for filename in file_list:\n",
    "#     filename = filename.strip()\n",
    "#     source_path = '{}/{}'.format(blast_in_dir_rem, filename)\n",
    "#     dest_filename = os.path.splitext(filename)[0]\n",
    "#     destination_path = '{}/hmmsearch_blast_{}.txt'.format(blast_out_dir_rem, dest_filename)\n",
    "    \n",
    "#     # Run hmmsearch command using ssh without redirection\n",
    "# #     !ssh {name}@{server} \"/software/packages/hmmer/hmmer-3.3.2/usr/bin/hmmsearch {source_path} /db/rp/rp-seqs-75.fasta.gz > {destination_path} 2>&1\"\n",
    "#     command = 'ssh {}@{} \"/software/packages/hmmer/hmmer-3.3.2/usr/bin/hmmsearch {} /db/rp/rp-seqs-75.fasta.gz > {} 2>&1\"'.format(name, server, source_path, destination_path)\n",
    "#     !{command}   \n",
    "        \n",
    "#     print('hmmsearch completed for {} - Results saved to {}'.format(filename, destination_path))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Copy BLAST results to the local folder\n",
    "hmmsearch_dir_blast = '{}/results/hmms/hmmsearch/blast'.format(directory)\n",
    "\n",
    "# for filename in file_list:\n",
    "#     filename = filename.strip()\n",
    "#     source_path = '{}/{}'.format(blast_in_dir_rem, filename)\n",
    "#     dest_filename = os.path.splitext(filename)[0]\n",
    "    \n",
    "#     # Copy files\n",
    "#     !scp {name}@{server}:~/hmm/blast/hmmsearch/hmmsearch_blast_{dest_filename}.txt {hmmsearch_dir_blast}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Construct the source and the target directories for ClustalOmega\n",
    "clustal_in_dir_rem = '/home/{}/hmm/clustal/hmmbuild'.format(name)\n",
    "clustal_out_dir_rem = '/home/{}/hmm/clustal/hmmsearch'.format(name)\n",
    "\n",
    "# List files in the source directory on the remote machine\n",
    "ls_command = 'ssh {}@{} \"ls {}\"'.format(name, server, clustal_in_dir_rem)\n",
    "file_list = get_ipython().getoutput(ls_command)\n",
    "\n",
    "# # Iterate over each filename\n",
    "# for filename in file_list:\n",
    "#     filename = filename.strip()\n",
    "#     source_path = '{}/{}'.format(clustal_in_dir_rem, filename)\n",
    "#     dest_filename = os.path.splitext(filename)[0]\n",
    "#     destination_path = '{}/hmmsearch_clustal_{}.txt'.format(clustal_out_dir_rem, dest_filename)\n",
    "    \n",
    "#     # Run hmmsearch command using ssh without redirection\n",
    "# #     !ssh {name}@{server} \"/software/packages/hmmer/hmmer-3.3.2/usr/bin/hmmsearch {source_path} /db/rp/rp-seqs-75.fasta.gz > {destination_path} 2>&1\"\n",
    "#     command = 'ssh {}@{} \"/software/packages/hmmer/hmmer-3.3.2/usr/bin/hmmsearch {} /db/rp/rp-seqs-75.fasta.gz > {} 2>&1\"'.format(name, server, source_path, destination_path)\n",
    "#     !{command}     \n",
    "#     print('hmmsearch completed for {} - Results saved to {}'.format(filename, destination_path))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Copy ClustalOmega results to the local folder\n",
    "hmmsearch_dir_clustal = '{}/results/hmms/hmmsearch/clustal'.format(directory)\n",
    "\n",
    "# for filename in file_list:\n",
    "#     filename = filename.strip()\n",
    "#     source_path = '{}/{}'.format(clustal_in_dir_rem, filename)\n",
    "#     dest_filename = os.path.splitext(filename)[0]\n",
    "    \n",
    "#     # Copy files\n",
    "#     !scp {name}@{server}:~/hmm/clustal/hmmsearch/hmmsearch_clustal_{dest_filename}.txt {hmmsearch_dir_clustal}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error processing file hmmsearch_blast_O15922_230-240.txt: '  ------ inclusion threshold ------\\n' is not in list\n",
      "Error processing file hmmsearch_blast_O15922_315-324.txt: '  ------ inclusion threshold ------\\n' is not in list\n",
      "Error processing file hmmsearch_blast_J8TM36_236-249.txt: '  ------ inclusion threshold ------\\n' is not in list\n",
      "Error processing file hmmsearch_blast_Q5T4W7_108-120.txt: '  ------ inclusion threshold ------\\n' is not in list\n",
      "The number of retrieved hmmsearch results for BLAST: 186445\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>E-value</th>\n",
       "      <th>score</th>\n",
       "      <th>bias</th>\n",
       "      <th>E-value</th>\n",
       "      <th>score</th>\n",
       "      <th>bias</th>\n",
       "      <th>exp</th>\n",
       "      <th>N</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>hmm_from</th>\n",
       "      <th>hmm_to</th>\n",
       "      <th>hmm_length</th>\n",
       "      <th>ali_from</th>\n",
       "      <th>ali_to</th>\n",
       "      <th>ali_length</th>\n",
       "      <th>env_from</th>\n",
       "      <th>env_to</th>\n",
       "      <th>env_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>5406.7</td>\n",
       "      <td>5695.0</td>\n",
       "      <td>3.6e-05</td>\n",
       "      <td>36.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>378.4</td>\n",
       "      <td>365</td>\n",
       "      <td>A0A7M7NU54</td>\n",
       "      <td>1</td>\n",
       "      <td>35</td>\n",
       "      <td>35</td>\n",
       "      <td>246</td>\n",
       "      <td>282</td>\n",
       "      <td>37</td>\n",
       "      <td>246</td>\n",
       "      <td>283</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>2473.7</td>\n",
       "      <td>2447.2</td>\n",
       "      <td>3.9e-05</td>\n",
       "      <td>35.9</td>\n",
       "      <td>5.8</td>\n",
       "      <td>148.9</td>\n",
       "      <td>146</td>\n",
       "      <td>A0A7I8W901</td>\n",
       "      <td>1</td>\n",
       "      <td>34</td>\n",
       "      <td>34</td>\n",
       "      <td>273</td>\n",
       "      <td>306</td>\n",
       "      <td>34</td>\n",
       "      <td>273</td>\n",
       "      <td>308</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>2023.3</td>\n",
       "      <td>1971.1</td>\n",
       "      <td>8.5e-06</td>\n",
       "      <td>38.0</td>\n",
       "      <td>8.9</td>\n",
       "      <td>138.2</td>\n",
       "      <td>128</td>\n",
       "      <td>A0A8B7ZVU6</td>\n",
       "      <td>1</td>\n",
       "      <td>32</td>\n",
       "      <td>32</td>\n",
       "      <td>454</td>\n",
       "      <td>485</td>\n",
       "      <td>32</td>\n",
       "      <td>454</td>\n",
       "      <td>487</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>2013.1</td>\n",
       "      <td>1940.1</td>\n",
       "      <td>8.4e-06</td>\n",
       "      <td>38.0</td>\n",
       "      <td>8.9</td>\n",
       "      <td>136.2</td>\n",
       "      <td>127</td>\n",
       "      <td>A0A8B7ZRF1</td>\n",
       "      <td>1</td>\n",
       "      <td>32</td>\n",
       "      <td>32</td>\n",
       "      <td>382</td>\n",
       "      <td>413</td>\n",
       "      <td>32</td>\n",
       "      <td>382</td>\n",
       "      <td>415</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>1740.0</td>\n",
       "      <td>1752.7</td>\n",
       "      <td>7.7e-06</td>\n",
       "      <td>38.1</td>\n",
       "      <td>8.9</td>\n",
       "      <td>122.2</td>\n",
       "      <td>112</td>\n",
       "      <td>A0A8B7ZPY7</td>\n",
       "      <td>1</td>\n",
       "      <td>32</td>\n",
       "      <td>32</td>\n",
       "      <td>454</td>\n",
       "      <td>485</td>\n",
       "      <td>32</td>\n",
       "      <td>454</td>\n",
       "      <td>487</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  E-value   score    bias  E-value score bias    exp    N    Sequence  \\\n",
       "0       0  5406.7  5695.0  3.6e-05  36.0  9.0  378.4  365  A0A7M7NU54   \n",
       "1       0  2473.7  2447.2  3.9e-05  35.9  5.8  148.9  146  A0A7I8W901   \n",
       "2       0  2023.3  1971.1  8.5e-06  38.0  8.9  138.2  128  A0A8B7ZVU6   \n",
       "3       0  2013.1  1940.1  8.4e-06  38.0  8.9  136.2  127  A0A8B7ZRF1   \n",
       "4       0  1740.0  1752.7  7.7e-06  38.1  8.9  122.2  112  A0A8B7ZPY7   \n",
       "\n",
       "  hmm_from hmm_to hmm_length ali_from ali_to ali_length env_from env_to  \\\n",
       "0        1     35         35      246    282         37      246    283   \n",
       "1        1     34         34      273    306         34      273    308   \n",
       "2        1     32         32      454    485         32      454    487   \n",
       "3        1     32         32      382    413         32      382    415   \n",
       "4        1     32         32      454    485         32      454    487   \n",
       "\n",
       "  env_length  \n",
       "0         38  \n",
       "1         36  \n",
       "2         34  \n",
       "3         34  \n",
       "4         34  "
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Preprocess hmmsearch results - BLAST\n",
    "hmmsearch_dir_blast = '{}/results/hmms/hmmsearch/blast'.format(directory)\n",
    "hmmsearch_dir_blast_prepr_output = '{}/results/hmms/hmmsearch/blast/hmmsearch_df'.format(directory)\n",
    "\n",
    "blast_results_list = []\n",
    "\n",
    "for filename in os.listdir(hmmsearch_dir_blast):\n",
    "    if filename.endswith('.txt'):\n",
    "        file_path = os.path.join(hmmsearch_dir_blast, filename)\n",
    "        parts = filename.split('_')\n",
    "        protein_id = '{}_{}'.format(parts[2], parts[3].split(\".\")[0])\n",
    "        try:\n",
    "            blast_stat = process_hmmsearch_file(file_path) \n",
    "            blast_reg = extract_table_from_output(file_path)\n",
    "            blast_result = pd.merge(blast_stat, blast_reg, left_on='Sequence', right_on='id', how='inner')\n",
    "            blast_result = blast_result.drop(columns=['Description', 'id'])\n",
    "            \n",
    "            # Save individual DataFrames to a list\n",
    "            blast_results_list.append(blast_result)\n",
    "        \n",
    "            # Save the DataFrame to a CSV file\n",
    "            output_file = os.path.join(hmmsearch_dir_blast_prepr_output, 'hmmsearch_blast_df_{}.csv'.format(protein_id))\n",
    "            blast_result.to_csv(output_file, index=False)\n",
    "            \n",
    "        except ValueError as e:\n",
    "            print('Error processing file {}: {}'.format(filename, e))\n",
    "            \n",
    "# Merge all DataFrames into one\n",
    "hmmsearch_results_blast = pd.concat(blast_results_list, ignore_index=True)\n",
    "\n",
    "# Save the merged DataFrame to a CSV file\n",
    "merged_output_file = os.path.join('{}/results/hmms/hmmsearch/'.format(directory), 'hmmsearch_results_blast.csv')\n",
    "hmmsearch_results_blast.to_csv(merged_output_file, index=False)\n",
    "print('The number of retrieved hmmsearch results for BLAST: {}'.format(len(hmmsearch_results_blast)))\n",
    "hmmsearch_results_blast.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error processing file hmmsearch_clustal_J8TM36_236-249.txt: '  ------ inclusion threshold ------\\n' is not in list\n",
      "Error processing file hmmsearch_clustal_O15922_315-324.txt: '  ------ inclusion threshold ------\\n' is not in list\n",
      "Error processing file hmmsearch_clustal_Q5T4W7_108-120.txt: '  ------ inclusion threshold ------\\n' is not in list\n",
      "The number of retrieved hmmsearch results for ClustalOmega: 206952\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>E-value</th>\n",
       "      <th>score</th>\n",
       "      <th>bias</th>\n",
       "      <th>E-value</th>\n",
       "      <th>score</th>\n",
       "      <th>bias</th>\n",
       "      <th>exp</th>\n",
       "      <th>N</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>hmm_from</th>\n",
       "      <th>hmm_to</th>\n",
       "      <th>hmm_length</th>\n",
       "      <th>ali_from</th>\n",
       "      <th>ali_to</th>\n",
       "      <th>ali_length</th>\n",
       "      <th>env_from</th>\n",
       "      <th>env_to</th>\n",
       "      <th>env_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7.1e-32</td>\n",
       "      <td>121.2</td>\n",
       "      <td>13.5</td>\n",
       "      <td>8.1e-32</td>\n",
       "      <td>121.0</td>\n",
       "      <td>13.5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>A0A8C7BFT3</td>\n",
       "      <td>1</td>\n",
       "      <td>61</td>\n",
       "      <td>61</td>\n",
       "      <td>9</td>\n",
       "      <td>69</td>\n",
       "      <td>61</td>\n",
       "      <td>9</td>\n",
       "      <td>69</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.2e-30</td>\n",
       "      <td>117.3</td>\n",
       "      <td>13.5</td>\n",
       "      <td>5.9e-30</td>\n",
       "      <td>115.1</td>\n",
       "      <td>13.5</td>\n",
       "      <td>2.3</td>\n",
       "      <td>1</td>\n",
       "      <td>A0A2U3ZCX9</td>\n",
       "      <td>1</td>\n",
       "      <td>61</td>\n",
       "      <td>61</td>\n",
       "      <td>451</td>\n",
       "      <td>511</td>\n",
       "      <td>61</td>\n",
       "      <td>451</td>\n",
       "      <td>511</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.2e-30</td>\n",
       "      <td>117.2</td>\n",
       "      <td>13.5</td>\n",
       "      <td>5.9e-30</td>\n",
       "      <td>115.0</td>\n",
       "      <td>13.5</td>\n",
       "      <td>2.3</td>\n",
       "      <td>1</td>\n",
       "      <td>A0A2U3VQT6</td>\n",
       "      <td>1</td>\n",
       "      <td>61</td>\n",
       "      <td>61</td>\n",
       "      <td>454</td>\n",
       "      <td>514</td>\n",
       "      <td>61</td>\n",
       "      <td>454</td>\n",
       "      <td>514</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.3e-30</td>\n",
       "      <td>117.2</td>\n",
       "      <td>13.5</td>\n",
       "      <td>5.9e-30</td>\n",
       "      <td>115.1</td>\n",
       "      <td>13.5</td>\n",
       "      <td>2.2</td>\n",
       "      <td>1</td>\n",
       "      <td>A0A2Y9JGP3</td>\n",
       "      <td>1</td>\n",
       "      <td>61</td>\n",
       "      <td>61</td>\n",
       "      <td>450</td>\n",
       "      <td>510</td>\n",
       "      <td>61</td>\n",
       "      <td>450</td>\n",
       "      <td>510</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.3e-30</td>\n",
       "      <td>117.1</td>\n",
       "      <td>13.5</td>\n",
       "      <td>5.9e-30</td>\n",
       "      <td>115.1</td>\n",
       "      <td>13.5</td>\n",
       "      <td>2.2</td>\n",
       "      <td>1</td>\n",
       "      <td>A0A7N5P5S9</td>\n",
       "      <td>1</td>\n",
       "      <td>61</td>\n",
       "      <td>61</td>\n",
       "      <td>451</td>\n",
       "      <td>511</td>\n",
       "      <td>61</td>\n",
       "      <td>451</td>\n",
       "      <td>511</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   E-value  score  bias  E-value  score  bias  exp  N    Sequence hmm_from  \\\n",
       "0  7.1e-32  121.2  13.5  8.1e-32  121.0  13.5  1.0  1  A0A8C7BFT3        1   \n",
       "1  1.2e-30  117.3  13.5  5.9e-30  115.1  13.5  2.3  1  A0A2U3ZCX9        1   \n",
       "2  1.2e-30  117.2  13.5  5.9e-30  115.0  13.5  2.3  1  A0A2U3VQT6        1   \n",
       "3  1.3e-30  117.2  13.5  5.9e-30  115.1  13.5  2.2  1  A0A2Y9JGP3        1   \n",
       "4  1.3e-30  117.1  13.5  5.9e-30  115.1  13.5  2.2  1  A0A7N5P5S9        1   \n",
       "\n",
       "  hmm_to hmm_length ali_from ali_to ali_length env_from env_to env_length  \n",
       "0     61         61        9     69         61        9     69         61  \n",
       "1     61         61      451    511         61      451    511         61  \n",
       "2     61         61      454    514         61      454    514         61  \n",
       "3     61         61      450    510         61      450    510         61  \n",
       "4     61         61      451    511         61      451    511         61  "
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Preprocess hmmsearch results - ClustalOmega\n",
    "hmmsearch_dir_clustal = '{}/results/hmms/hmmsearch/clustal'.format(directory)\n",
    "hmmsearch_dir_clustal_prepr_output = '{}/results/hmms/hmmsearch/clustal/hmmsearch_df'.format(directory)\n",
    "\n",
    "clustal_results_list = []\n",
    "\n",
    "for filename in os.listdir(hmmsearch_dir_clustal):\n",
    "    if filename.endswith('.txt'):\n",
    "        file_path = os.path.join(hmmsearch_dir_clustal, filename)\n",
    "        parts = filename.split('_')\n",
    "        protein_id = '{}_{}'.format(parts[2], parts[3].split(\".\")[0])\n",
    "        try:\n",
    "            clustal_stat = process_hmmsearch_file(file_path) \n",
    "            clustal_reg = extract_table_from_output(file_path)\n",
    "            clustal_result = pd.merge(clustal_stat, clustal_reg, left_on='Sequence', right_on='id', how='inner')\n",
    "            clustal_result = clustal_result.drop(columns=['Description', 'id'])\n",
    "            \n",
    "            # Save individual DataFrames to a list\n",
    "            clustal_results_list.append(clustal_result)\n",
    "        \n",
    "            # Save the DataFrame to a CSV file\n",
    "            output_file = os.path.join(hmmsearch_dir_clustal_prepr_output, 'hmmsearch_clustal_df_{}.csv'.format(protein_id))\n",
    "            clustal_result.to_csv(output_file, index=False)\n",
    "            \n",
    "        except ValueError as e:\n",
    "            print('Error processing file {}: {}'.format(filename, e))\n",
    "            \n",
    "# Merge all DataFrames into one\n",
    "hmmsearch_results_clustal = pd.concat(clustal_results_list, ignore_index=True)\n",
    "\n",
    "# Save the merged DataFrame to a CSV file\n",
    "merged_output_file = os.path.join('{}/results/hmms/hmmsearch/'.format(directory), 'hmmsearch_results_clustal.csv')\n",
    "hmmsearch_results_clustal.to_csv(merged_output_file, index=False)\n",
    "print('The number of retrieved hmmsearch results for ClustalOmega: {}'.format(len(hmmsearch_results_clustal)))\n",
    "hmmsearch_results_clustal.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In terms of region we are interested in the 3 fields:\n",
    "- `hmm_from`-`hmm_to`: the endpoints of the hmm profile. In our case the beginning is usually 1, since we parse not the initial alignment as an input, but the separate disordered regions where the position mostly starts from 1 (in case there are no gaps in the subject sequences).\n",
    "- `ali_from`-`ali_to`: the endpoints of the target sequence. They are obtained by matching HMM to the sequence, thus obtaining the positions.\n",
    "- `env_from`-`env_to`: the envelope of the domain's location. It's usually a bit wider than the alignment."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Pfam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-25T07:30:38.221338Z",
     "start_time": "2023-08-25T07:30:38.183994Z"
    }
   },
   "outputs": [],
   "source": [
    "# # Copy the files with the hmmsearch statistics to the remote computer\n",
    "# !scp {directory}/results/hmms/hmmsearch/hmmsearch_results_blast.csv {name}@{server}:~/stats/hmmsearch_results_blast.csv\n",
    "# !scp {directory}/results/hmms/hmmsearch/hmmsearch_results_clustal.csv {name}@{server}:~/stats/hmmsearch_results_clustal.csv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The file `filtered.tsv.gz` contains only the domains from Interpro. We check the overlap of the found results from `hmmsearch`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Check the overlaps with Interpro domains (BLAST)\n",
    "# !ssh {name}@{server} '/home/alina/protein2ipr.py /home/alina/stats/blast /home/alina/filtered.tsv.gz protein2ipr_blast.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # # Copy the files to the local folder\n",
    "# !scp {name}@{server}:~/protein2ipr_blast.tsv {directory}/results/pfam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Check the overlaps with Interpro domains (ClustalOmega)\n",
    "# !ssh {name}@{server} '/home/alina/protein2ipr.py /home/alina/stats/clustal /home/alina/filtered.tsv.gz protein2ipr_clustal.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-25T07:40:00.248249Z",
     "start_time": "2023-08-25T07:40:00.231919Z"
    }
   },
   "outputs": [],
   "source": [
    "# # Copy the files to the local folder\n",
    "# !scp {name}@{server}:~/protein2ipr_clustal.tsv {directory}/results/pfam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Results preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of retrieved Pfam instances for BLAST: 927771\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>pfam_id</th>\n",
       "      <th>ipr_id</th>\n",
       "      <th>start_pfam</th>\n",
       "      <th>end_pfam</th>\n",
       "      <th>length_pfam</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF02775</td>\n",
       "      <td>IPR011766</td>\n",
       "      <td>499</td>\n",
       "      <td>646</td>\n",
       "      <td>148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF00205</td>\n",
       "      <td>IPR012000</td>\n",
       "      <td>289</td>\n",
       "      <td>434</td>\n",
       "      <td>146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF02776</td>\n",
       "      <td>IPR012001</td>\n",
       "      <td>91</td>\n",
       "      <td>205</td>\n",
       "      <td>115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A010Q7P7</td>\n",
       "      <td>PF00018</td>\n",
       "      <td>IPR001452</td>\n",
       "      <td>435</td>\n",
       "      <td>481</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A010Q7P7</td>\n",
       "      <td>PF03114</td>\n",
       "      <td>IPR004148</td>\n",
       "      <td>113</td>\n",
       "      <td>226</td>\n",
       "      <td>114</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   uniprot_id  pfam_id     ipr_id  start_pfam  end_pfam  length_pfam\n",
       "0  A0A010Q304  PF02775  IPR011766         499       646          148\n",
       "1  A0A010Q304  PF00205  IPR012000         289       434          146\n",
       "2  A0A010Q304  PF02776  IPR012001          91       205          115\n",
       "3  A0A010Q7P7  PF00018  IPR001452         435       481           47\n",
       "4  A0A010Q7P7  PF03114  IPR004148         113       226          114"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter only entries with Pfam ID and intercepting regions with the curated_disprot instances - BLAST\n",
    "filename_blast = '{}/results/pfam/protein2ipr_blast.tsv'.format(directory)\n",
    "pfam_blast = pfam_processing(filename_blast)\n",
    "\n",
    "print('The number of retrieved Pfam instances for BLAST: {}'.format(len(pfam_blast)))\n",
    "pfam_blast.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of retrieved Pfam instances for ClustalOmega: 1073417\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>pfam_id</th>\n",
       "      <th>ipr_id</th>\n",
       "      <th>start_pfam</th>\n",
       "      <th>end_pfam</th>\n",
       "      <th>length_pfam</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF02775</td>\n",
       "      <td>IPR011766</td>\n",
       "      <td>499</td>\n",
       "      <td>646</td>\n",
       "      <td>148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF00205</td>\n",
       "      <td>IPR012000</td>\n",
       "      <td>289</td>\n",
       "      <td>434</td>\n",
       "      <td>146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF02776</td>\n",
       "      <td>IPR012001</td>\n",
       "      <td>91</td>\n",
       "      <td>205</td>\n",
       "      <td>115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A010Q7P7</td>\n",
       "      <td>PF00018</td>\n",
       "      <td>IPR001452</td>\n",
       "      <td>435</td>\n",
       "      <td>481</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A010Q7P7</td>\n",
       "      <td>PF03114</td>\n",
       "      <td>IPR004148</td>\n",
       "      <td>113</td>\n",
       "      <td>226</td>\n",
       "      <td>114</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   uniprot_id  pfam_id     ipr_id  start_pfam  end_pfam  length_pfam\n",
       "0  A0A010Q304  PF02775  IPR011766         499       646          148\n",
       "1  A0A010Q304  PF00205  IPR012000         289       434          146\n",
       "2  A0A010Q304  PF02776  IPR012001          91       205          115\n",
       "3  A0A010Q7P7  PF00018  IPR001452         435       481           47\n",
       "4  A0A010Q7P7  PF03114  IPR004148         113       226          114"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter only entries with Pfam ID and intercepting regions with the curated_disprot instances - ClustalOmega\n",
    "filename_clustal = '{}/results/pfam/protein2ipr_clustal.tsv'.format(directory)\n",
    "pfam_clustal = pfam_processing(filename_clustal)\n",
    "\n",
    "print('The number of retrieved Pfam instances for ClustalOmega: {}'.format(len(pfam_clustal)))\n",
    "pfam_clustal.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of retrieved Pfam instances for BLAST: 949687\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>pfam_id</th>\n",
       "      <th>ipr_id</th>\n",
       "      <th>start_pfam</th>\n",
       "      <th>end_pfam</th>\n",
       "      <th>length_pfam</th>\n",
       "      <th>hmm_from</th>\n",
       "      <th>hmm_to</th>\n",
       "      <th>hmm_length</th>\n",
       "      <th>ali_from</th>\n",
       "      <th>ali_to</th>\n",
       "      <th>ali_length</th>\n",
       "      <th>env_from</th>\n",
       "      <th>env_to</th>\n",
       "      <th>env_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF02775</td>\n",
       "      <td>IPR011766</td>\n",
       "      <td>499</td>\n",
       "      <td>646</td>\n",
       "      <td>148</td>\n",
       "      <td>1</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>650</td>\n",
       "      <td>688</td>\n",
       "      <td>39</td>\n",
       "      <td>650</td>\n",
       "      <td>689</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF02775</td>\n",
       "      <td>IPR011766</td>\n",
       "      <td>499</td>\n",
       "      <td>646</td>\n",
       "      <td>148</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>583</td>\n",
       "      <td>597</td>\n",
       "      <td>15</td>\n",
       "      <td>583</td>\n",
       "      <td>597</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF00205</td>\n",
       "      <td>IPR012000</td>\n",
       "      <td>289</td>\n",
       "      <td>434</td>\n",
       "      <td>146</td>\n",
       "      <td>1</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>650</td>\n",
       "      <td>688</td>\n",
       "      <td>39</td>\n",
       "      <td>650</td>\n",
       "      <td>689</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF00205</td>\n",
       "      <td>IPR012000</td>\n",
       "      <td>289</td>\n",
       "      <td>434</td>\n",
       "      <td>146</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>583</td>\n",
       "      <td>597</td>\n",
       "      <td>15</td>\n",
       "      <td>583</td>\n",
       "      <td>597</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF02776</td>\n",
       "      <td>IPR012001</td>\n",
       "      <td>91</td>\n",
       "      <td>205</td>\n",
       "      <td>115</td>\n",
       "      <td>1</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>650</td>\n",
       "      <td>688</td>\n",
       "      <td>39</td>\n",
       "      <td>650</td>\n",
       "      <td>689</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   uniprot_id  pfam_id     ipr_id  start_pfam  end_pfam  length_pfam hmm_from  \\\n",
       "0  A0A010Q304  PF02775  IPR011766         499       646          148        1   \n",
       "1  A0A010Q304  PF02775  IPR011766         499       646          148        1   \n",
       "2  A0A010Q304  PF00205  IPR012000         289       434          146        1   \n",
       "3  A0A010Q304  PF00205  IPR012000         289       434          146        1   \n",
       "4  A0A010Q304  PF02776  IPR012001          91       205          115        1   \n",
       "\n",
       "  hmm_to hmm_length ali_from ali_to ali_length env_from env_to env_length  \n",
       "0     39         39      650    688         39      650    689         40  \n",
       "1     15         15      583    597         15      583    597         15  \n",
       "2     39         39      650    688         39      650    689         40  \n",
       "3     15         15      583    597         15      583    597         15  \n",
       "4     39         39      650    688         39      650    689         40  "
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge hmmsearch results with Pfam - BLAST\n",
    "pfam_blast = pd.merge(pfam_blast, \n",
    "                      hmmsearch_results_blast[['Sequence', \n",
    "                                               'hmm_from', 'hmm_to', 'hmm_length',\n",
    "                                               'ali_from', 'ali_to', 'ali_length',\n",
    "                                               'env_from', 'env_to', 'env_length']], \n",
    "                      left_on='uniprot_id', right_on='Sequence', how='left')\n",
    "\n",
    "pfam_blast = pfam_blast.dropna(axis=0)\n",
    "pfam_blast = pfam_blast.drop(columns='Sequence')\n",
    "print('The number of retrieved Pfam instances for BLAST: {}'.format(len(pfam_blast)))\n",
    "pfam_blast.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of retrieved Pfam instances for ClustalOmega: 1109890\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>pfam_id</th>\n",
       "      <th>ipr_id</th>\n",
       "      <th>start_pfam</th>\n",
       "      <th>end_pfam</th>\n",
       "      <th>length_pfam</th>\n",
       "      <th>hmm_from</th>\n",
       "      <th>hmm_to</th>\n",
       "      <th>hmm_length</th>\n",
       "      <th>ali_from</th>\n",
       "      <th>ali_to</th>\n",
       "      <th>ali_length</th>\n",
       "      <th>env_from</th>\n",
       "      <th>env_to</th>\n",
       "      <th>env_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF02775</td>\n",
       "      <td>IPR011766</td>\n",
       "      <td>499</td>\n",
       "      <td>646</td>\n",
       "      <td>148</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>583</td>\n",
       "      <td>597</td>\n",
       "      <td>15</td>\n",
       "      <td>583</td>\n",
       "      <td>597</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF02775</td>\n",
       "      <td>IPR011766</td>\n",
       "      <td>499</td>\n",
       "      <td>646</td>\n",
       "      <td>148</td>\n",
       "      <td>1</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>650</td>\n",
       "      <td>688</td>\n",
       "      <td>39</td>\n",
       "      <td>650</td>\n",
       "      <td>689</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF00205</td>\n",
       "      <td>IPR012000</td>\n",
       "      <td>289</td>\n",
       "      <td>434</td>\n",
       "      <td>146</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>583</td>\n",
       "      <td>597</td>\n",
       "      <td>15</td>\n",
       "      <td>583</td>\n",
       "      <td>597</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF00205</td>\n",
       "      <td>IPR012000</td>\n",
       "      <td>289</td>\n",
       "      <td>434</td>\n",
       "      <td>146</td>\n",
       "      <td>1</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>650</td>\n",
       "      <td>688</td>\n",
       "      <td>39</td>\n",
       "      <td>650</td>\n",
       "      <td>689</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF02776</td>\n",
       "      <td>IPR012001</td>\n",
       "      <td>91</td>\n",
       "      <td>205</td>\n",
       "      <td>115</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>583</td>\n",
       "      <td>597</td>\n",
       "      <td>15</td>\n",
       "      <td>583</td>\n",
       "      <td>597</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   uniprot_id  pfam_id     ipr_id  start_pfam  end_pfam  length_pfam hmm_from  \\\n",
       "0  A0A010Q304  PF02775  IPR011766         499       646          148        1   \n",
       "1  A0A010Q304  PF02775  IPR011766         499       646          148        1   \n",
       "2  A0A010Q304  PF00205  IPR012000         289       434          146        1   \n",
       "3  A0A010Q304  PF00205  IPR012000         289       434          146        1   \n",
       "4  A0A010Q304  PF02776  IPR012001          91       205          115        1   \n",
       "\n",
       "  hmm_to hmm_length ali_from ali_to ali_length env_from env_to env_length  \n",
       "0     15         15      583    597         15      583    597         15  \n",
       "1     39         39      650    688         39      650    689         40  \n",
       "2     15         15      583    597         15      583    597         15  \n",
       "3     39         39      650    688         39      650    689         40  \n",
       "4     15         15      583    597         15      583    597         15  "
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge hmmsearch results with Pfam - ClustalOmega\n",
    "pfam_clustal = pd.merge(pfam_clustal, \n",
    "                        hmmsearch_results_clustal[['Sequence', \n",
    "                                                   'hmm_from', 'hmm_to', 'hmm_length',\n",
    "                                                   'ali_from', 'ali_to', 'ali_length',\n",
    "                                                   'env_from', 'env_to', 'env_length']], \n",
    "                        left_on='uniprot_id', right_on='Sequence', how='left')\n",
    "\n",
    "pfam_clustal = pfam_clustal.dropna(axis=0)\n",
    "pfam_clustal = pfam_clustal.drop(columns='Sequence')\n",
    "print('The number of retrieved Pfam instances for ClustalOmega: {}'.format(len(pfam_clustal)))\n",
    "pfam_clustal.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to calculate overlaps between DisProt-HMM and Pfam-HMM\n",
    "def pfam_hmm_overlap(row_pfam):\n",
    "    # Calculate the overlap of hmm (ali in this case) and the Pfam\n",
    "    overlap_len = min(row_pfam['end_pfam'], row_pfam['ali_to']) - max(row_pfam['start_pfam'], \n",
    "                                                                      row_pfam['ali_from']) + 1 \n",
    "    overlap_pfam = overlap_len / row_pfam['length_pfam'] * 100 # % of length overlap with Pfam\n",
    "    overlap_ali = overlap_len / row_pfam['ali_length'] * 100 # % of length overlap with HMM (ali)\n",
    "    max_length = row_pfam['length_pfam'] + row_pfam['ali_length'] - overlap_len # the whole region covered by hmm and pfam\n",
    "    non_overlap_len = max_length - overlap_len + 1 # non covered region\n",
    "\n",
    "    if overlap_len > 0:\n",
    "        overlap_perc = (overlap_len / max_length) * 100 # which region\n",
    "        overlap_sym = 2 * overlap_len / (row_pfam['ali_length'] + row_pfam['length_pfam'])\n",
    "    else:\n",
    "        overlap_perc = 0\n",
    "        overlap_sym = 0\n",
    "\n",
    "    return overlap_len, overlap_pfam, overlap_ali, non_overlap_len, overlap_perc, overlap_sym"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>pfam_id</th>\n",
       "      <th>ipr_id</th>\n",
       "      <th>start_pfam</th>\n",
       "      <th>end_pfam</th>\n",
       "      <th>length_pfam</th>\n",
       "      <th>hmm_from</th>\n",
       "      <th>hmm_to</th>\n",
       "      <th>hmm_length</th>\n",
       "      <th>ali_from</th>\n",
       "      <th>...</th>\n",
       "      <th>ali_length</th>\n",
       "      <th>env_from</th>\n",
       "      <th>env_to</th>\n",
       "      <th>env_length</th>\n",
       "      <th>overl_len</th>\n",
       "      <th>overl_pfam</th>\n",
       "      <th>overl_ali</th>\n",
       "      <th>non_overl_len</th>\n",
       "      <th>overl_perc</th>\n",
       "      <th>overlap_sym</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF02775</td>\n",
       "      <td>IPR011766</td>\n",
       "      <td>499</td>\n",
       "      <td>646</td>\n",
       "      <td>148</td>\n",
       "      <td>1</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>650</td>\n",
       "      <td>...</td>\n",
       "      <td>39</td>\n",
       "      <td>650</td>\n",
       "      <td>689</td>\n",
       "      <td>40</td>\n",
       "      <td>-3</td>\n",
       "      <td>-2.03</td>\n",
       "      <td>-7.69</td>\n",
       "      <td>194</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF02775</td>\n",
       "      <td>IPR011766</td>\n",
       "      <td>499</td>\n",
       "      <td>646</td>\n",
       "      <td>148</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>583</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>583</td>\n",
       "      <td>597</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>10.14</td>\n",
       "      <td>100.00</td>\n",
       "      <td>134</td>\n",
       "      <td>10.14</td>\n",
       "      <td>0.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF00205</td>\n",
       "      <td>IPR012000</td>\n",
       "      <td>289</td>\n",
       "      <td>434</td>\n",
       "      <td>146</td>\n",
       "      <td>1</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>650</td>\n",
       "      <td>...</td>\n",
       "      <td>39</td>\n",
       "      <td>650</td>\n",
       "      <td>689</td>\n",
       "      <td>40</td>\n",
       "      <td>-215</td>\n",
       "      <td>-147.26</td>\n",
       "      <td>-551.28</td>\n",
       "      <td>616</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF00205</td>\n",
       "      <td>IPR012000</td>\n",
       "      <td>289</td>\n",
       "      <td>434</td>\n",
       "      <td>146</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>583</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>583</td>\n",
       "      <td>597</td>\n",
       "      <td>15</td>\n",
       "      <td>-148</td>\n",
       "      <td>-101.37</td>\n",
       "      <td>-986.67</td>\n",
       "      <td>458</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF02776</td>\n",
       "      <td>IPR012001</td>\n",
       "      <td>91</td>\n",
       "      <td>205</td>\n",
       "      <td>115</td>\n",
       "      <td>1</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>650</td>\n",
       "      <td>...</td>\n",
       "      <td>39</td>\n",
       "      <td>650</td>\n",
       "      <td>689</td>\n",
       "      <td>40</td>\n",
       "      <td>-444</td>\n",
       "      <td>-386.09</td>\n",
       "      <td>-1138.46</td>\n",
       "      <td>1043</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   uniprot_id  pfam_id     ipr_id  start_pfam  end_pfam  length_pfam hmm_from  \\\n",
       "0  A0A010Q304  PF02775  IPR011766         499       646          148        1   \n",
       "1  A0A010Q304  PF02775  IPR011766         499       646          148        1   \n",
       "2  A0A010Q304  PF00205  IPR012000         289       434          146        1   \n",
       "3  A0A010Q304  PF00205  IPR012000         289       434          146        1   \n",
       "4  A0A010Q304  PF02776  IPR012001          91       205          115        1   \n",
       "\n",
       "  hmm_to hmm_length ali_from  ... ali_length env_from env_to env_length  \\\n",
       "0     39         39      650  ...         39      650    689         40   \n",
       "1     15         15      583  ...         15      583    597         15   \n",
       "2     39         39      650  ...         39      650    689         40   \n",
       "3     15         15      583  ...         15      583    597         15   \n",
       "4     39         39      650  ...         39      650    689         40   \n",
       "\n",
       "  overl_len  overl_pfam  overl_ali  non_overl_len  overl_perc  overlap_sym  \n",
       "0        -3       -2.03      -7.69            194        0.00         0.00  \n",
       "1        15       10.14     100.00            134       10.14         0.18  \n",
       "2      -215     -147.26    -551.28            616        0.00         0.00  \n",
       "3      -148     -101.37    -986.67            458        0.00         0.00  \n",
       "4      -444     -386.09   -1138.46           1043        0.00         0.00  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add the overlaps to the dataframe - BLAST\n",
    "overlap_pfam_hmm = []\n",
    "\n",
    "for index_pfam, row_pfam in pfam_blast.iterrows():\n",
    "    overl_len, overl_pfam, overl_ali, non_overl_len, overl_perc, overlap_sym = pfam_hmm_overlap(row_pfam)\n",
    "    overl_pfam = round(overl_pfam, 2)\n",
    "    overl_ali = round(overl_ali, 2)\n",
    "    overl_perc = round(overl_perc, 2)\n",
    "    overlap_sym = round(overlap_sym, 2)\n",
    "    overlap_pfam_hmm.append((overl_len, overl_pfam, overl_ali, non_overl_len, overl_perc, overlap_sym))\n",
    "\n",
    "# Extract overlap_pfam and overlap_hmm from the list of tuples\n",
    "pfam_blast['overl_len'], pfam_blast['overl_pfam'], pfam_blast['overl_ali'], pfam_blast['non_overl_len'], pfam_blast['overl_perc'], pfam_blast['overlap_sym'] = zip(*overlap_pfam_hmm)\n",
    "pfam_blast.to_csv(f'results/pfam/pfam_overlap/pfam_blast.csv', index=False)\n",
    "pfam_blast.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>pfam_id</th>\n",
       "      <th>ipr_id</th>\n",
       "      <th>start_pfam</th>\n",
       "      <th>end_pfam</th>\n",
       "      <th>length_pfam</th>\n",
       "      <th>hmm_from</th>\n",
       "      <th>hmm_to</th>\n",
       "      <th>hmm_length</th>\n",
       "      <th>ali_from</th>\n",
       "      <th>...</th>\n",
       "      <th>ali_length</th>\n",
       "      <th>env_from</th>\n",
       "      <th>env_to</th>\n",
       "      <th>env_length</th>\n",
       "      <th>overl_len</th>\n",
       "      <th>overl_pfam</th>\n",
       "      <th>overl_ali</th>\n",
       "      <th>non_overl_len</th>\n",
       "      <th>overl_perc</th>\n",
       "      <th>overlap_sym</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF02775</td>\n",
       "      <td>IPR011766</td>\n",
       "      <td>499</td>\n",
       "      <td>646</td>\n",
       "      <td>148</td>\n",
       "      <td>1</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>650</td>\n",
       "      <td>...</td>\n",
       "      <td>39</td>\n",
       "      <td>650</td>\n",
       "      <td>689</td>\n",
       "      <td>40</td>\n",
       "      <td>-3</td>\n",
       "      <td>-2.03</td>\n",
       "      <td>-7.69</td>\n",
       "      <td>194</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF02775</td>\n",
       "      <td>IPR011766</td>\n",
       "      <td>499</td>\n",
       "      <td>646</td>\n",
       "      <td>148</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>583</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>583</td>\n",
       "      <td>597</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>10.14</td>\n",
       "      <td>100.00</td>\n",
       "      <td>134</td>\n",
       "      <td>10.14</td>\n",
       "      <td>0.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF00205</td>\n",
       "      <td>IPR012000</td>\n",
       "      <td>289</td>\n",
       "      <td>434</td>\n",
       "      <td>146</td>\n",
       "      <td>1</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>650</td>\n",
       "      <td>...</td>\n",
       "      <td>39</td>\n",
       "      <td>650</td>\n",
       "      <td>689</td>\n",
       "      <td>40</td>\n",
       "      <td>-215</td>\n",
       "      <td>-147.26</td>\n",
       "      <td>-551.28</td>\n",
       "      <td>616</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF00205</td>\n",
       "      <td>IPR012000</td>\n",
       "      <td>289</td>\n",
       "      <td>434</td>\n",
       "      <td>146</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>583</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>583</td>\n",
       "      <td>597</td>\n",
       "      <td>15</td>\n",
       "      <td>-148</td>\n",
       "      <td>-101.37</td>\n",
       "      <td>-986.67</td>\n",
       "      <td>458</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF02776</td>\n",
       "      <td>IPR012001</td>\n",
       "      <td>91</td>\n",
       "      <td>205</td>\n",
       "      <td>115</td>\n",
       "      <td>1</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>650</td>\n",
       "      <td>...</td>\n",
       "      <td>39</td>\n",
       "      <td>650</td>\n",
       "      <td>689</td>\n",
       "      <td>40</td>\n",
       "      <td>-444</td>\n",
       "      <td>-386.09</td>\n",
       "      <td>-1138.46</td>\n",
       "      <td>1043</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>949682</th>\n",
       "      <td>X6N7N7</td>\n",
       "      <td>PF00018</td>\n",
       "      <td>IPR001452</td>\n",
       "      <td>132</td>\n",
       "      <td>177</td>\n",
       "      <td>46</td>\n",
       "      <td>10</td>\n",
       "      <td>66</td>\n",
       "      <td>57</td>\n",
       "      <td>55</td>\n",
       "      <td>...</td>\n",
       "      <td>58</td>\n",
       "      <td>52</td>\n",
       "      <td>116</td>\n",
       "      <td>65</td>\n",
       "      <td>-19</td>\n",
       "      <td>-41.30</td>\n",
       "      <td>-32.76</td>\n",
       "      <td>143</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>949683</th>\n",
       "      <td>X6N7N7</td>\n",
       "      <td>PF02209</td>\n",
       "      <td>IPR003128</td>\n",
       "      <td>314</td>\n",
       "      <td>349</td>\n",
       "      <td>36</td>\n",
       "      <td>10</td>\n",
       "      <td>66</td>\n",
       "      <td>57</td>\n",
       "      <td>55</td>\n",
       "      <td>...</td>\n",
       "      <td>58</td>\n",
       "      <td>52</td>\n",
       "      <td>116</td>\n",
       "      <td>65</td>\n",
       "      <td>-201</td>\n",
       "      <td>-558.33</td>\n",
       "      <td>-346.55</td>\n",
       "      <td>497</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>949684</th>\n",
       "      <td>X6NHX1</td>\n",
       "      <td>PF00018</td>\n",
       "      <td>IPR001452</td>\n",
       "      <td>41</td>\n",
       "      <td>85</td>\n",
       "      <td>45</td>\n",
       "      <td>6</td>\n",
       "      <td>66</td>\n",
       "      <td>61</td>\n",
       "      <td>30</td>\n",
       "      <td>...</td>\n",
       "      <td>61</td>\n",
       "      <td>25</td>\n",
       "      <td>93</td>\n",
       "      <td>69</td>\n",
       "      <td>45</td>\n",
       "      <td>100.00</td>\n",
       "      <td>73.77</td>\n",
       "      <td>17</td>\n",
       "      <td>73.77</td>\n",
       "      <td>0.85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>949685</th>\n",
       "      <td>X6NKT1</td>\n",
       "      <td>PF14604</td>\n",
       "      <td>IPR001452</td>\n",
       "      <td>10</td>\n",
       "      <td>56</td>\n",
       "      <td>47</td>\n",
       "      <td>15</td>\n",
       "      <td>69</td>\n",
       "      <td>55</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>55</td>\n",
       "      <td>4</td>\n",
       "      <td>61</td>\n",
       "      <td>58</td>\n",
       "      <td>47</td>\n",
       "      <td>100.00</td>\n",
       "      <td>85.45</td>\n",
       "      <td>9</td>\n",
       "      <td>85.45</td>\n",
       "      <td>0.92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>949686</th>\n",
       "      <td>X6PEP5</td>\n",
       "      <td>PF14604</td>\n",
       "      <td>IPR001452</td>\n",
       "      <td>9</td>\n",
       "      <td>56</td>\n",
       "      <td>48</td>\n",
       "      <td>16</td>\n",
       "      <td>67</td>\n",
       "      <td>52</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>52</td>\n",
       "      <td>4</td>\n",
       "      <td>61</td>\n",
       "      <td>58</td>\n",
       "      <td>48</td>\n",
       "      <td>100.00</td>\n",
       "      <td>92.31</td>\n",
       "      <td>5</td>\n",
       "      <td>92.31</td>\n",
       "      <td>0.96</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>947815 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        uniprot_id  pfam_id     ipr_id  start_pfam  end_pfam  length_pfam  \\\n",
       "0       A0A010Q304  PF02775  IPR011766         499       646          148   \n",
       "1       A0A010Q304  PF02775  IPR011766         499       646          148   \n",
       "2       A0A010Q304  PF00205  IPR012000         289       434          146   \n",
       "3       A0A010Q304  PF00205  IPR012000         289       434          146   \n",
       "4       A0A010Q304  PF02776  IPR012001          91       205          115   \n",
       "...            ...      ...        ...         ...       ...          ...   \n",
       "949682      X6N7N7  PF00018  IPR001452         132       177           46   \n",
       "949683      X6N7N7  PF02209  IPR003128         314       349           36   \n",
       "949684      X6NHX1  PF00018  IPR001452          41        85           45   \n",
       "949685      X6NKT1  PF14604  IPR001452          10        56           47   \n",
       "949686      X6PEP5  PF14604  IPR001452           9        56           48   \n",
       "\n",
       "       hmm_from hmm_to hmm_length ali_from  ... ali_length env_from env_to  \\\n",
       "0             1     39         39      650  ...         39      650    689   \n",
       "1             1     15         15      583  ...         15      583    597   \n",
       "2             1     39         39      650  ...         39      650    689   \n",
       "3             1     15         15      583  ...         15      583    597   \n",
       "4             1     39         39      650  ...         39      650    689   \n",
       "...         ...    ...        ...      ...  ...        ...      ...    ...   \n",
       "949682       10     66         57       55  ...         58       52    116   \n",
       "949683       10     66         57       55  ...         58       52    116   \n",
       "949684        6     66         61       30  ...         61       25     93   \n",
       "949685       15     69         55        6  ...         55        4     61   \n",
       "949686       16     67         52        7  ...         52        4     61   \n",
       "\n",
       "       env_length overl_len  overl_pfam  overl_ali  non_overl_len  overl_perc  \\\n",
       "0              40        -3       -2.03      -7.69            194        0.00   \n",
       "1              15        15       10.14     100.00            134       10.14   \n",
       "2              40      -215     -147.26    -551.28            616        0.00   \n",
       "3              15      -148     -101.37    -986.67            458        0.00   \n",
       "4              40      -444     -386.09   -1138.46           1043        0.00   \n",
       "...           ...       ...         ...        ...            ...         ...   \n",
       "949682         65       -19      -41.30     -32.76            143        0.00   \n",
       "949683         65      -201     -558.33    -346.55            497        0.00   \n",
       "949684         69        45      100.00      73.77             17       73.77   \n",
       "949685         58        47      100.00      85.45              9       85.45   \n",
       "949686         58        48      100.00      92.31              5       92.31   \n",
       "\n",
       "        overlap_sym  \n",
       "0              0.00  \n",
       "1              0.18  \n",
       "2              0.00  \n",
       "3              0.00  \n",
       "4              0.00  \n",
       "...             ...  \n",
       "949682         0.00  \n",
       "949683         0.00  \n",
       "949684         0.85  \n",
       "949685         0.92  \n",
       "949686         0.96  \n",
       "\n",
       "[947815 rows x 21 columns]"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pfam_blast[pfam_blast['overl_len'] != 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>pfam_id</th>\n",
       "      <th>ipr_id</th>\n",
       "      <th>start_pfam</th>\n",
       "      <th>end_pfam</th>\n",
       "      <th>length_pfam</th>\n",
       "      <th>hmm_from</th>\n",
       "      <th>hmm_to</th>\n",
       "      <th>hmm_length</th>\n",
       "      <th>ali_from</th>\n",
       "      <th>...</th>\n",
       "      <th>ali_length</th>\n",
       "      <th>env_from</th>\n",
       "      <th>env_to</th>\n",
       "      <th>env_length</th>\n",
       "      <th>overl_len</th>\n",
       "      <th>overl_pfam</th>\n",
       "      <th>overl_ali</th>\n",
       "      <th>non_overl_len</th>\n",
       "      <th>overl_perc</th>\n",
       "      <th>overlap_sym</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF02775</td>\n",
       "      <td>IPR011766</td>\n",
       "      <td>499</td>\n",
       "      <td>646</td>\n",
       "      <td>148</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>583</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>583</td>\n",
       "      <td>597</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>10.14</td>\n",
       "      <td>100.00</td>\n",
       "      <td>134</td>\n",
       "      <td>10.14</td>\n",
       "      <td>0.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF02775</td>\n",
       "      <td>IPR011766</td>\n",
       "      <td>499</td>\n",
       "      <td>646</td>\n",
       "      <td>148</td>\n",
       "      <td>1</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>650</td>\n",
       "      <td>...</td>\n",
       "      <td>39</td>\n",
       "      <td>650</td>\n",
       "      <td>689</td>\n",
       "      <td>40</td>\n",
       "      <td>-3</td>\n",
       "      <td>-2.03</td>\n",
       "      <td>-7.69</td>\n",
       "      <td>194</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF00205</td>\n",
       "      <td>IPR012000</td>\n",
       "      <td>289</td>\n",
       "      <td>434</td>\n",
       "      <td>146</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>583</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>583</td>\n",
       "      <td>597</td>\n",
       "      <td>15</td>\n",
       "      <td>-148</td>\n",
       "      <td>-101.37</td>\n",
       "      <td>-986.67</td>\n",
       "      <td>458</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF00205</td>\n",
       "      <td>IPR012000</td>\n",
       "      <td>289</td>\n",
       "      <td>434</td>\n",
       "      <td>146</td>\n",
       "      <td>1</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>650</td>\n",
       "      <td>...</td>\n",
       "      <td>39</td>\n",
       "      <td>650</td>\n",
       "      <td>689</td>\n",
       "      <td>40</td>\n",
       "      <td>-215</td>\n",
       "      <td>-147.26</td>\n",
       "      <td>-551.28</td>\n",
       "      <td>616</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF02776</td>\n",
       "      <td>IPR012001</td>\n",
       "      <td>91</td>\n",
       "      <td>205</td>\n",
       "      <td>115</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>583</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>583</td>\n",
       "      <td>597</td>\n",
       "      <td>15</td>\n",
       "      <td>-377</td>\n",
       "      <td>-327.83</td>\n",
       "      <td>-2513.33</td>\n",
       "      <td>885</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   uniprot_id  pfam_id     ipr_id  start_pfam  end_pfam  length_pfam hmm_from  \\\n",
       "0  A0A010Q304  PF02775  IPR011766         499       646          148        1   \n",
       "1  A0A010Q304  PF02775  IPR011766         499       646          148        1   \n",
       "2  A0A010Q304  PF00205  IPR012000         289       434          146        1   \n",
       "3  A0A010Q304  PF00205  IPR012000         289       434          146        1   \n",
       "4  A0A010Q304  PF02776  IPR012001          91       205          115        1   \n",
       "\n",
       "  hmm_to hmm_length ali_from  ... ali_length env_from env_to env_length  \\\n",
       "0     15         15      583  ...         15      583    597         15   \n",
       "1     39         39      650  ...         39      650    689         40   \n",
       "2     15         15      583  ...         15      583    597         15   \n",
       "3     39         39      650  ...         39      650    689         40   \n",
       "4     15         15      583  ...         15      583    597         15   \n",
       "\n",
       "  overl_len  overl_pfam  overl_ali  non_overl_len  overl_perc  overlap_sym  \n",
       "0        15       10.14     100.00            134       10.14         0.18  \n",
       "1        -3       -2.03      -7.69            194        0.00         0.00  \n",
       "2      -148     -101.37    -986.67            458        0.00         0.00  \n",
       "3      -215     -147.26    -551.28            616        0.00         0.00  \n",
       "4      -377     -327.83   -2513.33            885        0.00         0.00  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add the overlaps to the dataframe - ClustalOmega\n",
    "overlap_pfam_hmm = []\n",
    "\n",
    "for index_pfam, row_pfam in pfam_clustal.iterrows():\n",
    "    overl_len, overl_pfam, overl_ali, non_overl_len, overl_perc, overlap_sym = pfam_hmm_overlap(row_pfam)\n",
    "    overl_pfam = round(overl_pfam, 2)\n",
    "    overl_ali = round(overl_ali, 2)\n",
    "    overl_perc = round(overl_perc, 2)\n",
    "    overlap_sym = round(overlap_sym, 2)\n",
    "    overlap_pfam_hmm.append((overl_len, overl_pfam, overl_ali, non_overl_len, overl_perc, overlap_sym))\n",
    "\n",
    "# Extract overlap_pfam and overlap_hmm from the list of tuples\n",
    "pfam_clustal['overl_len'], pfam_clustal['overl_pfam'], pfam_clustal['overl_ali'], pfam_clustal['non_overl_len'], pfam_clustal['overl_perc'], pfam_clustal['overlap_sym'] = zip(*overlap_pfam_hmm)\n",
    "pfam_clustal.to_csv(f'results/pfam/pfam_overlap/pfam_clustal.csv', index=False)\n",
    "pfam_clustal.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>pfam_id</th>\n",
       "      <th>ipr_id</th>\n",
       "      <th>start_pfam</th>\n",
       "      <th>end_pfam</th>\n",
       "      <th>length_pfam</th>\n",
       "      <th>hmm_from</th>\n",
       "      <th>hmm_to</th>\n",
       "      <th>hmm_length</th>\n",
       "      <th>ali_from</th>\n",
       "      <th>...</th>\n",
       "      <th>ali_length</th>\n",
       "      <th>env_from</th>\n",
       "      <th>env_to</th>\n",
       "      <th>env_length</th>\n",
       "      <th>overl_len</th>\n",
       "      <th>overl_pfam</th>\n",
       "      <th>overl_ali</th>\n",
       "      <th>non_overl_len</th>\n",
       "      <th>overl_perc</th>\n",
       "      <th>overlap_sym</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF02775</td>\n",
       "      <td>IPR011766</td>\n",
       "      <td>499</td>\n",
       "      <td>646</td>\n",
       "      <td>148</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>583</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>583</td>\n",
       "      <td>597</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>10.14</td>\n",
       "      <td>100.00</td>\n",
       "      <td>134</td>\n",
       "      <td>10.14</td>\n",
       "      <td>0.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF02775</td>\n",
       "      <td>IPR011766</td>\n",
       "      <td>499</td>\n",
       "      <td>646</td>\n",
       "      <td>148</td>\n",
       "      <td>1</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>650</td>\n",
       "      <td>...</td>\n",
       "      <td>39</td>\n",
       "      <td>650</td>\n",
       "      <td>689</td>\n",
       "      <td>40</td>\n",
       "      <td>-3</td>\n",
       "      <td>-2.03</td>\n",
       "      <td>-7.69</td>\n",
       "      <td>194</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF00205</td>\n",
       "      <td>IPR012000</td>\n",
       "      <td>289</td>\n",
       "      <td>434</td>\n",
       "      <td>146</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>583</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>583</td>\n",
       "      <td>597</td>\n",
       "      <td>15</td>\n",
       "      <td>-148</td>\n",
       "      <td>-101.37</td>\n",
       "      <td>-986.67</td>\n",
       "      <td>458</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF00205</td>\n",
       "      <td>IPR012000</td>\n",
       "      <td>289</td>\n",
       "      <td>434</td>\n",
       "      <td>146</td>\n",
       "      <td>1</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>650</td>\n",
       "      <td>...</td>\n",
       "      <td>39</td>\n",
       "      <td>650</td>\n",
       "      <td>689</td>\n",
       "      <td>40</td>\n",
       "      <td>-215</td>\n",
       "      <td>-147.26</td>\n",
       "      <td>-551.28</td>\n",
       "      <td>616</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF02776</td>\n",
       "      <td>IPR012001</td>\n",
       "      <td>91</td>\n",
       "      <td>205</td>\n",
       "      <td>115</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>583</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>583</td>\n",
       "      <td>597</td>\n",
       "      <td>15</td>\n",
       "      <td>-377</td>\n",
       "      <td>-327.83</td>\n",
       "      <td>-2513.33</td>\n",
       "      <td>885</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1109885</th>\n",
       "      <td>X7YG90</td>\n",
       "      <td>PF02776</td>\n",
       "      <td>IPR012001</td>\n",
       "      <td>35</td>\n",
       "      <td>150</td>\n",
       "      <td>116</td>\n",
       "      <td>2</td>\n",
       "      <td>15</td>\n",
       "      <td>14</td>\n",
       "      <td>507</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>507</td>\n",
       "      <td>520</td>\n",
       "      <td>14</td>\n",
       "      <td>-356</td>\n",
       "      <td>-306.90</td>\n",
       "      <td>-2542.86</td>\n",
       "      <td>843</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1109886</th>\n",
       "      <td>X7ZSI2</td>\n",
       "      <td>PF02775</td>\n",
       "      <td>IPR011766</td>\n",
       "      <td>422</td>\n",
       "      <td>575</td>\n",
       "      <td>154</td>\n",
       "      <td>2</td>\n",
       "      <td>15</td>\n",
       "      <td>14</td>\n",
       "      <td>507</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>507</td>\n",
       "      <td>520</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>9.09</td>\n",
       "      <td>100.00</td>\n",
       "      <td>141</td>\n",
       "      <td>9.09</td>\n",
       "      <td>0.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1109887</th>\n",
       "      <td>X7ZSI2</td>\n",
       "      <td>PF00205</td>\n",
       "      <td>IPR012000</td>\n",
       "      <td>224</td>\n",
       "      <td>359</td>\n",
       "      <td>136</td>\n",
       "      <td>2</td>\n",
       "      <td>15</td>\n",
       "      <td>14</td>\n",
       "      <td>507</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>507</td>\n",
       "      <td>520</td>\n",
       "      <td>14</td>\n",
       "      <td>-147</td>\n",
       "      <td>-108.09</td>\n",
       "      <td>-1050.00</td>\n",
       "      <td>445</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1109888</th>\n",
       "      <td>X7ZSI2</td>\n",
       "      <td>PF02776</td>\n",
       "      <td>IPR012001</td>\n",
       "      <td>35</td>\n",
       "      <td>150</td>\n",
       "      <td>116</td>\n",
       "      <td>2</td>\n",
       "      <td>15</td>\n",
       "      <td>14</td>\n",
       "      <td>507</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>507</td>\n",
       "      <td>520</td>\n",
       "      <td>14</td>\n",
       "      <td>-356</td>\n",
       "      <td>-306.90</td>\n",
       "      <td>-2542.86</td>\n",
       "      <td>843</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1109889</th>\n",
       "      <td>X8FDQ9</td>\n",
       "      <td>PF02775</td>\n",
       "      <td>IPR011766</td>\n",
       "      <td>1</td>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>2</td>\n",
       "      <td>15</td>\n",
       "      <td>14</td>\n",
       "      <td>27</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>27</td>\n",
       "      <td>40</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>14.74</td>\n",
       "      <td>100.00</td>\n",
       "      <td>82</td>\n",
       "      <td>14.74</td>\n",
       "      <td>0.26</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1106621 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         uniprot_id  pfam_id     ipr_id  start_pfam  end_pfam  length_pfam  \\\n",
       "0        A0A010Q304  PF02775  IPR011766         499       646          148   \n",
       "1        A0A010Q304  PF02775  IPR011766         499       646          148   \n",
       "2        A0A010Q304  PF00205  IPR012000         289       434          146   \n",
       "3        A0A010Q304  PF00205  IPR012000         289       434          146   \n",
       "4        A0A010Q304  PF02776  IPR012001          91       205          115   \n",
       "...             ...      ...        ...         ...       ...          ...   \n",
       "1109885      X7YG90  PF02776  IPR012001          35       150          116   \n",
       "1109886      X7ZSI2  PF02775  IPR011766         422       575          154   \n",
       "1109887      X7ZSI2  PF00205  IPR012000         224       359          136   \n",
       "1109888      X7ZSI2  PF02776  IPR012001          35       150          116   \n",
       "1109889      X8FDQ9  PF02775  IPR011766           1        95           95   \n",
       "\n",
       "        hmm_from hmm_to hmm_length ali_from  ... ali_length env_from env_to  \\\n",
       "0              1     15         15      583  ...         15      583    597   \n",
       "1              1     39         39      650  ...         39      650    689   \n",
       "2              1     15         15      583  ...         15      583    597   \n",
       "3              1     39         39      650  ...         39      650    689   \n",
       "4              1     15         15      583  ...         15      583    597   \n",
       "...          ...    ...        ...      ...  ...        ...      ...    ...   \n",
       "1109885        2     15         14      507  ...         14      507    520   \n",
       "1109886        2     15         14      507  ...         14      507    520   \n",
       "1109887        2     15         14      507  ...         14      507    520   \n",
       "1109888        2     15         14      507  ...         14      507    520   \n",
       "1109889        2     15         14       27  ...         14       27     40   \n",
       "\n",
       "        env_length overl_len  overl_pfam  overl_ali  non_overl_len  \\\n",
       "0               15        15       10.14     100.00            134   \n",
       "1               40        -3       -2.03      -7.69            194   \n",
       "2               15      -148     -101.37    -986.67            458   \n",
       "3               40      -215     -147.26    -551.28            616   \n",
       "4               15      -377     -327.83   -2513.33            885   \n",
       "...            ...       ...         ...        ...            ...   \n",
       "1109885         14      -356     -306.90   -2542.86            843   \n",
       "1109886         14        14        9.09     100.00            141   \n",
       "1109887         14      -147     -108.09   -1050.00            445   \n",
       "1109888         14      -356     -306.90   -2542.86            843   \n",
       "1109889         14        14       14.74     100.00             82   \n",
       "\n",
       "         overl_perc  overlap_sym  \n",
       "0             10.14         0.18  \n",
       "1              0.00         0.00  \n",
       "2              0.00         0.00  \n",
       "3              0.00         0.00  \n",
       "4              0.00         0.00  \n",
       "...             ...          ...  \n",
       "1109885        0.00         0.00  \n",
       "1109886        9.09         0.17  \n",
       "1109887        0.00         0.00  \n",
       "1109888        0.00         0.00  \n",
       "1109889       14.74         0.26  \n",
       "\n",
       "[1106621 rows x 21 columns]"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pfam_clustal[pfam_clustal['overl_len'] != 0]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
