{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is the notebook related to HMM data preprocessing. To run this, please uncomment all the cells an press 'Run all' button."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set the path and input parameters\n",
    "import os\n",
    "directory = os.getcwd() # the main directory of the project\n",
    "\n",
    "# The credentials for the remote cluster\n",
    "name = 'alina'\n",
    "server = 'ecate'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-10-02T07:55:55.284130Z",
     "start_time": "2023-10-02T07:55:53.625394Z"
    }
   },
   "outputs": [],
   "source": [
    "# Importing the libraries\n",
    "from functions import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## HMM: Data preparation\n",
    "In this part we generate HMM models based on MSA and retrieve the results from hmmsearch and then from Pfam. We start with the loading of the `disordered` dataframe and analyse how many proteins and disordered regions we have."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set the maximum width of the columns\n",
    "pd.set_option('display.max_colwidth', 20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of rows with the disordered regions: 7393\n"
     ]
    }
   ],
   "source": [
    "# Load the disordered and curated_disprot dataframes\n",
    "disordered_df = pd.read_csv('{}/disordered_df.csv'.format(directory))\n",
    "curated_disprot_df = pd.read_csv('{}/curated_disprot.csv'.format(directory))\n",
    "\n",
    "print('The number of rows with the disordered regions: {}'.format(len(disordered_df)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of proteins with disordered regions: 39\n",
      "The number of disordered regions in the proteins: 53\n"
     ]
    }
   ],
   "source": [
    "# Collect the disordered regions in a dictionary\n",
    "dis_regs = set()\n",
    "\n",
    "for i, row in disordered_df.iterrows():\n",
    "    dis_id = row[0]\n",
    "    matching_row = curated_disprot_df[curated_disprot_df['acc'] == dis_id]\n",
    "    if not matching_row.empty:\n",
    "        region = matching_row['region']\n",
    "        dis_regs.update(region)\n",
    "        dis_regs_list = list(dis_regs) # convert set to a list\n",
    "        \n",
    "# Define an array of disordered regions ids\n",
    "disprot_ids = disordered_df['query_id'].unique()\n",
    "\n",
    "print('The number of proteins with disordered regions: {}'.format(len(disprot_ids)))\n",
    "print('The number of disordered regions in the proteins: {}'.format(len(dis_regs_list)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The results above show how many proteins with the disordered regions are in the provided data, as well as how many disordered regions do they contain."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-02T13:38:09.492889Z",
     "start_time": "2023-08-02T13:38:08.673895Z"
    }
   },
   "source": [
    "## 1. hmmbuild\n",
    "We build an HMM of each disordered region, using trimmed MSA as an input. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Directory /Users/alina/HMM/results/alignments/output_files/disordered/clustal already exists.\n",
      "Directory /Users/alina/HMM/results/hmms/hmmbuild/clustal already exists.\n"
     ]
    }
   ],
   "source": [
    "# Set the paths to  MSA (input) and HMM (output) files - ClustalOmega\n",
    "msa_dir_clustal = '{}/results/alignments/output_files/disordered/clustal'.format(directory)\n",
    "hmmbuild_dir_clustal = '{}/results/hmms/hmmbuild/clustal'.format(directory)\n",
    "\n",
    "# Create directories if they don't exist\n",
    "create_directory(msa_dir_clustal)\n",
    "create_directory(hmmbuild_dir_clustal)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Build HMM for ClustalOmega\n",
    "# for filename in os.listdir(msa_dir_clustal):\n",
    "#     if filename.endswith('.fasta'):\n",
    "#         input_file = os.path.join(msa_dir_clustal, filename)\n",
    "#         output_file = os.path.join(hmmbuild_dir_clustal, os.path.splitext(filename)[0] + '.hmm')\n",
    "\n",
    "#         subprocess.run(['hmmbuild', output_file, input_file])\n",
    "#         print('hmmbuild completed for {}'.format(filename))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. hmmsearch\n",
    "\n",
    "After building the models, our objective is to assess if overlaps with the profiles in the sequence database (Reference Proteome 75%) exist. We retrive the data containing the most significant sequences, with a default E-value threshold of 0.01.\n",
    "\n",
    "The same as before, we do it for BLAST and ClustalOmega separately."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Copy the HMMs from the ClustalOmega folder\n",
    "# for filename in os.listdir(hmmbuild_dir_clustal):\n",
    "#     source_path = os.path.join(hmmbuild_dir_clustal, filename)\n",
    "#     destination_path = '{}@{}:~/hmm/clustal/hmmbuild/{}'.format(name, server, filename)\n",
    "#     subprocess.run(['scp', source_path, destination_path])\n",
    "#     print('File {} copied to {}'.format(filename, destination_path))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After copying the files, we run `hmmsearch` command. \n",
    "It is done remotely by running the command `sbatch start_array_jobs.sh <ali_type>`.\n",
    "`<ali_type>` can be either `blast` or `clustal`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Construct the source and the target directories for ClustalOmega\n",
    "clustal_in_dir_rem = '/home/{}/hmm/clustal/hmmbuild'.format(name)\n",
    "clustal_out_dir_rem = '/home/{}/hmm/clustal/hmmsearch'.format(name)\n",
    "\n",
    "# # List files in the source directory on the remote machine\n",
    "# ls_command_clustal = 'ssh {}@{} \"ls {}\"'.format(name, server, clustal_in_dir_rem)\n",
    "# file_list_clustal = get_ipython().getoutput(ls_command_clustal)\n",
    "\n",
    "# Then run sbatch start_array_jobs.sh clustal on the remote computer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Copy ClustalOmega results to the local folder\n",
    "# hmmsearch_dir_clustal = '{}/results/hmms/hmmsearch/clustal'.format(directory)\n",
    "\n",
    "# for filename in file_list_clustal:\n",
    "#     filename = filename.strip()\n",
    "#     source_path = '{}/{}'.format(clustal_in_dir_rem, filename)\n",
    "#     dest_filename = os.path.splitext(filename)[0]\n",
    "    \n",
    "#     # Copy files\n",
    "#     !scp {name}@{server}:~/hmm/clustal/hmmsearch/hmmsearch_clustal_{dest_filename}.txt {hmmsearch_dir_clustal}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Directory /Users/alina/HMM/results/hmms/hmmsearch/clustal already exists.\n",
      "Directory /Users/alina/HMM/results/hmms/hmmsearch/clustal/hmmsearch_df already exists.\n",
      "Error processing file hmmsearch_clustal_J8TM36_236-249.txt: '  ------ inclusion threshold ------\\n' is not in list\n",
      "Error processing file hmmsearch_clustal_O15922_315-324.txt: '  ------ inclusion threshold ------\\n' is not in list\n",
      "Error processing file hmmsearch_clustal_Q5T4W7_108-120.txt: '  ------ inclusion threshold ------\\n' is not in list\n",
      "The number of retrieved hmmsearch results for ClustalOmega: 211312\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>E-value</th>\n",
       "      <th>score</th>\n",
       "      <th>bias</th>\n",
       "      <th>E-value</th>\n",
       "      <th>score</th>\n",
       "      <th>bias</th>\n",
       "      <th>exp</th>\n",
       "      <th>N</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>hmm_from</th>\n",
       "      <th>hmm_to</th>\n",
       "      <th>hmm_length</th>\n",
       "      <th>ali_from</th>\n",
       "      <th>ali_to</th>\n",
       "      <th>ali_length</th>\n",
       "      <th>env_from</th>\n",
       "      <th>env_to</th>\n",
       "      <th>env_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3.8e-32</td>\n",
       "      <td>122.0</td>\n",
       "      <td>13.1</td>\n",
       "      <td>4.4e-32</td>\n",
       "      <td>121.8</td>\n",
       "      <td>13.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>A0A8C7BFT3</td>\n",
       "      <td>1</td>\n",
       "      <td>61</td>\n",
       "      <td>61</td>\n",
       "      <td>9</td>\n",
       "      <td>69</td>\n",
       "      <td>61</td>\n",
       "      <td>9</td>\n",
       "      <td>70</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.4e-31</td>\n",
       "      <td>119.5</td>\n",
       "      <td>13.6</td>\n",
       "      <td>9.9e-31</td>\n",
       "      <td>117.5</td>\n",
       "      <td>13.6</td>\n",
       "      <td>2.2</td>\n",
       "      <td>1</td>\n",
       "      <td>A0A3Q7NSQ2</td>\n",
       "      <td>1</td>\n",
       "      <td>62</td>\n",
       "      <td>62</td>\n",
       "      <td>352</td>\n",
       "      <td>413</td>\n",
       "      <td>62</td>\n",
       "      <td>352</td>\n",
       "      <td>413</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3e-31</td>\n",
       "      <td>119.2</td>\n",
       "      <td>13.6</td>\n",
       "      <td>1.4e-30</td>\n",
       "      <td>117.0</td>\n",
       "      <td>13.6</td>\n",
       "      <td>2.3</td>\n",
       "      <td>1</td>\n",
       "      <td>A0A2U3ZCX9</td>\n",
       "      <td>1</td>\n",
       "      <td>62</td>\n",
       "      <td>62</td>\n",
       "      <td>451</td>\n",
       "      <td>512</td>\n",
       "      <td>62</td>\n",
       "      <td>451</td>\n",
       "      <td>512</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3e-31</td>\n",
       "      <td>119.1</td>\n",
       "      <td>13.6</td>\n",
       "      <td>1.4e-30</td>\n",
       "      <td>117.0</td>\n",
       "      <td>13.6</td>\n",
       "      <td>2.2</td>\n",
       "      <td>1</td>\n",
       "      <td>A0A2U3VQT6</td>\n",
       "      <td>1</td>\n",
       "      <td>62</td>\n",
       "      <td>62</td>\n",
       "      <td>454</td>\n",
       "      <td>515</td>\n",
       "      <td>62</td>\n",
       "      <td>454</td>\n",
       "      <td>515</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3.1e-31</td>\n",
       "      <td>119.1</td>\n",
       "      <td>13.6</td>\n",
       "      <td>1.4e-30</td>\n",
       "      <td>117.0</td>\n",
       "      <td>13.6</td>\n",
       "      <td>2.2</td>\n",
       "      <td>1</td>\n",
       "      <td>A0A3Q7P1S3</td>\n",
       "      <td>1</td>\n",
       "      <td>62</td>\n",
       "      <td>62</td>\n",
       "      <td>451</td>\n",
       "      <td>512</td>\n",
       "      <td>62</td>\n",
       "      <td>451</td>\n",
       "      <td>512</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>3.3e-31</td>\n",
       "      <td>119.0</td>\n",
       "      <td>13.6</td>\n",
       "      <td>1.4e-30</td>\n",
       "      <td>117.0</td>\n",
       "      <td>13.6</td>\n",
       "      <td>2.2</td>\n",
       "      <td>1</td>\n",
       "      <td>A0A384CTJ6</td>\n",
       "      <td>1</td>\n",
       "      <td>62</td>\n",
       "      <td>62</td>\n",
       "      <td>452</td>\n",
       "      <td>513</td>\n",
       "      <td>62</td>\n",
       "      <td>452</td>\n",
       "      <td>513</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>3.3e-31</td>\n",
       "      <td>119.0</td>\n",
       "      <td>13.6</td>\n",
       "      <td>1.4e-30</td>\n",
       "      <td>117.0</td>\n",
       "      <td>13.6</td>\n",
       "      <td>2.2</td>\n",
       "      <td>1</td>\n",
       "      <td>A0A452RB54</td>\n",
       "      <td>1</td>\n",
       "      <td>62</td>\n",
       "      <td>62</td>\n",
       "      <td>452</td>\n",
       "      <td>513</td>\n",
       "      <td>62</td>\n",
       "      <td>452</td>\n",
       "      <td>513</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>3.3e-31</td>\n",
       "      <td>119.0</td>\n",
       "      <td>13.6</td>\n",
       "      <td>1.4e-30</td>\n",
       "      <td>117.0</td>\n",
       "      <td>13.6</td>\n",
       "      <td>2.2</td>\n",
       "      <td>1</td>\n",
       "      <td>A0A7N5P5S9</td>\n",
       "      <td>1</td>\n",
       "      <td>62</td>\n",
       "      <td>62</td>\n",
       "      <td>451</td>\n",
       "      <td>512</td>\n",
       "      <td>62</td>\n",
       "      <td>451</td>\n",
       "      <td>512</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>7.2e-31</td>\n",
       "      <td>117.9</td>\n",
       "      <td>13.1</td>\n",
       "      <td>3.3e-30</td>\n",
       "      <td>115.8</td>\n",
       "      <td>13.1</td>\n",
       "      <td>2.2</td>\n",
       "      <td>1</td>\n",
       "      <td>A0A2Y9JGP3</td>\n",
       "      <td>1</td>\n",
       "      <td>61</td>\n",
       "      <td>61</td>\n",
       "      <td>450</td>\n",
       "      <td>510</td>\n",
       "      <td>61</td>\n",
       "      <td>450</td>\n",
       "      <td>511</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1.2e-30</td>\n",
       "      <td>117.2</td>\n",
       "      <td>14.1</td>\n",
       "      <td>5.2e-30</td>\n",
       "      <td>115.2</td>\n",
       "      <td>14.1</td>\n",
       "      <td>2.2</td>\n",
       "      <td>1</td>\n",
       "      <td>G1RG64</td>\n",
       "      <td>1</td>\n",
       "      <td>62</td>\n",
       "      <td>62</td>\n",
       "      <td>446</td>\n",
       "      <td>507</td>\n",
       "      <td>62</td>\n",
       "      <td>446</td>\n",
       "      <td>507</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   E-value  score  bias  E-value  score  bias  exp  N    Sequence hmm_from  \\\n",
       "0  3.8e-32  122.0  13.1  4.4e-32  121.8  13.1  1.0  1  A0A8C7BFT3        1   \n",
       "1  2.4e-31  119.5  13.6  9.9e-31  117.5  13.6  2.2  1  A0A3Q7NSQ2        1   \n",
       "2    3e-31  119.2  13.6  1.4e-30  117.0  13.6  2.3  1  A0A2U3ZCX9        1   \n",
       "3    3e-31  119.1  13.6  1.4e-30  117.0  13.6  2.2  1  A0A2U3VQT6        1   \n",
       "4  3.1e-31  119.1  13.6  1.4e-30  117.0  13.6  2.2  1  A0A3Q7P1S3        1   \n",
       "5  3.3e-31  119.0  13.6  1.4e-30  117.0  13.6  2.2  1  A0A384CTJ6        1   \n",
       "6  3.3e-31  119.0  13.6  1.4e-30  117.0  13.6  2.2  1  A0A452RB54        1   \n",
       "7  3.3e-31  119.0  13.6  1.4e-30  117.0  13.6  2.2  1  A0A7N5P5S9        1   \n",
       "8  7.2e-31  117.9  13.1  3.3e-30  115.8  13.1  2.2  1  A0A2Y9JGP3        1   \n",
       "9  1.2e-30  117.2  14.1  5.2e-30  115.2  14.1  2.2  1      G1RG64        1   \n",
       "\n",
       "  hmm_to hmm_length ali_from ali_to ali_length env_from env_to env_length  \n",
       "0     61         61        9     69         61        9     70         62  \n",
       "1     62         62      352    413         62      352    413         62  \n",
       "2     62         62      451    512         62      451    512         62  \n",
       "3     62         62      454    515         62      454    515         62  \n",
       "4     62         62      451    512         62      451    512         62  \n",
       "5     62         62      452    513         62      452    513         62  \n",
       "6     62         62      452    513         62      452    513         62  \n",
       "7     62         62      451    512         62      451    512         62  \n",
       "8     61         61      450    510         61      450    511         62  \n",
       "9     62         62      446    507         62      446    507         62  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Preprocess hmmsearch results - ClustalOmega\n",
    "hmmsearch_dir_clustal = '{}/results/hmms/hmmsearch/clustal'.format(directory)\n",
    "hmmsearch_dir_clustal_prepr_output = '{}/results/hmms/hmmsearch/clustal/hmmsearch_df'.format(directory)\n",
    "create_directory(hmmsearch_dir_clustal)\n",
    "create_directory(hmmsearch_dir_clustal_prepr_output)\n",
    "\n",
    "clustal_results_list = []\n",
    "\n",
    "for filename in os.listdir(hmmsearch_dir_clustal):\n",
    "    if filename.endswith('.txt'):\n",
    "        file_path = os.path.join(hmmsearch_dir_clustal, filename)\n",
    "        parts = filename.split('_')\n",
    "        protein_id = '{}_{}'.format(parts[2], parts[3].split(\".\")[0])\n",
    "        try:\n",
    "            clustal_stat = process_hmmsearch_file(file_path) \n",
    "            clustal_reg = extract_table_from_output(file_path)\n",
    "            clustal_result = pd.merge(clustal_stat, clustal_reg, left_on='Sequence', right_on='id', how='inner')\n",
    "            clustal_result = clustal_result.drop(columns=['Description', 'id'])\n",
    "            \n",
    "            # Save individual DataFrames to a list\n",
    "            clustal_results_list.append(clustal_result)\n",
    "        \n",
    "            # Save the DataFrame to a CSV file\n",
    "            output_file = os.path.join(hmmsearch_dir_clustal_prepr_output, 'hmmsearch_clustal_df_{}.csv'.format(protein_id))\n",
    "            clustal_result.to_csv(output_file, index=False)\n",
    "            \n",
    "        except ValueError as e:\n",
    "            print('Error processing file {}: {}'.format(filename, e))\n",
    "            \n",
    "# Merge all DataFrames into one\n",
    "hmmsearch_results_clustal = pd.concat(clustal_results_list, ignore_index=True)\n",
    "\n",
    "# Save the merged DataFrame to a CSV file\n",
    "merged_output_file = os.path.join('{}/results/hmms/hmmsearch/'.format(directory), 'hmmsearch_results_clustal.csv')\n",
    "hmmsearch_results_clustal.to_csv(merged_output_file, index=False)\n",
    "print('The number of retrieved hmmsearch results for ClustalOmega: {}'.format(len(hmmsearch_results_clustal)))\n",
    "hmmsearch_results_clustal[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see that 3 regions of proteins do not have HMM matches in RP 75% database: **O15922_315-324, J8TM36_236-249, Q5T4W7_108-120**.\n",
    "\n",
    "The possible reason may lie in the size of the region - all of them are quite small, about 10-14 amino acids. By analysing other short disordered regions like **O43791_169-178** (the length of 10 amino acids) we can see that there are reported the results only below the threshold."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In terms of region we are interested in the 3 fields:\n",
    "- `hmm_from`-`hmm_to`: the endpoints of the hmm profile. In our case the beginning is usually 1, since we parse not the initial alignment as an input, but the separate disordered regions where the position mostly starts from 1 (in case there are no gaps in the subject sequences).\n",
    "- `ali_from`-`ali_to`: the endpoints of the target sequence. They are obtained by matching HMM to the sequence, thus obtaining the positions.\n",
    "- `env_from`-`env_to`: the envelope of the domain's location. It's usually a bit wider than the alignment."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Pfam\n",
    "In this part we will use the proteins obtained in the dataframes with the `hmmsearch` results to iterate over the Protein2ipr database to find which HMM models do already exist in Pfam."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-25T07:30:38.221338Z",
     "start_time": "2023-08-25T07:30:38.183994Z"
    }
   },
   "outputs": [],
   "source": [
    "# # Copy the files with the hmmsearch statistics to the remote computer\n",
    "# !scp {directory}/results/hmms/hmmsearch/hmmsearch_results_clustal.csv {name}@{server}:~/stats/hmmsearch_results_clustal.csv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The file `filtered.tsv.gz` is a shorter archive derived from `protein2ipr` database. **It contains only the domains from Interpro**. We check the overlap of the found results from `hmmsearch`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Check the overlaps with Interpro domains (ClustalOmega)\n",
    "# !ssh {name}@{server} '/home/alina/protein2ipr.py /home/alina/stats/clustal /home/alina/filtered.tsv.gz protein2ipr_clustal.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-25T07:40:00.248249Z",
     "start_time": "2023-08-25T07:40:00.231919Z"
    }
   },
   "outputs": [],
   "source": [
    "# # Copy the files to the local folder\n",
    "# !scp {name}@{server}:~/protein2ipr_clustal.tsv {directory}/results/pfam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Results preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of retrieved Pfam instances for ClustalOmega: 1073417\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>pfam_id</th>\n",
       "      <th>ipr_id</th>\n",
       "      <th>start_pfam</th>\n",
       "      <th>end_pfam</th>\n",
       "      <th>length_pfam</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF02775</td>\n",
       "      <td>IPR011766</td>\n",
       "      <td>499</td>\n",
       "      <td>646</td>\n",
       "      <td>148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF00205</td>\n",
       "      <td>IPR012000</td>\n",
       "      <td>289</td>\n",
       "      <td>434</td>\n",
       "      <td>146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF02776</td>\n",
       "      <td>IPR012001</td>\n",
       "      <td>91</td>\n",
       "      <td>205</td>\n",
       "      <td>115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A010Q7P7</td>\n",
       "      <td>PF00018</td>\n",
       "      <td>IPR001452</td>\n",
       "      <td>435</td>\n",
       "      <td>481</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A010Q7P7</td>\n",
       "      <td>PF03114</td>\n",
       "      <td>IPR004148</td>\n",
       "      <td>113</td>\n",
       "      <td>226</td>\n",
       "      <td>114</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   uniprot_id  pfam_id     ipr_id  start_pfam  end_pfam  length_pfam\n",
       "0  A0A010Q304  PF02775  IPR011766         499       646          148\n",
       "1  A0A010Q304  PF00205  IPR012000         289       434          146\n",
       "2  A0A010Q304  PF02776  IPR012001          91       205          115\n",
       "3  A0A010Q7P7  PF00018  IPR001452         435       481           47\n",
       "4  A0A010Q7P7  PF03114  IPR004148         113       226          114"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter only entries with Pfam ID and intercepting regions with the curated_disprot instances - ClustalOmega\n",
    "filename_clustal = '{}/results/pfam/protein2ipr_clustal.tsv'.format(directory)\n",
    "pfam_clustal = pfam_processing(filename_clustal)\n",
    "\n",
    "print('The number of retrieved Pfam instances for ClustalOmega: {}'.format(len(pfam_clustal)))\n",
    "pfam_clustal.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of retrieved Pfam instances for ClustalOmega: 1110208\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>pfam_id</th>\n",
       "      <th>ipr_id</th>\n",
       "      <th>start_pfam</th>\n",
       "      <th>end_pfam</th>\n",
       "      <th>length_pfam</th>\n",
       "      <th>hmm_from</th>\n",
       "      <th>hmm_to</th>\n",
       "      <th>hmm_length</th>\n",
       "      <th>ali_from</th>\n",
       "      <th>ali_to</th>\n",
       "      <th>ali_length</th>\n",
       "      <th>env_from</th>\n",
       "      <th>env_to</th>\n",
       "      <th>env_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF02775</td>\n",
       "      <td>IPR011766</td>\n",
       "      <td>499</td>\n",
       "      <td>646</td>\n",
       "      <td>148</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>583</td>\n",
       "      <td>598</td>\n",
       "      <td>16</td>\n",
       "      <td>583</td>\n",
       "      <td>598</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF02775</td>\n",
       "      <td>IPR011766</td>\n",
       "      <td>499</td>\n",
       "      <td>646</td>\n",
       "      <td>148</td>\n",
       "      <td>1</td>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "      <td>650</td>\n",
       "      <td>689</td>\n",
       "      <td>40</td>\n",
       "      <td>650</td>\n",
       "      <td>690</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF00205</td>\n",
       "      <td>IPR012000</td>\n",
       "      <td>289</td>\n",
       "      <td>434</td>\n",
       "      <td>146</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>583</td>\n",
       "      <td>598</td>\n",
       "      <td>16</td>\n",
       "      <td>583</td>\n",
       "      <td>598</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF00205</td>\n",
       "      <td>IPR012000</td>\n",
       "      <td>289</td>\n",
       "      <td>434</td>\n",
       "      <td>146</td>\n",
       "      <td>1</td>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "      <td>650</td>\n",
       "      <td>689</td>\n",
       "      <td>40</td>\n",
       "      <td>650</td>\n",
       "      <td>690</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF02776</td>\n",
       "      <td>IPR012001</td>\n",
       "      <td>91</td>\n",
       "      <td>205</td>\n",
       "      <td>115</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>583</td>\n",
       "      <td>598</td>\n",
       "      <td>16</td>\n",
       "      <td>583</td>\n",
       "      <td>598</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF02776</td>\n",
       "      <td>IPR012001</td>\n",
       "      <td>91</td>\n",
       "      <td>205</td>\n",
       "      <td>115</td>\n",
       "      <td>1</td>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "      <td>650</td>\n",
       "      <td>689</td>\n",
       "      <td>40</td>\n",
       "      <td>650</td>\n",
       "      <td>690</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A0A010Q7P7</td>\n",
       "      <td>PF00018</td>\n",
       "      <td>IPR001452</td>\n",
       "      <td>435</td>\n",
       "      <td>481</td>\n",
       "      <td>47</td>\n",
       "      <td>10</td>\n",
       "      <td>65</td>\n",
       "      <td>56</td>\n",
       "      <td>428</td>\n",
       "      <td>485</td>\n",
       "      <td>58</td>\n",
       "      <td>424</td>\n",
       "      <td>487</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>A0A010Q7P7</td>\n",
       "      <td>PF03114</td>\n",
       "      <td>IPR004148</td>\n",
       "      <td>113</td>\n",
       "      <td>226</td>\n",
       "      <td>114</td>\n",
       "      <td>10</td>\n",
       "      <td>65</td>\n",
       "      <td>56</td>\n",
       "      <td>428</td>\n",
       "      <td>485</td>\n",
       "      <td>58</td>\n",
       "      <td>424</td>\n",
       "      <td>487</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A0A010QBJ0</td>\n",
       "      <td>PF00018</td>\n",
       "      <td>IPR001452</td>\n",
       "      <td>1080</td>\n",
       "      <td>1125</td>\n",
       "      <td>46</td>\n",
       "      <td>18</td>\n",
       "      <td>68</td>\n",
       "      <td>51</td>\n",
       "      <td>1082</td>\n",
       "      <td>1132</td>\n",
       "      <td>51</td>\n",
       "      <td>1073</td>\n",
       "      <td>1135</td>\n",
       "      <td>63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>A0A010QBJ0</td>\n",
       "      <td>PF00063</td>\n",
       "      <td>IPR001609</td>\n",
       "      <td>44</td>\n",
       "      <td>702</td>\n",
       "      <td>659</td>\n",
       "      <td>18</td>\n",
       "      <td>68</td>\n",
       "      <td>51</td>\n",
       "      <td>1082</td>\n",
       "      <td>1132</td>\n",
       "      <td>51</td>\n",
       "      <td>1073</td>\n",
       "      <td>1135</td>\n",
       "      <td>63</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   uniprot_id  pfam_id     ipr_id  start_pfam  end_pfam  length_pfam hmm_from  \\\n",
       "0  A0A010Q304  PF02775  IPR011766         499       646          148        1   \n",
       "1  A0A010Q304  PF02775  IPR011766         499       646          148        1   \n",
       "2  A0A010Q304  PF00205  IPR012000         289       434          146        1   \n",
       "3  A0A010Q304  PF00205  IPR012000         289       434          146        1   \n",
       "4  A0A010Q304  PF02776  IPR012001          91       205          115        1   \n",
       "5  A0A010Q304  PF02776  IPR012001          91       205          115        1   \n",
       "6  A0A010Q7P7  PF00018  IPR001452         435       481           47       10   \n",
       "7  A0A010Q7P7  PF03114  IPR004148         113       226          114       10   \n",
       "8  A0A010QBJ0  PF00018  IPR001452        1080      1125           46       18   \n",
       "9  A0A010QBJ0  PF00063  IPR001609          44       702          659       18   \n",
       "\n",
       "  hmm_to hmm_length ali_from ali_to ali_length env_from env_to env_length  \n",
       "0     16         16      583    598         16      583    598         16  \n",
       "1     40         40      650    689         40      650    690         41  \n",
       "2     16         16      583    598         16      583    598         16  \n",
       "3     40         40      650    689         40      650    690         41  \n",
       "4     16         16      583    598         16      583    598         16  \n",
       "5     40         40      650    689         40      650    690         41  \n",
       "6     65         56      428    485         58      424    487         64  \n",
       "7     65         56      428    485         58      424    487         64  \n",
       "8     68         51     1082   1132         51     1073   1135         63  \n",
       "9     68         51     1082   1132         51     1073   1135         63  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge hmmsearch results with Pfam - ClustalOmega\n",
    "pfam_clustal = pd.merge(pfam_clustal, \n",
    "                        hmmsearch_results_clustal[['Sequence', \n",
    "                                                   'hmm_from', 'hmm_to', 'hmm_length',\n",
    "                                                   'ali_from', 'ali_to', 'ali_length',\n",
    "                                                   'env_from', 'env_to', 'env_length']], \n",
    "                        left_on='uniprot_id', right_on='Sequence', how='left')\n",
    "\n",
    "pfam_clustal = pfam_clustal.dropna(axis=0)\n",
    "pfam_clustal = pfam_clustal.drop(columns=['Sequence'])\n",
    "print('The number of retrieved Pfam instances for ClustalOmega: {}'.format(len(pfam_clustal)))\n",
    "pfam_clustal[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then we add the calculated overlaps in the dataframe using the function `pfam_hmm_overlap`.\n",
    "\n",
    "- `overl_len`: the common region of the DisProt-HMM and Pfam-HMM. We find the minimum of the end position both in Pfam and HMM and the maximum of start position. Their difference will give us the overlap.\n",
    "- `overlap_pfam`: % of overlap with Pfam only. We divide `overl_len` by the length of Pfam.\n",
    "- `overlap_ali`: % of overlap with HMM only. We divide `overl_len` by the length of HMM.\n",
    "- `non_overlap_len`: the non-covered region, the rest of the Pfam and HMM regions.\n",
    "- `overlap_perc`: % of covered region. Calculated by division of `overl_len` to the whole length of both Pfam and HMM."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>pfam_id</th>\n",
       "      <th>ipr_id</th>\n",
       "      <th>start_pfam</th>\n",
       "      <th>end_pfam</th>\n",
       "      <th>length_pfam</th>\n",
       "      <th>hmm_from</th>\n",
       "      <th>hmm_to</th>\n",
       "      <th>hmm_length</th>\n",
       "      <th>ali_from</th>\n",
       "      <th>...</th>\n",
       "      <th>ali_length</th>\n",
       "      <th>env_from</th>\n",
       "      <th>env_to</th>\n",
       "      <th>env_length</th>\n",
       "      <th>overl_len</th>\n",
       "      <th>overl_pfam</th>\n",
       "      <th>overl_ali</th>\n",
       "      <th>non_overl_len</th>\n",
       "      <th>overl_perc</th>\n",
       "      <th>overlap_sym</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF02775</td>\n",
       "      <td>IPR011766</td>\n",
       "      <td>499</td>\n",
       "      <td>646</td>\n",
       "      <td>148</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>583</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>583</td>\n",
       "      <td>598</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>10.81</td>\n",
       "      <td>100.0</td>\n",
       "      <td>133</td>\n",
       "      <td>10.81</td>\n",
       "      <td>19.51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF02775</td>\n",
       "      <td>IPR011766</td>\n",
       "      <td>499</td>\n",
       "      <td>646</td>\n",
       "      <td>148</td>\n",
       "      <td>1</td>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "      <td>650</td>\n",
       "      <td>...</td>\n",
       "      <td>40</td>\n",
       "      <td>650</td>\n",
       "      <td>690</td>\n",
       "      <td>41</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>195</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF00205</td>\n",
       "      <td>IPR012000</td>\n",
       "      <td>289</td>\n",
       "      <td>434</td>\n",
       "      <td>146</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>583</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>583</td>\n",
       "      <td>598</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>459</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF00205</td>\n",
       "      <td>IPR012000</td>\n",
       "      <td>289</td>\n",
       "      <td>434</td>\n",
       "      <td>146</td>\n",
       "      <td>1</td>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "      <td>650</td>\n",
       "      <td>...</td>\n",
       "      <td>40</td>\n",
       "      <td>650</td>\n",
       "      <td>690</td>\n",
       "      <td>41</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>617</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A010Q304</td>\n",
       "      <td>PF02776</td>\n",
       "      <td>IPR012001</td>\n",
       "      <td>91</td>\n",
       "      <td>205</td>\n",
       "      <td>115</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>583</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>583</td>\n",
       "      <td>598</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>886</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   uniprot_id  pfam_id     ipr_id  start_pfam  end_pfam  length_pfam hmm_from  \\\n",
       "0  A0A010Q304  PF02775  IPR011766         499       646          148        1   \n",
       "1  A0A010Q304  PF02775  IPR011766         499       646          148        1   \n",
       "2  A0A010Q304  PF00205  IPR012000         289       434          146        1   \n",
       "3  A0A010Q304  PF00205  IPR012000         289       434          146        1   \n",
       "4  A0A010Q304  PF02776  IPR012001          91       205          115        1   \n",
       "\n",
       "  hmm_to hmm_length ali_from  ... ali_length env_from env_to env_length  \\\n",
       "0     16         16      583  ...         16      583    598         16   \n",
       "1     40         40      650  ...         40      650    690         41   \n",
       "2     16         16      583  ...         16      583    598         16   \n",
       "3     40         40      650  ...         40      650    690         41   \n",
       "4     16         16      583  ...         16      583    598         16   \n",
       "\n",
       "  overl_len  overl_pfam  overl_ali  non_overl_len  overl_perc  overlap_sym  \n",
       "0        16       10.81      100.0            133       10.81        19.51  \n",
       "1         0        0.00        0.0            195        0.00         0.00  \n",
       "2         0        0.00        0.0            459        0.00         0.00  \n",
       "3         0        0.00        0.0            617        0.00         0.00  \n",
       "4         0        0.00        0.0            886        0.00         0.00  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add the overlaps to the dataframe - ClustalOmega\n",
    "overlap_pfam_hmm = []\n",
    "\n",
    "for index_pfam, row_pfam in pfam_clustal.iterrows():\n",
    "    overl_len, overl_pfam, overl_ali, non_overl_len, overl_perc, overlap_sym = pfam_hmm_overlap(row_pfam)\n",
    "    overl_pfam = round(overl_pfam, 2)\n",
    "    overl_ali = round(overl_ali, 2)\n",
    "    overl_perc = round(overl_perc, 2)\n",
    "    overlap_sym = round(overlap_sym, 2)\n",
    "    overlap_pfam_hmm.append((overl_len, overl_pfam, overl_ali, non_overl_len, overl_perc, overlap_sym))\n",
    "\n",
    "# Extract overlap_pfam and overlap_hmm from the list of tuples\n",
    "pfam_clustal['overl_len'], pfam_clustal['overl_pfam'], pfam_clustal['overl_ali'], pfam_clustal['non_overl_len'], pfam_clustal['overl_perc'], pfam_clustal['overlap_sym'] = zip(*overlap_pfam_hmm)\n",
    "pfam_clustal.to_csv(f'results/pfam/pfam_overlap/pfam_clustal.csv', index=False)\n",
    "pfam_clustal.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
